<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Sparse Lazy Array Format - High-performance single-cell data storage and analysis"><meta name=author content="Pavan Ramkumar"><link href=https://slaf-project.github.io/slaf/getting-started/quickstart/ rel=canonical><link rel=prev href=../..><link href=../../user-guide/how-slaf-works/ rel=next><link rel=icon href=../../assets/slaf-icon-transparent-dark-mono.svg><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>Quick Start - SLAF Documentation</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../assets/_mkdocstrings.css><link rel=stylesheet href=../../stylesheets/extra.css><link rel=stylesheet href=../../stylesheets/logo-theme.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config",""),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#quick-start class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="SLAF Documentation" class="md-header__button md-logo" aria-label="SLAF Documentation" data-md-component=logo> <img src=../../assets/slaf-icon-transparent-dark-mono.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> SLAF Documentation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Quick Start </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/slaf-project/slaf title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> slaf-project/slaf </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="SLAF Documentation" class="md-nav__button md-logo" aria-label="SLAF Documentation" data-md-component=logo> <img src=../../assets/slaf-icon-transparent-dark-mono.svg alt=logo> </a> SLAF Documentation </label> <div class=md-nav__source> <a href=https://github.com/slaf-project/slaf title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> slaf-project/slaf </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Quick Start </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Quick Start </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#installation class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=#your-first-slaf-experience class=md-nav__link> <span class=md-ellipsis> Your First SLAF Experience </span> </a> <nav class=md-nav aria-label="Your First SLAF Experience"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#option-1-create-a-synthetic-dataset-recommended-for-first-time class=md-nav__link> <span class=md-ellipsis> Option 1: Create a Synthetic Dataset (Recommended for First Time) </span> </a> </li> <li class=md-nav__item> <a href=#option-2-use-a-real-dataset-pbmc3k class=md-nav__link> <span class=md-ellipsis> Option 2: Use a Real Dataset (PBMC3K) </span> </a> </li> <li class=md-nav__item> <a href=#step-2-convert-to-slaf-format class=md-nav__link> <span class=md-ellipsis> Step 2: Convert to SLAF Format </span> </a> </li> <li class=md-nav__item> <a href=#step-3-load-and-explore-your-data class=md-nav__link> <span class=md-ellipsis> Step 3: Load and Explore Your Data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#your-first-sql-queries class=md-nav__link> <span class=md-ellipsis> Your First SQL Queries </span> </a> <nav class=md-nav aria-label="Your First SQL Queries"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#basic-queries class=md-nav__link> <span class=md-ellipsis> Basic Queries </span> </a> </li> <li class=md-nav__item> <a href=#advanced-queries-with-joins class=md-nav__link> <span class=md-ellipsis> Advanced Queries with Joins </span> </a> </li> <li class=md-nav__item> <a href=#complex-analysis class=md-nav__link> <span class=md-ellipsis> Complex Analysis </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#lazy-evaluation-with-anndata-integration class=md-nav__link> <span class=md-ellipsis> Lazy Evaluation with Anndata Integration </span> </a> </li> <li class=md-nav__item> <a href=#working-with-metadata-and-layers class=md-nav__link> <span class=md-ellipsis> Working with Metadata and Layers </span> </a> </li> <li class=md-nav__item> <a href=#lazy-scanpy-preprocessing class=md-nav__link> <span class=md-ellipsis> Lazy Scanpy Preprocessing </span> </a> <nav class=md-nav aria-label="Lazy Scanpy Preprocessing"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#chaining-transformations class=md-nav__link> <span class=md-ellipsis> Chaining Transformations </span> </a> </li> <li class=md-nav__item> <a href=#key-benefits class=md-nav__link> <span class=md-ellipsis> Key Benefits </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#efficient-filtering class=md-nav__link> <span class=md-ellipsis> Efficient Filtering </span> </a> </li> <li class=md-nav__item> <a href=#ml-training-with-dataloaders class=md-nav__link> <span class=md-ellipsis> ML Training with Dataloaders </span> </a> <nav class=md-nav aria-label="ML Training with Dataloaders"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#tokenization class=md-nav__link> <span class=md-ellipsis> Tokenization </span> </a> </li> <li class=md-nav__item> <a href=#dataloader-for-training class=md-nav__link> <span class=md-ellipsis> DataLoader for Training </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#using-the-slaf-cli class=md-nav__link> <span class=md-ellipsis> Using the SLAF CLI </span> </a> <nav class=md-nav aria-label="Using the SLAF CLI"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#getting-help class=md-nav__link> <span class=md-ellipsis> Getting Help </span> </a> </li> <li class=md-nav__item> <a href=#basic-commands class=md-nav__link> <span class=md-ellipsis> Basic Commands </span> </a> </li> <li class=md-nav__item> <a href=#common-use-cases class=md-nav__link> <span class=md-ellipsis> Common Use Cases </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex> <span class=md-ellipsis> User Guide </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> User Guide </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../user-guide/how-slaf-works/ class=md-nav__link> <span class=md-ellipsis> How SLAF Works </span> </a> </li> <li class=md-nav__item> <a href=../../user-guide/migrating-to-slaf/ class=md-nav__link> <span class=md-ellipsis> Migrating to SLAF </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex> <span class=md-ellipsis> Examples </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Examples </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../examples/getting-started/ class=md-nav__link> <span class=md-ellipsis> Getting Started </span> </a> </li> <li class=md-nav__item> <a href=../../examples/lazy-processing/ class=md-nav__link> <span class=md-ellipsis> Lazy Processing </span> </a> </li> <li class=md-nav__item> <a href=../../examples/ml-training/ class=md-nav__link> <span class=md-ellipsis> ML Training </span> </a> </li> <li class=md-nav__item> <a href=../../examples/sql-queries/ class=md-nav__link> <span class=md-ellipsis> SQL Queries </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex> <span class=md-ellipsis> Benchmarks </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Benchmarks </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../benchmarks/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../benchmarks/bioinformatics_benchmarks/ class=md-nav__link> <span class=md-ellipsis> For Bioinformaticians </span> </a> </li> <li class=md-nav__item> <a href=../../benchmarks/ml_benchmarks/ class=md-nav__link> <span class=md-ellipsis> For ML Engineers </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/core/ class=md-nav__link> <span class=md-ellipsis> Core </span> </a> </li> <li class=md-nav__item> <a href=../../api/data/ class=md-nav__link> <span class=md-ellipsis> Data </span> </a> </li> <li class=md-nav__item> <a href=../../api/integrations/ class=md-nav__link> <span class=md-ellipsis> Integrations </span> </a> </li> <li class=md-nav__item> <a href=../../api/ml/ class=md-nav__link> <span class=md-ellipsis> ML </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex> <span class=md-ellipsis> Development </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Development </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../development/contributing/ class=md-nav__link> <span class=md-ellipsis> For Contributors </span> </a> </li> <li class=md-nav__item> <a href=../../development/maintaining/ class=md-nav__link> <span class=md-ellipsis> For Maintainers </span> </a> </li> <li class=md-nav__item> <a href=../../development/benchmarks/ class=md-nav__link> <span class=md-ellipsis> Benchmarks </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex> <span class=md-ellipsis> Blog </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> Blog </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../blog/ class=md-nav__link> <span class=md-ellipsis> Blog Home </span> </a> </li> <li class=md-nav__item> <a href=../../blog/huggingface/ class=md-nav__link> <span class=md-ellipsis> SLAF on Hugging Face </span> </a> </li> <li class=md-nav__item> <a href=../../blog/introducing-slaf/ class=md-nav__link> <span class=md-ellipsis> Introducing SLAF </span> </a> </li> <li class=md-nav__item> <a href=../../blog/blazing-fast-dataloaders/ class=md-nav__link> <span class=md-ellipsis> Blazing Fast Dataloaders </span> </a> </li> <li class=md-nav__item> <a href=../../blog/blazing-fast-dataloaders-2/ class=md-nav__link> <span class=md-ellipsis> Mixture of Scanners </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#installation class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=#your-first-slaf-experience class=md-nav__link> <span class=md-ellipsis> Your First SLAF Experience </span> </a> <nav class=md-nav aria-label="Your First SLAF Experience"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#option-1-create-a-synthetic-dataset-recommended-for-first-time class=md-nav__link> <span class=md-ellipsis> Option 1: Create a Synthetic Dataset (Recommended for First Time) </span> </a> </li> <li class=md-nav__item> <a href=#option-2-use-a-real-dataset-pbmc3k class=md-nav__link> <span class=md-ellipsis> Option 2: Use a Real Dataset (PBMC3K) </span> </a> </li> <li class=md-nav__item> <a href=#step-2-convert-to-slaf-format class=md-nav__link> <span class=md-ellipsis> Step 2: Convert to SLAF Format </span> </a> </li> <li class=md-nav__item> <a href=#step-3-load-and-explore-your-data class=md-nav__link> <span class=md-ellipsis> Step 3: Load and Explore Your Data </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#your-first-sql-queries class=md-nav__link> <span class=md-ellipsis> Your First SQL Queries </span> </a> <nav class=md-nav aria-label="Your First SQL Queries"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#basic-queries class=md-nav__link> <span class=md-ellipsis> Basic Queries </span> </a> </li> <li class=md-nav__item> <a href=#advanced-queries-with-joins class=md-nav__link> <span class=md-ellipsis> Advanced Queries with Joins </span> </a> </li> <li class=md-nav__item> <a href=#complex-analysis class=md-nav__link> <span class=md-ellipsis> Complex Analysis </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#lazy-evaluation-with-anndata-integration class=md-nav__link> <span class=md-ellipsis> Lazy Evaluation with Anndata Integration </span> </a> </li> <li class=md-nav__item> <a href=#working-with-metadata-and-layers class=md-nav__link> <span class=md-ellipsis> Working with Metadata and Layers </span> </a> </li> <li class=md-nav__item> <a href=#lazy-scanpy-preprocessing class=md-nav__link> <span class=md-ellipsis> Lazy Scanpy Preprocessing </span> </a> <nav class=md-nav aria-label="Lazy Scanpy Preprocessing"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#chaining-transformations class=md-nav__link> <span class=md-ellipsis> Chaining Transformations </span> </a> </li> <li class=md-nav__item> <a href=#key-benefits class=md-nav__link> <span class=md-ellipsis> Key Benefits </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#efficient-filtering class=md-nav__link> <span class=md-ellipsis> Efficient Filtering </span> </a> </li> <li class=md-nav__item> <a href=#ml-training-with-dataloaders class=md-nav__link> <span class=md-ellipsis> ML Training with Dataloaders </span> </a> <nav class=md-nav aria-label="ML Training with Dataloaders"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#tokenization class=md-nav__link> <span class=md-ellipsis> Tokenization </span> </a> </li> <li class=md-nav__item> <a href=#dataloader-for-training class=md-nav__link> <span class=md-ellipsis> DataLoader for Training </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#using-the-slaf-cli class=md-nav__link> <span class=md-ellipsis> Using the SLAF CLI </span> </a> <nav class=md-nav aria-label="Using the SLAF CLI"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#getting-help class=md-nav__link> <span class=md-ellipsis> Getting Help </span> </a> </li> <li class=md-nav__item> <a href=#basic-commands class=md-nav__link> <span class=md-ellipsis> Basic Commands </span> </a> </li> <li class=md-nav__item> <a href=#common-use-cases class=md-nav__link> <span class=md-ellipsis> Common Use Cases </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#next-steps class=md-nav__link> <span class=md-ellipsis> Next Steps </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=quick-start>Quick Start<a class=headerlink href=#quick-start title="Permanent link">&para;</a></h1> <p>Get up and running with SLAF in minutes! Experience the joy of lightning-fast SQL queries on your single-cell data.</p> <h2 id=installation>Installation<a class=headerlink href=#installation title="Permanent link">&para;</a></h2> <p>Install SLAF using your preferred package manager:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1># Using uv (recommended)</span>
</span><span id=__span-0-2><a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>uv<span class=w> </span>add<span class=w> </span>slafdb
</span><span id=__span-0-3><a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a>
</span><span id=__span-0-4><a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=c1># Or pip</span>
</span><span id=__span-0-5><a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a>pip<span class=w> </span>install<span class=w> </span>slafdb
</span></code></pre></div> <p>For development dependencies (including documentation):</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1># Using uv (recommended)</span>
</span><span id=__span-1-2><a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>uv<span class=w> </span>pip<span class=w> </span>install<span class=w> </span>-e<span class=w> </span><span class=s2>&quot;.[docs,dev,test]&quot;</span>
</span><span id=__span-1-3><a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>
</span><span id=__span-1-4><a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=c1># Or using pip</span>
</span><span id=__span-1-5><a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a>pip<span class=w> </span>install<span class=w> </span>slafdb<span class=o>[</span>docs,dev,test<span class=o>]</span>
</span></code></pre></div> <h2 id=your-first-slaf-experience>Your First SLAF Experience<a class=headerlink href=#your-first-slaf-experience title="Permanent link">&para;</a></h2> <p>Let's simulate a single-cell dataset similar to what you'd get from a real experiment.</p> <h3 id=option-1-create-a-synthetic-dataset-recommended-for-first-time>Option 1: Create a Synthetic Dataset (Recommended for First Time)<a class=headerlink href=#option-1-create-a-synthetic-dataset-recommended-for-first-time title="Permanent link">&para;</a></h3> <p>This creates a realistic single-cell dataset from scratch:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=kn>import</span><span class=w> </span><span class=nn>numpy</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>np</span>
</span><span id=__span-2-2><a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=kn>import</span><span class=w> </span><span class=nn>polars</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>pl</span>
</span><span id=__span-2-3><a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a><span class=kn>import</span><span class=w> </span><span class=nn>scanpy</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>sc</span>
</span><span id=__span-2-4><a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a><span class=kn>from</span><span class=w> </span><span class=nn>scipy.sparse</span><span class=w> </span><span class=kn>import</span> <span class=n>csr_matrix</span>
</span><span id=__span-2-5><a id=__codelineno-2-5 name=__codelineno-2-5 href=#__codelineno-2-5></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf.data.converter</span><span class=w> </span><span class=kn>import</span> <span class=n>SLAFConverter</span>
</span><span id=__span-2-6><a id=__codelineno-2-6 name=__codelineno-2-6 href=#__codelineno-2-6></a>
</span><span id=__span-2-7><a id=__codelineno-2-7 name=__codelineno-2-7 href=#__codelineno-2-7></a><span class=c1># Set random seed for reproducible results</span>
</span><span id=__span-2-8><a id=__codelineno-2-8 name=__codelineno-2-8 href=#__codelineno-2-8></a><span class=n>np</span><span class=o>.</span><span class=n>random</span><span class=o>.</span><span class=n>seed</span><span class=p>(</span><span class=mi>42</span><span class=p>)</span>
</span><span id=__span-2-9><a id=__codelineno-2-9 name=__codelineno-2-9 href=#__codelineno-2-9></a>
</span><span id=__span-2-10><a id=__codelineno-2-10 name=__codelineno-2-10 href=#__codelineno-2-10></a><span class=c1># Create a realistic single-cell dataset</span>
</span><span id=__span-2-11><a id=__codelineno-2-11 name=__codelineno-2-11 href=#__codelineno-2-11></a><span class=n>n_cells</span><span class=p>,</span> <span class=n>n_genes</span> <span class=o>=</span> <span class=mi>1000</span><span class=p>,</span> <span class=mi>2000</span>
</span><span id=__span-2-12><a id=__codelineno-2-12 name=__codelineno-2-12 href=#__codelineno-2-12></a>
</span><span id=__span-2-13><a id=__codelineno-2-13 name=__codelineno-2-13 href=#__codelineno-2-13></a><span class=c1># Create sparse matrix with realistic sparsity</span>
</span><span id=__span-2-14><a id=__codelineno-2-14 name=__codelineno-2-14 href=#__codelineno-2-14></a><span class=n>density</span> <span class=o>=</span> <span class=mf>0.1</span>  <span class=c1># 10% sparsity - typical for single-cell data</span>
</span><span id=__span-2-15><a id=__codelineno-2-15 name=__codelineno-2-15 href=#__codelineno-2-15></a><span class=n>n_nonzero</span> <span class=o>=</span> <span class=nb>int</span><span class=p>(</span><span class=n>n_cells</span> <span class=o>*</span> <span class=n>n_genes</span> <span class=o>*</span> <span class=n>density</span><span class=p>)</span>
</span><span id=__span-2-16><a id=__codelineno-2-16 name=__codelineno-2-16 href=#__codelineno-2-16></a>
</span><span id=__span-2-17><a id=__codelineno-2-17 name=__codelineno-2-17 href=#__codelineno-2-17></a><span class=c1># Generate realistic expression data (log-normal distribution)</span>
</span><span id=__span-2-18><a id=__codelineno-2-18 name=__codelineno-2-18 href=#__codelineno-2-18></a><span class=n>data</span> <span class=o>=</span> <span class=n>np</span><span class=o>.</span><span class=n>random</span><span class=o>.</span><span class=n>lognormal</span><span class=p>(</span><span class=mi>0</span><span class=p>,</span> <span class=mi>1</span><span class=p>,</span> <span class=n>n_nonzero</span><span class=p>)</span><span class=o>.</span><span class=n>astype</span><span class=p>(</span><span class=n>np</span><span class=o>.</span><span class=n>float32</span><span class=p>)</span>
</span><span id=__span-2-19><a id=__codelineno-2-19 name=__codelineno-2-19 href=#__codelineno-2-19></a><span class=n>row_indices</span> <span class=o>=</span> <span class=n>np</span><span class=o>.</span><span class=n>random</span><span class=o>.</span><span class=n>randint</span><span class=p>(</span><span class=mi>0</span><span class=p>,</span> <span class=n>n_cells</span><span class=p>,</span> <span class=n>n_nonzero</span><span class=p>)</span>
</span><span id=__span-2-20><a id=__codelineno-2-20 name=__codelineno-2-20 href=#__codelineno-2-20></a><span class=n>col_indices</span> <span class=o>=</span> <span class=n>np</span><span class=o>.</span><span class=n>random</span><span class=o>.</span><span class=n>randint</span><span class=p>(</span><span class=mi>0</span><span class=p>,</span> <span class=n>n_genes</span><span class=p>,</span> <span class=n>n_nonzero</span><span class=p>)</span>
</span><span id=__span-2-21><a id=__codelineno-2-21 name=__codelineno-2-21 href=#__codelineno-2-21></a>
</span><span id=__span-2-22><a id=__codelineno-2-22 name=__codelineno-2-22 href=#__codelineno-2-22></a><span class=c1># Create sparse matrix</span>
</span><span id=__span-2-23><a id=__codelineno-2-23 name=__codelineno-2-23 href=#__codelineno-2-23></a><span class=n>X</span> <span class=o>=</span> <span class=n>csr_matrix</span><span class=p>((</span><span class=n>data</span><span class=p>,</span> <span class=p>(</span><span class=n>row_indices</span><span class=p>,</span> <span class=n>col_indices</span><span class=p>)),</span> <span class=n>shape</span><span class=o>=</span><span class=p>(</span><span class=n>n_cells</span><span class=p>,</span> <span class=n>n_genes</span><span class=p>))</span>
</span><span id=__span-2-24><a id=__codelineno-2-24 name=__codelineno-2-24 href=#__codelineno-2-24></a>
</span><span id=__span-2-25><a id=__codelineno-2-25 name=__codelineno-2-25 href=#__codelineno-2-25></a><span class=c1># Create cell metadata</span>
</span><span id=__span-2-26><a id=__codelineno-2-26 name=__codelineno-2-26 href=#__codelineno-2-26></a><span class=n>obs</span> <span class=o>=</span> <span class=n>pl</span><span class=o>.</span><span class=n>DataFrame</span><span class=p>({</span>
</span><span id=__span-2-27><a id=__codelineno-2-27 name=__codelineno-2-27 href=#__codelineno-2-27></a>    <span class=s2>&quot;cell_id&quot;</span><span class=p>:</span> <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;cell_</span><span class=si>{</span><span class=n>i</span><span class=si>:</span><span class=s2>04d</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=nb>range</span><span class=p>(</span><span class=n>n_cells</span><span class=p>)],</span>
</span><span id=__span-2-28><a id=__codelineno-2-28 name=__codelineno-2-28 href=#__codelineno-2-28></a>    <span class=s2>&quot;cell_type&quot;</span><span class=p>:</span> <span class=n>np</span><span class=o>.</span><span class=n>random</span><span class=o>.</span><span class=n>choice</span><span class=p>([</span><span class=s2>&quot;T_cell&quot;</span><span class=p>,</span> <span class=s2>&quot;B_cell&quot;</span><span class=p>,</span> <span class=s2>&quot;NK_cell&quot;</span><span class=p>,</span> <span class=s2>&quot;Monocyte&quot;</span><span class=p>],</span> <span class=n>n_cells</span><span class=p>),</span>
</span><span id=__span-2-29><a id=__codelineno-2-29 name=__codelineno-2-29 href=#__codelineno-2-29></a>    <span class=s2>&quot;batch&quot;</span><span class=p>:</span> <span class=n>np</span><span class=o>.</span><span class=n>random</span><span class=o>.</span><span class=n>choice</span><span class=p>([</span><span class=s2>&quot;batch1&quot;</span><span class=p>,</span> <span class=s2>&quot;batch2&quot;</span><span class=p>,</span> <span class=s2>&quot;batch3&quot;</span><span class=p>],</span> <span class=n>n_cells</span><span class=p>),</span>
</span><span id=__span-2-30><a id=__codelineno-2-30 name=__codelineno-2-30 href=#__codelineno-2-30></a>    <span class=s2>&quot;total_counts&quot;</span><span class=p>:</span> <span class=n>X</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>A1</span><span class=p>,</span>
</span><span id=__span-2-31><a id=__codelineno-2-31 name=__codelineno-2-31 href=#__codelineno-2-31></a>    <span class=s2>&quot;n_genes_by_counts&quot;</span><span class=p>:</span> <span class=p>(</span><span class=n>X</span> <span class=o>&gt;</span> <span class=mi>0</span><span class=p>)</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>A1</span><span class=p>,</span>
</span><span id=__span-2-32><a id=__codelineno-2-32 name=__codelineno-2-32 href=#__codelineno-2-32></a>    <span class=s2>&quot;high_mito&quot;</span><span class=p>:</span> <span class=n>np</span><span class=o>.</span><span class=n>random</span><span class=o>.</span><span class=n>choice</span><span class=p>([</span><span class=kc>True</span><span class=p>,</span> <span class=kc>False</span><span class=p>],</span> <span class=n>n_cells</span><span class=p>,</span> <span class=n>p</span><span class=o>=</span><span class=p>[</span><span class=mf>0.1</span><span class=p>,</span> <span class=mf>0.9</span><span class=p>]),</span>
</span><span id=__span-2-33><a id=__codelineno-2-33 name=__codelineno-2-33 href=#__codelineno-2-33></a><span class=p>})</span>
</span><span id=__span-2-34><a id=__codelineno-2-34 name=__codelineno-2-34 href=#__codelineno-2-34></a>
</span><span id=__span-2-35><a id=__codelineno-2-35 name=__codelineno-2-35 href=#__codelineno-2-35></a><span class=c1># Create gene metadata</span>
</span><span id=__span-2-36><a id=__codelineno-2-36 name=__codelineno-2-36 href=#__codelineno-2-36></a><span class=n>var</span> <span class=o>=</span> <span class=n>pl</span><span class=o>.</span><span class=n>DataFrame</span><span class=p>({</span>
</span><span id=__span-2-37><a id=__codelineno-2-37 name=__codelineno-2-37 href=#__codelineno-2-37></a>    <span class=s2>&quot;gene_id&quot;</span><span class=p>:</span> <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;ENSG_</span><span class=si>{</span><span class=n>i</span><span class=si>:</span><span class=s2>08d</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=nb>range</span><span class=p>(</span><span class=n>n_genes</span><span class=p>)],</span>
</span><span id=__span-2-38><a id=__codelineno-2-38 name=__codelineno-2-38 href=#__codelineno-2-38></a>    <span class=s2>&quot;gene_symbol&quot;</span><span class=p>:</span> <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;GENE_</span><span class=si>{</span><span class=n>i</span><span class=si>:</span><span class=s2>06d</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=nb>range</span><span class=p>(</span><span class=n>n_genes</span><span class=p>)],</span>
</span><span id=__span-2-39><a id=__codelineno-2-39 name=__codelineno-2-39 href=#__codelineno-2-39></a>    <span class=s2>&quot;highly_variable&quot;</span><span class=p>:</span> <span class=n>np</span><span class=o>.</span><span class=n>random</span><span class=o>.</span><span class=n>choice</span><span class=p>([</span><span class=kc>True</span><span class=p>,</span> <span class=kc>False</span><span class=p>],</span> <span class=n>n_genes</span><span class=p>,</span> <span class=n>p</span><span class=o>=</span><span class=p>[</span><span class=mf>0.2</span><span class=p>,</span> <span class=mf>0.8</span><span class=p>]),</span>
</span><span id=__span-2-40><a id=__codelineno-2-40 name=__codelineno-2-40 href=#__codelineno-2-40></a>    <span class=s2>&quot;total_counts&quot;</span><span class=p>:</span> <span class=n>X</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>0</span><span class=p>)</span><span class=o>.</span><span class=n>A1</span><span class=p>,</span>
</span><span id=__span-2-41><a id=__codelineno-2-41 name=__codelineno-2-41 href=#__codelineno-2-41></a>    <span class=s2>&quot;n_cells_by_counts&quot;</span><span class=p>:</span> <span class=p>(</span><span class=n>X</span> <span class=o>&gt;</span> <span class=mi>0</span><span class=p>)</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>0</span><span class=p>)</span><span class=o>.</span><span class=n>A1</span><span class=p>,</span>
</span><span id=__span-2-42><a id=__codelineno-2-42 name=__codelineno-2-42 href=#__codelineno-2-42></a><span class=p>})</span>
</span><span id=__span-2-43><a id=__codelineno-2-43 name=__codelineno-2-43 href=#__codelineno-2-43></a>
</span><span id=__span-2-44><a id=__codelineno-2-44 name=__codelineno-2-44 href=#__codelineno-2-44></a><span class=c1># Create AnnData object</span>
</span><span id=__span-2-45><a id=__codelineno-2-45 name=__codelineno-2-45 href=#__codelineno-2-45></a><span class=n>adata</span> <span class=o>=</span> <span class=n>sc</span><span class=o>.</span><span class=n>AnnData</span><span class=p>(</span><span class=n>X</span><span class=o>=</span><span class=n>X</span><span class=p>,</span> <span class=n>obs</span><span class=o>=</span><span class=n>obs</span><span class=p>,</span> <span class=n>var</span><span class=o>=</span><span class=n>var</span><span class=p>)</span>
</span><span id=__span-2-46><a id=__codelineno-2-46 name=__codelineno-2-46 href=#__codelineno-2-46></a>
</span><span id=__span-2-47><a id=__codelineno-2-47 name=__codelineno-2-47 href=#__codelineno-2-47></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;✅ Created dataset: </span><span class=si>{</span><span class=n>adata</span><span class=o>.</span><span class=n>n_obs</span><span class=si>:</span><span class=s2>,</span><span class=si>}</span><span class=s2> cells × </span><span class=si>{</span><span class=n>adata</span><span class=o>.</span><span class=n>n_vars</span><span class=si>:</span><span class=s2>,</span><span class=si>}</span><span class=s2> genes&quot;</span><span class=p>)</span>
</span><span id=__span-2-48><a id=__codelineno-2-48 name=__codelineno-2-48 href=#__codelineno-2-48></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;   Sparsity: </span><span class=si>{</span><span class=mi>1</span><span class=w> </span><span class=o>-</span><span class=w> </span><span class=n>adata</span><span class=o>.</span><span class=n>X</span><span class=o>.</span><span class=n>nnz</span><span class=w> </span><span class=o>/</span><span class=w> </span><span class=p>(</span><span class=n>adata</span><span class=o>.</span><span class=n>n_obs</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=n>adata</span><span class=o>.</span><span class=n>n_vars</span><span class=p>)</span><span class=si>:</span><span class=s2>.1%</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span></code></pre></div> <h3 id=option-2-use-a-real-dataset-pbmc3k>Option 2: Use a Real Dataset (PBMC3K)<a class=headerlink href=#option-2-use-a-real-dataset-pbmc3k title="Permanent link">&para;</a></h3> <p>For a more authentic experience, you can use the popular PBMC3K dataset:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-3-1><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=kn>import</span><span class=w> </span><span class=nn>scanpy</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>sc</span>
</span><span id=__span-3-2><a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf.data.converter</span><span class=w> </span><span class=kn>import</span> <span class=n>SLAFConverter</span>
</span><span id=__span-3-3><a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a>
</span><span id=__span-3-4><a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a><span class=c1># Download PBMC3K dataset (this will take a moment)</span>
</span><span id=__span-3-5><a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Downloading PBMC3K dataset...&quot;</span><span class=p>)</span>
</span><span id=__span-3-6><a id=__codelineno-3-6 name=__codelineno-3-6 href=#__codelineno-3-6></a><span class=n>adata</span> <span class=o>=</span> <span class=n>sc</span><span class=o>.</span><span class=n>datasets</span><span class=o>.</span><span class=n>pbmc3k</span><span class=p>()</span>
</span><span id=__span-3-7><a id=__codelineno-3-7 name=__codelineno-3-7 href=#__codelineno-3-7></a><span class=n>adata</span><span class=o>.</span><span class=n>var_names_make_unique</span><span class=p>()</span>
</span><span id=__span-3-8><a id=__codelineno-3-8 name=__codelineno-3-8 href=#__codelineno-3-8></a>
</span><span id=__span-3-9><a id=__codelineno-3-9 name=__codelineno-3-9 href=#__codelineno-3-9></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;✅ Downloaded PBMC3K: </span><span class=si>{</span><span class=n>adata</span><span class=o>.</span><span class=n>n_obs</span><span class=si>:</span><span class=s2>,</span><span class=si>}</span><span class=s2> cells × </span><span class=si>{</span><span class=n>adata</span><span class=o>.</span><span class=n>n_vars</span><span class=si>:</span><span class=s2>,</span><span class=si>}</span><span class=s2> genes&quot;</span><span class=p>)</span>
</span><span id=__span-3-10><a id=__codelineno-3-10 name=__codelineno-3-10 href=#__codelineno-3-10></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;   Sparsity: </span><span class=si>{</span><span class=mi>1</span><span class=w> </span><span class=o>-</span><span class=w> </span><span class=n>adata</span><span class=o>.</span><span class=n>X</span><span class=o>.</span><span class=n>nnz</span><span class=w> </span><span class=o>/</span><span class=w> </span><span class=p>(</span><span class=n>adata</span><span class=o>.</span><span class=n>n_obs</span><span class=w> </span><span class=o>*</span><span class=w> </span><span class=n>adata</span><span class=o>.</span><span class=n>n_vars</span><span class=p>)</span><span class=si>:</span><span class=s2>.1%</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-3-11><a id=__codelineno-3-11 name=__codelineno-3-11 href=#__codelineno-3-11></a>
</span><span id=__span-3-12><a id=__codelineno-3-12 name=__codelineno-3-12 href=#__codelineno-3-12></a><span class=c1># Add some basic metadata if not present</span>
</span><span id=__span-3-13><a id=__codelineno-3-13 name=__codelineno-3-13 href=#__codelineno-3-13></a><span class=k>if</span> <span class=s2>&quot;total_counts&quot;</span> <span class=ow>not</span> <span class=ow>in</span> <span class=n>adata</span><span class=o>.</span><span class=n>obs</span><span class=o>.</span><span class=n>columns</span><span class=p>:</span>
</span><span id=__span-3-14><a id=__codelineno-3-14 name=__codelineno-3-14 href=#__codelineno-3-14></a>    <span class=n>adata</span><span class=o>.</span><span class=n>obs</span><span class=p>[</span><span class=s2>&quot;total_counts&quot;</span><span class=p>]</span> <span class=o>=</span> <span class=n>adata</span><span class=o>.</span><span class=n>X</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>A1</span>
</span><span id=__span-3-15><a id=__codelineno-3-15 name=__codelineno-3-15 href=#__codelineno-3-15></a><span class=k>if</span> <span class=s2>&quot;n_genes_by_counts&quot;</span> <span class=ow>not</span> <span class=ow>in</span> <span class=n>adata</span><span class=o>.</span><span class=n>obs</span><span class=o>.</span><span class=n>columns</span><span class=p>:</span>
</span><span id=__span-3-16><a id=__codelineno-3-16 name=__codelineno-3-16 href=#__codelineno-3-16></a>    <span class=n>adata</span><span class=o>.</span><span class=n>obs</span><span class=p>[</span><span class=s2>&quot;n_genes_by_counts&quot;</span><span class=p>]</span> <span class=o>=</span> <span class=p>(</span><span class=n>adata</span><span class=o>.</span><span class=n>X</span> <span class=o>&gt;</span> <span class=mi>0</span><span class=p>)</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>A1</span>
</span><span id=__span-3-17><a id=__codelineno-3-17 name=__codelineno-3-17 href=#__codelineno-3-17></a><span class=k>if</span> <span class=s2>&quot;batch&quot;</span> <span class=ow>not</span> <span class=ow>in</span> <span class=n>adata</span><span class=o>.</span><span class=n>obs</span><span class=o>.</span><span class=n>columns</span><span class=p>:</span>
</span><span id=__span-3-18><a id=__codelineno-3-18 name=__codelineno-3-18 href=#__codelineno-3-18></a>    <span class=n>adata</span><span class=o>.</span><span class=n>obs</span><span class=p>[</span><span class=s2>&quot;batch&quot;</span><span class=p>]</span> <span class=o>=</span> <span class=s2>&quot;pbmc3k&quot;</span>
</span></code></pre></div> <h3 id=step-2-convert-to-slaf-format>Step 2: Convert to SLAF Format<a class=headerlink href=#step-2-convert-to-slaf-format title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-4-1><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1># Convert to SLAF format</span>
</span><span id=__span-4-2><a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=n>converter</span> <span class=o>=</span> <span class=n>SLAFConverter</span><span class=p>()</span>
</span><span id=__span-4-3><a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a><span class=n>slaf_path</span> <span class=o>=</span> <span class=s2>&quot;my_dataset.slaf&quot;</span>
</span><span id=__span-4-4><a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a><span class=n>converter</span><span class=o>.</span><span class=n>convert_anndata</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span> <span class=n>slaf_path</span><span class=p>)</span>
</span><span id=__span-4-5><a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a>
</span><span id=__span-4-6><a id=__codelineno-4-6 name=__codelineno-4-6 href=#__codelineno-4-6></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;✅ Converted to SLAF format: </span><span class=si>{</span><span class=n>slaf_path</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span></code></pre></div> <h3 id=step-3-load-and-explore-your-data>Step 3: Load and Explore Your Data<a class=headerlink href=#step-3-load-and-explore-your-data title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-5-1><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=kn>import</span><span class=w> </span><span class=nn>slaf</span>
</span><span id=__span-5-2><a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a>
</span><span id=__span-5-3><a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a><span class=c1># Load your SLAF dataset</span>
</span><span id=__span-5-4><a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=n>slaf_array</span> <span class=o>=</span> <span class=n>slaf</span><span class=o>.</span><span class=n>SLAFArray</span><span class=p>(</span><span class=n>slaf_path</span><span class=p>)</span>
</span><span id=__span-5-5><a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a>
</span><span id=__span-5-6><a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a><span class=c1># Check basic info</span>
</span><span id=__span-5-7><a id=__codelineno-5-7 name=__codelineno-5-7 href=#__codelineno-5-7></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Dataset shape: </span><span class=si>{</span><span class=n>slaf_array</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-5-8><a id=__codelineno-5-8 name=__codelineno-5-8 href=#__codelineno-5-8></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Number of cells: </span><span class=si>{</span><span class=n>slaf_array</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>0</span><span class=p>]</span><span class=si>:</span><span class=s2>,</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-5-9><a id=__codelineno-5-9 name=__codelineno-5-9 href=#__codelineno-5-9></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Number of genes: </span><span class=si>{</span><span class=n>slaf_array</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>1</span><span class=p>]</span><span class=si>:</span><span class=s2>,</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-5-10><a id=__codelineno-5-10 name=__codelineno-5-10 href=#__codelineno-5-10></a>
</span><span id=__span-5-11><a id=__codelineno-5-11 name=__codelineno-5-11 href=#__codelineno-5-11></a><span class=c1># View cell metadata</span>
</span><span id=__span-5-12><a id=__codelineno-5-12 name=__codelineno-5-12 href=#__codelineno-5-12></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;</span><span class=se>\n</span><span class=s2>Cell metadata columns:&quot;</span><span class=p>)</span>
</span><span id=__span-5-13><a id=__codelineno-5-13 name=__codelineno-5-13 href=#__codelineno-5-13></a><span class=nb>print</span><span class=p>(</span><span class=nb>list</span><span class=p>(</span><span class=n>slaf_array</span><span class=o>.</span><span class=n>obs</span><span class=o>.</span><span class=n>columns</span><span class=p>))</span>
</span><span id=__span-5-14><a id=__codelineno-5-14 name=__codelineno-5-14 href=#__codelineno-5-14></a>
</span><span id=__span-5-15><a id=__codelineno-5-15 name=__codelineno-5-15 href=#__codelineno-5-15></a><span class=c1># View gene metadata</span>
</span><span id=__span-5-16><a id=__codelineno-5-16 name=__codelineno-5-16 href=#__codelineno-5-16></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;</span><span class=se>\n</span><span class=s2>Gene metadata columns:&quot;</span><span class=p>)</span>
</span><span id=__span-5-17><a id=__codelineno-5-17 name=__codelineno-5-17 href=#__codelineno-5-17></a><span class=nb>print</span><span class=p>(</span><span class=nb>list</span><span class=p>(</span><span class=n>slaf_array</span><span class=o>.</span><span class=n>var</span><span class=o>.</span><span class=n>columns</span><span class=p>))</span>
</span></code></pre></div> <h2 id=your-first-sql-queries>Your First SQL Queries<a class=headerlink href=#your-first-sql-queries title="Permanent link">&para;</a></h2> <p>Now experience the power of SQL on your single-cell data!</p> <h3 id=basic-queries>Basic Queries<a class=headerlink href=#basic-queries title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-6-1><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=c1># Count cells by cell type</span>
</span><span id=__span-6-2><a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=n>cell_types</span> <span class=o>=</span> <span class=n>slaf_array</span><span class=o>.</span><span class=n>query</span><span class=p>(</span><span class=s2>&quot;&quot;&quot;</span>
</span><span id=__span-6-3><a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a><span class=s2>    SELECT cell_type, COUNT(*) as count</span>
</span><span id=__span-6-4><a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a><span class=s2>    FROM cells</span>
</span><span id=__span-6-5><a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a><span class=s2>    GROUP BY cell_type</span>
</span><span id=__span-6-6><a id=__codelineno-6-6 name=__codelineno-6-6 href=#__codelineno-6-6></a><span class=s2>    ORDER BY count DESC</span>
</span><span id=__span-6-7><a id=__codelineno-6-7 name=__codelineno-6-7 href=#__codelineno-6-7></a><span class=s2>&quot;&quot;&quot;</span><span class=p>)</span>
</span><span id=__span-6-8><a id=__codelineno-6-8 name=__codelineno-6-8 href=#__codelineno-6-8></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Cell type distribution:&quot;</span><span class=p>)</span>
</span><span id=__span-6-9><a id=__codelineno-6-9 name=__codelineno-6-9 href=#__codelineno-6-9></a><span class=nb>print</span><span class=p>(</span><span class=n>cell_types</span><span class=p>)</span>
</span><span id=__span-6-10><a id=__codelineno-6-10 name=__codelineno-6-10 href=#__codelineno-6-10></a>
</span><span id=__span-6-11><a id=__codelineno-6-11 name=__codelineno-6-11 href=#__codelineno-6-11></a><span class=c1># Find cells with high total counts</span>
</span><span id=__span-6-12><a id=__codelineno-6-12 name=__codelineno-6-12 href=#__codelineno-6-12></a><span class=n>high_count_cells</span> <span class=o>=</span> <span class=n>slaf_array</span><span class=o>.</span><span class=n>query</span><span class=p>(</span><span class=s2>&quot;&quot;&quot;</span>
</span><span id=__span-6-13><a id=__codelineno-6-13 name=__codelineno-6-13 href=#__codelineno-6-13></a><span class=s2>    SELECT cell_id, cell_type, total_counts</span>
</span><span id=__span-6-14><a id=__codelineno-6-14 name=__codelineno-6-14 href=#__codelineno-6-14></a><span class=s2>    FROM cells</span>
</span><span id=__span-6-15><a id=__codelineno-6-15 name=__codelineno-6-15 href=#__codelineno-6-15></a><span class=s2>    WHERE total_counts &gt; 1000</span>
</span><span id=__span-6-16><a id=__codelineno-6-16 name=__codelineno-6-16 href=#__codelineno-6-16></a><span class=s2>    ORDER BY total_counts DESC</span>
</span><span id=__span-6-17><a id=__codelineno-6-17 name=__codelineno-6-17 href=#__codelineno-6-17></a><span class=s2>    LIMIT 5</span>
</span><span id=__span-6-18><a id=__codelineno-6-18 name=__codelineno-6-18 href=#__codelineno-6-18></a><span class=s2>&quot;&quot;&quot;</span><span class=p>)</span>
</span><span id=__span-6-19><a id=__codelineno-6-19 name=__codelineno-6-19 href=#__codelineno-6-19></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;</span><span class=se>\n</span><span class=s2>Cells with high total counts:&quot;</span><span class=p>)</span>
</span><span id=__span-6-20><a id=__codelineno-6-20 name=__codelineno-6-20 href=#__codelineno-6-20></a><span class=nb>print</span><span class=p>(</span><span class=n>high_count_cells</span><span class=p>)</span>
</span></code></pre></div> <h3 id=advanced-queries-with-joins>Advanced Queries with Joins<a class=headerlink href=#advanced-queries-with-joins title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-7-1><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># Find highly variable genes with their expression stats</span>
</span><span id=__span-7-2><a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=n>variable_genes</span> <span class=o>=</span> <span class=n>slaf_array</span><span class=o>.</span><span class=n>query</span><span class=p>(</span><span class=s2>&quot;&quot;&quot;</span>
</span><span id=__span-7-3><a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=s2>    SELECT</span>
</span><span id=__span-7-4><a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a><span class=s2>        g.gene_id,</span>
</span><span id=__span-7-5><a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a><span class=s2>        g.gene_symbol,</span>
</span><span id=__span-7-6><a id=__codelineno-7-6 name=__codelineno-7-6 href=#__codelineno-7-6></a><span class=s2>        g.total_counts as gene_total_counts,</span>
</span><span id=__span-7-7><a id=__codelineno-7-7 name=__codelineno-7-7 href=#__codelineno-7-7></a><span class=s2>        COUNT(e.value) as cells_expressed,</span>
</span><span id=__span-7-8><a id=__codelineno-7-8 name=__codelineno-7-8 href=#__codelineno-7-8></a><span class=s2>        AVG(e.value) as avg_expression,</span>
</span><span id=__span-7-9><a id=__codelineno-7-9 name=__codelineno-7-9 href=#__codelineno-7-9></a><span class=s2>        MAX(e.value) as max_expression</span>
</span><span id=__span-7-10><a id=__codelineno-7-10 name=__codelineno-7-10 href=#__codelineno-7-10></a><span class=s2>    FROM genes g</span>
</span><span id=__span-7-11><a id=__codelineno-7-11 name=__codelineno-7-11 href=#__codelineno-7-11></a><span class=s2>    LEFT JOIN expression e ON g.gene_integer_id = e.gene_integer_id</span>
</span><span id=__span-7-12><a id=__codelineno-7-12 name=__codelineno-7-12 href=#__codelineno-7-12></a><span class=s2>    WHERE g.highly_variable = true</span>
</span><span id=__span-7-13><a id=__codelineno-7-13 name=__codelineno-7-13 href=#__codelineno-7-13></a><span class=s2>    GROUP BY g.gene_id, g.gene_symbol, g.total_counts</span>
</span><span id=__span-7-14><a id=__codelineno-7-14 name=__codelineno-7-14 href=#__codelineno-7-14></a><span class=s2>    ORDER BY g.total_counts DESC</span>
</span><span id=__span-7-15><a id=__codelineno-7-15 name=__codelineno-7-15 href=#__codelineno-7-15></a><span class=s2>    LIMIT 10</span>
</span><span id=__span-7-16><a id=__codelineno-7-16 name=__codelineno-7-16 href=#__codelineno-7-16></a><span class=s2>&quot;&quot;&quot;</span><span class=p>)</span>
</span><span id=__span-7-17><a id=__codelineno-7-17 name=__codelineno-7-17 href=#__codelineno-7-17></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Top highly variable genes:&quot;</span><span class=p>)</span>
</span><span id=__span-7-18><a id=__codelineno-7-18 name=__codelineno-7-18 href=#__codelineno-7-18></a><span class=nb>print</span><span class=p>(</span><span class=n>variable_genes</span><span class=p>)</span>
</span></code></pre></div> <h3 id=complex-analysis>Complex Analysis<a class=headerlink href=#complex-analysis title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-8-1><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># Analyze expression patterns by cell type</span>
</span><span id=__span-8-2><a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=n>cell_type_analysis</span> <span class=o>=</span> <span class=n>slaf_array</span><span class=o>.</span><span class=n>query</span><span class=p>(</span><span class=s2>&quot;&quot;&quot;</span>
</span><span id=__span-8-3><a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a><span class=s2>    SELECT</span>
</span><span id=__span-8-4><a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=s2>        c.cell_type,</span>
</span><span id=__span-8-5><a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a><span class=s2>        COUNT(DISTINCT c.cell_id) as num_cells,</span>
</span><span id=__span-8-6><a id=__codelineno-8-6 name=__codelineno-8-6 href=#__codelineno-8-6></a><span class=s2>        AVG(c.total_counts) as avg_total_counts,</span>
</span><span id=__span-8-7><a id=__codelineno-8-7 name=__codelineno-8-7 href=#__codelineno-8-7></a><span class=s2>        AVG(c.n_genes_by_counts) as avg_genes_per_cell,</span>
</span><span id=__span-8-8><a id=__codelineno-8-8 name=__codelineno-8-8 href=#__codelineno-8-8></a><span class=s2>        COUNT(CASE WHEN c.high_mito = true THEN 1 END) as high_mito_cells</span>
</span><span id=__span-8-9><a id=__codelineno-8-9 name=__codelineno-8-9 href=#__codelineno-8-9></a><span class=s2>    FROM cells c</span>
</span><span id=__span-8-10><a id=__codelineno-8-10 name=__codelineno-8-10 href=#__codelineno-8-10></a><span class=s2>    GROUP BY c.cell_type</span>
</span><span id=__span-8-11><a id=__codelineno-8-11 name=__codelineno-8-11 href=#__codelineno-8-11></a><span class=s2>    ORDER BY num_cells DESC</span>
</span><span id=__span-8-12><a id=__codelineno-8-12 name=__codelineno-8-12 href=#__codelineno-8-12></a><span class=s2>&quot;&quot;&quot;</span><span class=p>)</span>
</span><span id=__span-8-13><a id=__codelineno-8-13 name=__codelineno-8-13 href=#__codelineno-8-13></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Cell type analysis:&quot;</span><span class=p>)</span>
</span><span id=__span-8-14><a id=__codelineno-8-14 name=__codelineno-8-14 href=#__codelineno-8-14></a><span class=nb>print</span><span class=p>(</span><span class=n>cell_type_analysis</span><span class=p>)</span>
</span></code></pre></div> <h2 id=lazy-evaluation-with-anndata-integration>Lazy Evaluation with Anndata Integration<a class=headerlink href=#lazy-evaluation-with-anndata-integration title="Permanent link">&para;</a></h2> <p>SLAF works seamlessly with Scanpy for familiar workflows with lazy evaluation:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-9-1><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf.integrations.anndata</span><span class=w> </span><span class=kn>import</span> <span class=n>read_slaf</span>
</span><span id=__span-9-2><a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>
</span><span id=__span-9-3><a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=c1># Load as lazy AnnData</span>
</span><span id=__span-9-4><a id=__codelineno-9-4 name=__codelineno-9-4 href=#__codelineno-9-4></a><span class=n>adata</span> <span class=o>=</span> <span class=n>read_slaf</span><span class=p>(</span><span class=n>slaf_path</span><span class=p>)</span>
</span><span id=__span-9-5><a id=__codelineno-9-5 name=__codelineno-9-5 href=#__codelineno-9-5></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;✅ Loaded: </span><span class=si>{</span><span class=n>adata</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>0</span><span class=p>]</span><span class=si>:</span><span class=s2>,</span><span class=si>}</span><span class=s2> cells × &quot;</span>
</span><span id=__span-9-6><a id=__codelineno-9-6 name=__codelineno-9-6 href=#__codelineno-9-6></a>      <span class=sa>f</span><span class=s2>&quot;</span><span class=si>{</span><span class=n>adata</span><span class=o>.</span><span class=n>shape</span><span class=p>[</span><span class=mi>1</span><span class=p>]</span><span class=si>:</span><span class=s2>,</span><span class=si>}</span><span class=s2> genes&quot;</span><span class=p>)</span>
</span><span id=__span-9-7><a id=__codelineno-9-7 name=__codelineno-9-7 href=#__codelineno-9-7></a>
</span><span id=__span-9-8><a id=__codelineno-9-8 name=__codelineno-9-8 href=#__codelineno-9-8></a><span class=c1># Lazy slicing - no data loaded yet</span>
</span><span id=__span-9-9><a id=__codelineno-9-9 name=__codelineno-9-9 href=#__codelineno-9-9></a><span class=n>t_cells</span> <span class=o>=</span> <span class=n>adata</span><span class=p>[</span><span class=n>adata</span><span class=o>.</span><span class=n>obs</span><span class=o>.</span><span class=n>cell_type</span> <span class=o>==</span> <span class=s2>&quot;T_cell&quot;</span><span class=p>,</span> <span class=p>:]</span>
</span><span id=__span-9-10><a id=__codelineno-9-10 name=__codelineno-9-10 href=#__codelineno-9-10></a>
</span><span id=__span-9-11><a id=__codelineno-9-11 name=__codelineno-9-11 href=#__codelineno-9-11></a><span class=c1># Only load data when you need it</span>
</span><span id=__span-9-12><a id=__codelineno-9-12 name=__codelineno-9-12 href=#__codelineno-9-12></a><span class=n>expression_matrix</span> <span class=o>=</span> <span class=n>t_cells</span><span class=o>.</span><span class=n>X</span><span class=o>.</span><span class=n>compute</span><span class=p>()</span>
</span><span id=__span-9-13><a id=__codelineno-9-13 name=__codelineno-9-13 href=#__codelineno-9-13></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Computed expression: </span><span class=si>{</span><span class=n>expression_matrix</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span></code></pre></div> <h2 id=working-with-metadata-and-layers>Working with Metadata and Layers<a class=headerlink href=#working-with-metadata-and-layers title="Permanent link">&para;</a></h2> <p>SLAF provides lazy views for all AnnData metadata objects, allowing you to read, modify, and create metadata efficiently:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-10-1><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf.integrations.anndata</span><span class=w> </span><span class=kn>import</span> <span class=n>read_slaf</span>
</span><span id=__span-10-2><a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a>
</span><span id=__span-10-3><a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a><span class=c1># Load as lazy AnnData</span>
</span><span id=__span-10-4><a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=n>adata</span> <span class=o>=</span> <span class=n>read_slaf</span><span class=p>(</span><span class=n>slaf_path</span><span class=p>)</span>
</span><span id=__span-10-5><a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a>
</span><span id=__span-10-6><a id=__codelineno-10-6 name=__codelineno-10-6 href=#__codelineno-10-6></a><span class=c1># Access and modify obs columns (cell metadata)</span>
</span><span id=__span-10-7><a id=__codelineno-10-7 name=__codelineno-10-7 href=#__codelineno-10-7></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Available obs columns:&quot;</span><span class=p>,</span> <span class=nb>list</span><span class=p>(</span><span class=n>adata</span><span class=o>.</span><span class=n>obs</span><span class=o>.</span><span class=n>columns</span><span class=p>))</span>
</span><span id=__span-10-8><a id=__codelineno-10-8 name=__codelineno-10-8 href=#__codelineno-10-8></a><span class=n>adata</span><span class=o>.</span><span class=n>obs</span><span class=p>[</span><span class=s2>&quot;new_cluster&quot;</span><span class=p>]</span> <span class=o>=</span> <span class=n>cluster_labels</span>  <span class=c1># Create new column</span>
</span><span id=__span-10-9><a id=__codelineno-10-9 name=__codelineno-10-9 href=#__codelineno-10-9></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Cluster distribution:</span><span class=se>\n</span><span class=si>{</span><span class=n>adata</span><span class=o>.</span><span class=n>obs</span><span class=p>[</span><span class=s1>&#39;new_cluster&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>value_counts</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-10-10><a id=__codelineno-10-10 name=__codelineno-10-10 href=#__codelineno-10-10></a>
</span><span id=__span-10-11><a id=__codelineno-10-11 name=__codelineno-10-11 href=#__codelineno-10-11></a><span class=c1># Access and modify var columns (gene metadata)</span>
</span><span id=__span-10-12><a id=__codelineno-10-12 name=__codelineno-10-12 href=#__codelineno-10-12></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Available var columns:&quot;</span><span class=p>,</span> <span class=nb>list</span><span class=p>(</span><span class=n>adata</span><span class=o>.</span><span class=n>var</span><span class=o>.</span><span class=n>columns</span><span class=p>))</span>
</span><span id=__span-10-13><a id=__codelineno-10-13 name=__codelineno-10-13 href=#__codelineno-10-13></a><span class=n>adata</span><span class=o>.</span><span class=n>var</span><span class=p>[</span><span class=s2>&quot;is_highly_variable&quot;</span><span class=p>]</span> <span class=o>=</span> <span class=n>hvg_flags</span>  <span class=c1># Create new column</span>
</span><span id=__span-10-14><a id=__codelineno-10-14 name=__codelineno-10-14 href=#__codelineno-10-14></a>
</span><span id=__span-10-15><a id=__codelineno-10-15 name=__codelineno-10-15 href=#__codelineno-10-15></a><span class=c1># Store multi-dimensional arrays (obsm - cell embeddings)</span>
</span><span id=__span-10-16><a id=__codelineno-10-16 name=__codelineno-10-16 href=#__codelineno-10-16></a><span class=kn>import</span><span class=w> </span><span class=nn>numpy</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>np</span>
</span><span id=__span-10-17><a id=__codelineno-10-17 name=__codelineno-10-17 href=#__codelineno-10-17></a><span class=n>umap_coords</span> <span class=o>=</span> <span class=n>compute_umap</span><span class=p>(</span><span class=n>adata</span><span class=p>)</span>  <span class=c1># Your function returning (n_cells, 2) array</span>
</span><span id=__span-10-18><a id=__codelineno-10-18 name=__codelineno-10-18 href=#__codelineno-10-18></a><span class=n>adata</span><span class=o>.</span><span class=n>obsm</span><span class=p>[</span><span class=s2>&quot;X_umap&quot;</span><span class=p>]</span> <span class=o>=</span> <span class=n>umap_coords</span>
</span><span id=__span-10-19><a id=__codelineno-10-19 name=__codelineno-10-19 href=#__codelineno-10-19></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;UMAP shape: </span><span class=si>{</span><span class=n>adata</span><span class=o>.</span><span class=n>obsm</span><span class=p>[</span><span class=s1>&#39;X_umap&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-10-20><a id=__codelineno-10-20 name=__codelineno-10-20 href=#__codelineno-10-20></a>
</span><span id=__span-10-21><a id=__codelineno-10-21 name=__codelineno-10-21 href=#__codelineno-10-21></a><span class=c1># Store gene-level embeddings (varm)</span>
</span><span id=__span-10-22><a id=__codelineno-10-22 name=__codelineno-10-22 href=#__codelineno-10-22></a><span class=n>pca_loadings</span> <span class=o>=</span> <span class=n>compute_pca_loadings</span><span class=p>(</span><span class=n>adata</span><span class=p>)</span>  <span class=c1># Your function returning (n_genes, 50) array</span>
</span><span id=__span-10-23><a id=__codelineno-10-23 name=__codelineno-10-23 href=#__codelineno-10-23></a><span class=n>adata</span><span class=o>.</span><span class=n>varm</span><span class=p>[</span><span class=s2>&quot;PCs&quot;</span><span class=p>]</span> <span class=o>=</span> <span class=n>pca_loadings</span>
</span><span id=__span-10-24><a id=__codelineno-10-24 name=__codelineno-10-24 href=#__codelineno-10-24></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;PCA loadings shape: </span><span class=si>{</span><span class=n>adata</span><span class=o>.</span><span class=n>varm</span><span class=p>[</span><span class=s1>&#39;PCs&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-10-25><a id=__codelineno-10-25 name=__codelineno-10-25 href=#__codelineno-10-25></a>
</span><span id=__span-10-26><a id=__codelineno-10-26 name=__codelineno-10-26 href=#__codelineno-10-26></a><span class=c1># Store unstructured metadata (uns)</span>
</span><span id=__span-10-27><a id=__codelineno-10-27 name=__codelineno-10-27 href=#__codelineno-10-27></a><span class=n>adata</span><span class=o>.</span><span class=n>uns</span><span class=p>[</span><span class=s2>&quot;neighbors&quot;</span><span class=p>]</span> <span class=o>=</span> <span class=p>{</span><span class=s2>&quot;params&quot;</span><span class=p>:</span> <span class=p>{</span><span class=s2>&quot;n_neighbors&quot;</span><span class=p>:</span> <span class=mi>15</span><span class=p>,</span> <span class=s2>&quot;metric&quot;</span><span class=p>:</span> <span class=s2>&quot;euclidean&quot;</span><span class=p>}}</span>
</span><span id=__span-10-28><a id=__codelineno-10-28 name=__codelineno-10-28 href=#__codelineno-10-28></a><span class=n>adata</span><span class=o>.</span><span class=n>uns</span><span class=p>[</span><span class=s2>&quot;pca&quot;</span><span class=p>]</span> <span class=o>=</span> <span class=p>{</span><span class=s2>&quot;variance_ratio&quot;</span><span class=p>:</span> <span class=n>variance_ratios</span><span class=o>.</span><span class=n>tolist</span><span class=p>()}</span>
</span><span id=__span-10-29><a id=__codelineno-10-29 name=__codelineno-10-29 href=#__codelineno-10-29></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Unstructured metadata keys:&quot;</span><span class=p>,</span> <span class=nb>list</span><span class=p>(</span><span class=n>adata</span><span class=o>.</span><span class=n>uns</span><span class=o>.</span><span class=n>keys</span><span class=p>()))</span>
</span><span id=__span-10-30><a id=__codelineno-10-30 name=__codelineno-10-30 href=#__codelineno-10-30></a>
</span><span id=__span-10-31><a id=__codelineno-10-31 name=__codelineno-10-31 href=#__codelineno-10-31></a><span class=c1># Access layers (alternative expression matrices)</span>
</span><span id=__span-10-32><a id=__codelineno-10-32 name=__codelineno-10-32 href=#__codelineno-10-32></a><span class=k>if</span> <span class=s2>&quot;spliced&quot;</span> <span class=ow>in</span> <span class=n>adata</span><span class=o>.</span><span class=n>layers</span><span class=p>:</span>
</span><span id=__span-10-33><a id=__codelineno-10-33 name=__codelineno-10-33 href=#__codelineno-10-33></a>    <span class=n>spliced_matrix</span> <span class=o>=</span> <span class=n>adata</span><span class=o>.</span><span class=n>layers</span><span class=p>[</span><span class=s2>&quot;spliced&quot;</span><span class=p>]</span><span class=o>.</span><span class=n>compute</span><span class=p>()</span>
</span><span id=__span-10-34><a id=__codelineno-10-34 name=__codelineno-10-34 href=#__codelineno-10-34></a>    <span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Spliced layer shape: </span><span class=si>{</span><span class=n>spliced_matrix</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-10-35><a id=__codelineno-10-35 name=__codelineno-10-35 href=#__codelineno-10-35></a>
</span><span id=__span-10-36><a id=__codelineno-10-36 name=__codelineno-10-36 href=#__codelineno-10-36></a><span class=c1># Create new layers</span>
</span><span id=__span-10-37><a id=__codelineno-10-37 name=__codelineno-10-37 href=#__codelineno-10-37></a><span class=n>normalized</span> <span class=o>=</span> <span class=n>adata</span><span class=o>.</span><span class=n>X</span><span class=o>.</span><span class=n>copy</span><span class=p>()</span>
</span><span id=__span-10-38><a id=__codelineno-10-38 name=__codelineno-10-38 href=#__codelineno-10-38></a><span class=c1># ... apply normalization ...</span>
</span><span id=__span-10-39><a id=__codelineno-10-39 name=__codelineno-10-39 href=#__codelineno-10-39></a><span class=n>adata</span><span class=o>.</span><span class=n>layers</span><span class=p>[</span><span class=s2>&quot;normalized&quot;</span><span class=p>]</span> <span class=o>=</span> <span class=n>normalized</span>
</span><span id=__span-10-40><a id=__codelineno-10-40 name=__codelineno-10-40 href=#__codelineno-10-40></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Available layers:&quot;</span><span class=p>,</span> <span class=nb>list</span><span class=p>(</span><span class=n>adata</span><span class=o>.</span><span class=n>layers</span><span class=o>.</span><span class=n>keys</span><span class=p>()))</span>
</span></code></pre></div> <p><strong>Key Features:</strong></p> <ul> <li><strong>DataFrame-like access</strong>: <code>obs</code> and <code>var</code> behave like pandas DataFrames</li> <li><strong>Dictionary-like interface</strong>: <code>layers</code>, <code>obsm</code>, <code>varm</code>, and <code>uns</code> use dict-like syntax</li> <li><strong>Lazy evaluation</strong>: Metadata is accessed on-demand</li> <li><strong>Immutability protection</strong>: Columns/keys converted from h5ad are protected from accidental deletion</li> </ul> <h2 id=lazy-scanpy-preprocessing>Lazy Scanpy Preprocessing<a class=headerlink href=#lazy-scanpy-preprocessing title="Permanent link">&para;</a></h2> <p>SLAF provides lazy versions of scanpy preprocessing functions that work efficiently on large datasets:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-11-1><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf.integrations</span><span class=w> </span><span class=kn>import</span> <span class=n>scanpy</span> <span class=k>as</span> <span class=n>slaf_scanpy</span>
</span><span id=__span-11-2><a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf.integrations.anndata</span><span class=w> </span><span class=kn>import</span> <span class=n>read_slaf</span>
</span><span id=__span-11-3><a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a>
</span><span id=__span-11-4><a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a><span class=c1># Load as lazy AnnData</span>
</span><span id=__span-11-5><a id=__codelineno-11-5 name=__codelineno-11-5 href=#__codelineno-11-5></a><span class=n>adata</span> <span class=o>=</span> <span class=n>read_slaf</span><span class=p>(</span><span class=n>slaf_path</span><span class=p>)</span>
</span><span id=__span-11-6><a id=__codelineno-11-6 name=__codelineno-11-6 href=#__codelineno-11-6></a>
</span><span id=__span-11-7><a id=__codelineno-11-7 name=__codelineno-11-7 href=#__codelineno-11-7></a><span class=c1># Calculate QC metrics (lazy SQL aggregation)</span>
</span><span id=__span-11-8><a id=__codelineno-11-8 name=__codelineno-11-8 href=#__codelineno-11-8></a><span class=n>cell_qc</span><span class=p>,</span> <span class=n>gene_qc</span> <span class=o>=</span> <span class=n>slaf_scanpy</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>calculate_qc_metrics</span><span class=p>(</span>
</span><span id=__span-11-9><a id=__codelineno-11-9 name=__codelineno-11-9 href=#__codelineno-11-9></a>    <span class=n>adata</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>False</span>
</span><span id=__span-11-10><a id=__codelineno-11-10 name=__codelineno-11-10 href=#__codelineno-11-10></a><span class=p>)</span>
</span><span id=__span-11-11><a id=__codelineno-11-11 name=__codelineno-11-11 href=#__codelineno-11-11></a><span class=nb>print</span><span class=p>(</span><span class=s2>&quot;Cell QC metrics:&quot;</span><span class=p>)</span>
</span><span id=__span-11-12><a id=__codelineno-11-12 name=__codelineno-11-12 href=#__codelineno-11-12></a><span class=nb>print</span><span class=p>(</span><span class=n>cell_qc</span><span class=o>.</span><span class=n>head</span><span class=p>())</span>
</span><span id=__span-11-13><a id=__codelineno-11-13 name=__codelineno-11-13 href=#__codelineno-11-13></a>
</span><span id=__span-11-14><a id=__codelineno-11-14 name=__codelineno-11-14 href=#__codelineno-11-14></a><span class=c1># Filter cells (lazy - no data loaded)</span>
</span><span id=__span-11-15><a id=__codelineno-11-15 name=__codelineno-11-15 href=#__codelineno-11-15></a><span class=n>adata_filtered</span> <span class=o>=</span> <span class=n>slaf_scanpy</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>filter_cells</span><span class=p>(</span>
</span><span id=__span-11-16><a id=__codelineno-11-16 name=__codelineno-11-16 href=#__codelineno-11-16></a>    <span class=n>adata</span><span class=p>,</span>
</span><span id=__span-11-17><a id=__codelineno-11-17 name=__codelineno-11-17 href=#__codelineno-11-17></a>    <span class=n>min_genes</span><span class=o>=</span><span class=mi>200</span><span class=p>,</span>
</span><span id=__span-11-18><a id=__codelineno-11-18 name=__codelineno-11-18 href=#__codelineno-11-18></a>    <span class=n>min_counts</span><span class=o>=</span><span class=mi>1000</span><span class=p>,</span>
</span><span id=__span-11-19><a id=__codelineno-11-19 name=__codelineno-11-19 href=#__codelineno-11-19></a>    <span class=n>inplace</span><span class=o>=</span><span class=kc>False</span>
</span><span id=__span-11-20><a id=__codelineno-11-20 name=__codelineno-11-20 href=#__codelineno-11-20></a><span class=p>)</span>
</span><span id=__span-11-21><a id=__codelineno-11-21 name=__codelineno-11-21 href=#__codelineno-11-21></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;After filtering: </span><span class=si>{</span><span class=n>adata_filtered</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-11-22><a id=__codelineno-11-22 name=__codelineno-11-22 href=#__codelineno-11-22></a>
</span><span id=__span-11-23><a id=__codelineno-11-23 name=__codelineno-11-23 href=#__codelineno-11-23></a><span class=c1># Filter genes (lazy)</span>
</span><span id=__span-11-24><a id=__codelineno-11-24 name=__codelineno-11-24 href=#__codelineno-11-24></a><span class=n>adata_filtered</span> <span class=o>=</span> <span class=n>slaf_scanpy</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>filter_genes</span><span class=p>(</span>
</span><span id=__span-11-25><a id=__codelineno-11-25 name=__codelineno-11-25 href=#__codelineno-11-25></a>    <span class=n>adata_filtered</span><span class=p>,</span>
</span><span id=__span-11-26><a id=__codelineno-11-26 name=__codelineno-11-26 href=#__codelineno-11-26></a>    <span class=n>min_cells</span><span class=o>=</span><span class=mi>30</span><span class=p>,</span>
</span><span id=__span-11-27><a id=__codelineno-11-27 name=__codelineno-11-27 href=#__codelineno-11-27></a>    <span class=n>inplace</span><span class=o>=</span><span class=kc>False</span>
</span><span id=__span-11-28><a id=__codelineno-11-28 name=__codelineno-11-28 href=#__codelineno-11-28></a><span class=p>)</span>
</span><span id=__span-11-29><a id=__codelineno-11-29 name=__codelineno-11-29 href=#__codelineno-11-29></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;After gene filtering: </span><span class=si>{</span><span class=n>adata_filtered</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-11-30><a id=__codelineno-11-30 name=__codelineno-11-30 href=#__codelineno-11-30></a>
</span><span id=__span-11-31><a id=__codelineno-11-31 name=__codelineno-11-31 href=#__codelineno-11-31></a><span class=c1># Normalize total (lazy transformation)</span>
</span><span id=__span-11-32><a id=__codelineno-11-32 name=__codelineno-11-32 href=#__codelineno-11-32></a><span class=n>adata_norm</span> <span class=o>=</span> <span class=n>slaf_scanpy</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>normalize_total</span><span class=p>(</span>
</span><span id=__span-11-33><a id=__codelineno-11-33 name=__codelineno-11-33 href=#__codelineno-11-33></a>    <span class=n>adata_filtered</span><span class=p>,</span>
</span><span id=__span-11-34><a id=__codelineno-11-34 name=__codelineno-11-34 href=#__codelineno-11-34></a>    <span class=n>target_sum</span><span class=o>=</span><span class=mf>1e4</span><span class=p>,</span>
</span><span id=__span-11-35><a id=__codelineno-11-35 name=__codelineno-11-35 href=#__codelineno-11-35></a>    <span class=n>inplace</span><span class=o>=</span><span class=kc>False</span>
</span><span id=__span-11-36><a id=__codelineno-11-36 name=__codelineno-11-36 href=#__codelineno-11-36></a><span class=p>)</span>
</span><span id=__span-11-37><a id=__codelineno-11-37 name=__codelineno-11-37 href=#__codelineno-11-37></a>
</span><span id=__span-11-38><a id=__codelineno-11-38 name=__codelineno-11-38 href=#__codelineno-11-38></a><span class=c1># Apply log1p transformation (lazy)</span>
</span><span id=__span-11-39><a id=__codelineno-11-39 name=__codelineno-11-39 href=#__codelineno-11-39></a><span class=n>adata_log</span> <span class=o>=</span> <span class=n>slaf_scanpy</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>log1p</span><span class=p>(</span>
</span><span id=__span-11-40><a id=__codelineno-11-40 name=__codelineno-11-40 href=#__codelineno-11-40></a>    <span class=n>adata_norm</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>False</span>
</span><span id=__span-11-41><a id=__codelineno-11-41 name=__codelineno-11-41 href=#__codelineno-11-41></a><span class=p>)</span>
</span><span id=__span-11-42><a id=__codelineno-11-42 name=__codelineno-11-42 href=#__codelineno-11-42></a>
</span><span id=__span-11-43><a id=__codelineno-11-43 name=__codelineno-11-43 href=#__codelineno-11-43></a><span class=c1># Find highly variable genes (lazy SQL)</span>
</span><span id=__span-11-44><a id=__codelineno-11-44 name=__codelineno-11-44 href=#__codelineno-11-44></a><span class=n>hvg_stats</span> <span class=o>=</span> <span class=n>slaf_scanpy</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>highly_variable_genes</span><span class=p>(</span>
</span><span id=__span-11-45><a id=__codelineno-11-45 name=__codelineno-11-45 href=#__codelineno-11-45></a>    <span class=n>adata_log</span><span class=p>,</span>
</span><span id=__span-11-46><a id=__codelineno-11-46 name=__codelineno-11-46 href=#__codelineno-11-46></a>    <span class=n>min_mean</span><span class=o>=</span><span class=mf>0.0125</span><span class=p>,</span>
</span><span id=__span-11-47><a id=__codelineno-11-47 name=__codelineno-11-47 href=#__codelineno-11-47></a>    <span class=n>max_mean</span><span class=o>=</span><span class=mi>3</span><span class=p>,</span>
</span><span id=__span-11-48><a id=__codelineno-11-48 name=__codelineno-11-48 href=#__codelineno-11-48></a>    <span class=n>min_disp</span><span class=o>=</span><span class=mf>0.5</span><span class=p>,</span>
</span><span id=__span-11-49><a id=__codelineno-11-49 name=__codelineno-11-49 href=#__codelineno-11-49></a>    <span class=n>inplace</span><span class=o>=</span><span class=kc>False</span>
</span><span id=__span-11-50><a id=__codelineno-11-50 name=__codelineno-11-50 href=#__codelineno-11-50></a><span class=p>)</span>
</span><span id=__span-11-51><a id=__codelineno-11-51 name=__codelineno-11-51 href=#__codelineno-11-51></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Highly variable genes: </span><span class=si>{</span><span class=n>hvg_stats</span><span class=p>[</span><span class=s1>&#39;highly_variable&#39;</span><span class=p>]</span><span class=o>.</span><span class=n>sum</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span><span id=__span-11-52><a id=__codelineno-11-52 name=__codelineno-11-52 href=#__codelineno-11-52></a>
</span><span id=__span-11-53><a id=__codelineno-11-53 name=__codelineno-11-53 href=#__codelineno-11-53></a><span class=c1># All operations are lazy - compute when needed</span>
</span><span id=__span-11-54><a id=__codelineno-11-54 name=__codelineno-11-54 href=#__codelineno-11-54></a><span class=n>expression</span> <span class=o>=</span> <span class=n>adata_log</span><span class=o>.</span><span class=n>X</span><span class=o>.</span><span class=n>compute</span><span class=p>()</span>
</span><span id=__span-11-55><a id=__codelineno-11-55 name=__codelineno-11-55 href=#__codelineno-11-55></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Computed expression: </span><span class=si>{</span><span class=n>expression</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span></code></pre></div> <h3 id=chaining-transformations>Chaining Transformations<a class=headerlink href=#chaining-transformations title="Permanent link">&para;</a></h3> <p>You can chain lazy transformations efficiently:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-12-1><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=c1># Build complete preprocessing pipeline</span>
</span><span id=__span-12-2><a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=n>adata_processed</span> <span class=o>=</span> <span class=n>slaf_scanpy</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>normalize_total</span><span class=p>(</span>
</span><span id=__span-12-3><a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a>    <span class=n>adata</span><span class=p>,</span> <span class=n>target_sum</span><span class=o>=</span><span class=mf>1e4</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>False</span>
</span><span id=__span-12-4><a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a><span class=p>)</span>
</span><span id=__span-12-5><a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a><span class=n>adata_processed</span> <span class=o>=</span> <span class=n>slaf_scanpy</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>log1p</span><span class=p>(</span>
</span><span id=__span-12-6><a id=__codelineno-12-6 name=__codelineno-12-6 href=#__codelineno-12-6></a>    <span class=n>adata_processed</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>False</span>
</span><span id=__span-12-7><a id=__codelineno-12-7 name=__codelineno-12-7 href=#__codelineno-12-7></a><span class=p>)</span>
</span><span id=__span-12-8><a id=__codelineno-12-8 name=__codelineno-12-8 href=#__codelineno-12-8></a>
</span><span id=__span-12-9><a id=__codelineno-12-9 name=__codelineno-12-9 href=#__codelineno-12-9></a><span class=c1># Slice after transformations</span>
</span><span id=__span-12-10><a id=__codelineno-12-10 name=__codelineno-12-10 href=#__codelineno-12-10></a><span class=n>subset</span> <span class=o>=</span> <span class=n>adata_processed</span><span class=p>[:</span><span class=mi>1000</span><span class=p>,</span> <span class=p>:</span><span class=mi>500</span><span class=p>]</span>
</span><span id=__span-12-11><a id=__codelineno-12-11 name=__codelineno-12-11 href=#__codelineno-12-11></a>
</span><span id=__span-12-12><a id=__codelineno-12-12 name=__codelineno-12-12 href=#__codelineno-12-12></a><span class=c1># Compute only when needed</span>
</span><span id=__span-12-13><a id=__codelineno-12-13 name=__codelineno-12-13 href=#__codelineno-12-13></a><span class=n>result</span> <span class=o>=</span> <span class=n>subset</span><span class=o>.</span><span class=n>X</span><span class=o>.</span><span class=n>compute</span><span class=p>()</span>
</span><span id=__span-12-14><a id=__codelineno-12-14 name=__codelineno-12-14 href=#__codelineno-12-14></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Processed subset: </span><span class=si>{</span><span class=n>result</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span></code></pre></div> <h3 id=key-benefits>Key Benefits<a class=headerlink href=#key-benefits title="Permanent link">&para;</a></h3> <ul> <li><strong>Memory Efficient</strong>: Operations stored as instructions</li> <li><strong>SQL Performance</strong>: QC metrics use SQL aggregation</li> <li><strong>Composable</strong>: Chain transformations easily</li> <li><strong>Lazy</strong>: Compute only when you call <code>.compute()</code></li> </ul> <h2 id=efficient-filtering>Efficient Filtering<a class=headerlink href=#efficient-filtering title="Permanent link">&para;</a></h2> <p>SLAF provides optimized filtering methods:</p> <div class="language-python highlight"><pre><span></span><code><span id=__span-13-1><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=c1># Filter cells by metadata</span>
</span><span id=__span-13-2><a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=n>t_cells</span> <span class=o>=</span> <span class=n>slaf_array</span><span class=o>.</span><span class=n>filter_cells</span><span class=p>(</span><span class=n>cell_type</span><span class=o>=</span><span class=s2>&quot;T_cell&quot;</span><span class=p>,</span> <span class=n>total_counts</span><span class=o>=</span><span class=s2>&quot;&gt;1000&quot;</span><span class=p>)</span>
</span><span id=__span-13-3><a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Found </span><span class=si>{</span><span class=nb>len</span><span class=p>(</span><span class=n>t_cells</span><span class=p>)</span><span class=si>}</span><span class=s2> T cells with high counts&quot;</span><span class=p>)</span>
</span><span id=__span-13-4><a id=__codelineno-13-4 name=__codelineno-13-4 href=#__codelineno-13-4></a>
</span><span id=__span-13-5><a id=__codelineno-13-5 name=__codelineno-13-5 href=#__codelineno-13-5></a><span class=c1># Filter genes</span>
</span><span id=__span-13-6><a id=__codelineno-13-6 name=__codelineno-13-6 href=#__codelineno-13-6></a><span class=n>variable_genes</span> <span class=o>=</span> <span class=n>slaf_array</span><span class=o>.</span><span class=n>filter_genes</span><span class=p>(</span><span class=n>highly_variable</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
</span><span id=__span-13-7><a id=__codelineno-13-7 name=__codelineno-13-7 href=#__codelineno-13-7></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Found </span><span class=si>{</span><span class=nb>len</span><span class=p>(</span><span class=n>variable_genes</span><span class=p>)</span><span class=si>}</span><span class=s2> highly variable genes&quot;</span><span class=p>)</span>
</span><span id=__span-13-8><a id=__codelineno-13-8 name=__codelineno-13-8 href=#__codelineno-13-8></a>
</span><span id=__span-13-9><a id=__codelineno-13-9 name=__codelineno-13-9 href=#__codelineno-13-9></a><span class=c1># Get expression submatrix</span>
</span><span id=__span-13-10><a id=__codelineno-13-10 name=__codelineno-13-10 href=#__codelineno-13-10></a><span class=n>expression</span> <span class=o>=</span> <span class=n>slaf_array</span><span class=o>.</span><span class=n>get_submatrix</span><span class=p>(</span>
</span><span id=__span-13-11><a id=__codelineno-13-11 name=__codelineno-13-11 href=#__codelineno-13-11></a>    <span class=n>cell_selector</span><span class=o>=</span><span class=n>t_cells</span><span class=p>,</span>
</span><span id=__span-13-12><a id=__codelineno-13-12 name=__codelineno-13-12 href=#__codelineno-13-12></a>    <span class=n>gene_selector</span><span class=o>=</span><span class=n>variable_genes</span>
</span><span id=__span-13-13><a id=__codelineno-13-13 name=__codelineno-13-13 href=#__codelineno-13-13></a><span class=p>)</span>
</span><span id=__span-13-14><a id=__codelineno-13-14 name=__codelineno-13-14 href=#__codelineno-13-14></a><span class=nb>print</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Expression submatrix: </span><span class=si>{</span><span class=n>expression</span><span class=o>.</span><span class=n>shape</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
</span></code></pre></div> <h2 id=ml-training-with-dataloaders>ML Training with Dataloaders<a class=headerlink href=#ml-training-with-dataloaders title="Permanent link">&para;</a></h2> <p>SLAF provides efficient tokenization and dataloaders for training foundation models:</p> <h3 id=tokenization>Tokenization<a class=headerlink href=#tokenization title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-14-1><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf.ml</span><span class=w> </span><span class=kn>import</span> <span class=n>SLAFTokenizer</span>
</span><span id=__span-14-2><a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a>
</span><span id=__span-14-3><a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a><span class=c1># Create tokenizer for GeneFormer style tokenization</span>
</span><span id=__span-14-4><a id=__codelineno-14-4 name=__codelineno-14-4 href=#__codelineno-14-4></a><span class=n>tokenizer</span> <span class=o>=</span> <span class=n>SLAFTokenizer</span><span class=p>(</span>
</span><span id=__span-14-5><a id=__codelineno-14-5 name=__codelineno-14-5 href=#__codelineno-14-5></a>    <span class=n>slaf_array</span><span class=o>=</span><span class=n>slaf_array</span><span class=p>,</span>
</span><span id=__span-14-6><a id=__codelineno-14-6 name=__codelineno-14-6 href=#__codelineno-14-6></a>    <span class=n>tokenizer_type</span><span class=o>=</span><span class=s2>&quot;geneformer&quot;</span><span class=p>,</span>
</span><span id=__span-14-7><a id=__codelineno-14-7 name=__codelineno-14-7 href=#__codelineno-14-7></a>    <span class=n>vocab_size</span><span class=o>=</span><span class=mi>50000</span><span class=p>,</span>
</span><span id=__span-14-8><a id=__codelineno-14-8 name=__codelineno-14-8 href=#__codelineno-14-8></a>    <span class=n>n_expression_bins</span><span class=o>=</span><span class=mi>10</span>
</span><span id=__span-14-9><a id=__codelineno-14-9 name=__codelineno-14-9 href=#__codelineno-14-9></a><span class=p>)</span>
</span><span id=__span-14-10><a id=__codelineno-14-10 name=__codelineno-14-10 href=#__codelineno-14-10></a>
</span><span id=__span-14-11><a id=__codelineno-14-11 name=__codelineno-14-11 href=#__codelineno-14-11></a><span class=c1># Geneformer tokenization (gene sequence only)</span>
</span><span id=__span-14-12><a id=__codelineno-14-12 name=__codelineno-14-12 href=#__codelineno-14-12></a><span class=n>gene_sequences</span> <span class=o>=</span> <span class=p>[[</span><span class=mi>1</span><span class=p>,</span> <span class=mi>2</span><span class=p>,</span> <span class=mi>3</span><span class=p>],</span> <span class=p>[</span><span class=mi>4</span><span class=p>,</span> <span class=mi>5</span><span class=p>,</span> <span class=mi>6</span><span class=p>]]</span>  <span class=c1># Example gene IDs</span>
</span><span id=__span-14-13><a id=__codelineno-14-13 name=__codelineno-14-13 href=#__codelineno-14-13></a><span class=n>input_ids</span><span class=p>,</span> <span class=n>attention_mask</span> <span class=o>=</span> <span class=n>tokenizer</span><span class=o>.</span><span class=n>tokenize</span><span class=p>(</span>
</span><span id=__span-14-14><a id=__codelineno-14-14 name=__codelineno-14-14 href=#__codelineno-14-14></a>    <span class=n>gene_sequences</span><span class=p>,</span>
</span><span id=__span-14-15><a id=__codelineno-14-15 name=__codelineno-14-15 href=#__codelineno-14-15></a>    <span class=n>max_genes</span><span class=o>=</span><span class=mi>2048</span>
</span><span id=__span-14-16><a id=__codelineno-14-16 name=__codelineno-14-16 href=#__codelineno-14-16></a><span class=p>)</span>
</span><span id=__span-14-17><a id=__codelineno-14-17 name=__codelineno-14-17 href=#__codelineno-14-17></a>
</span><span id=__span-14-18><a id=__codelineno-14-18 name=__codelineno-14-18 href=#__codelineno-14-18></a><span class=c1># Create tokenizer for scGPT style tokenization</span>
</span><span id=__span-14-19><a id=__codelineno-14-19 name=__codelineno-14-19 href=#__codelineno-14-19></a><span class=n>tokenizer</span> <span class=o>=</span> <span class=n>SLAFTokenizer</span><span class=p>(</span>
</span><span id=__span-14-20><a id=__codelineno-14-20 name=__codelineno-14-20 href=#__codelineno-14-20></a>    <span class=n>slaf_array</span><span class=o>=</span><span class=n>slaf_array</span><span class=p>,</span>
</span><span id=__span-14-21><a id=__codelineno-14-21 name=__codelineno-14-21 href=#__codelineno-14-21></a>    <span class=n>tokenizer_type</span><span class=o>=</span><span class=s2>&quot;scgpt&quot;</span><span class=p>,</span>
</span><span id=__span-14-22><a id=__codelineno-14-22 name=__codelineno-14-22 href=#__codelineno-14-22></a>    <span class=n>vocab_size</span><span class=o>=</span><span class=mi>50000</span><span class=p>,</span>
</span><span id=__span-14-23><a id=__codelineno-14-23 name=__codelineno-14-23 href=#__codelineno-14-23></a>    <span class=n>n_expression_bins</span><span class=o>=</span><span class=mi>10</span>
</span><span id=__span-14-24><a id=__codelineno-14-24 name=__codelineno-14-24 href=#__codelineno-14-24></a><span class=p>)</span>
</span><span id=__span-14-25><a id=__codelineno-14-25 name=__codelineno-14-25 href=#__codelineno-14-25></a>
</span><span id=__span-14-26><a id=__codelineno-14-26 name=__codelineno-14-26 href=#__codelineno-14-26></a><span class=c1># scGPT tokenization (gene-expression pairs)</span>
</span><span id=__span-14-27><a id=__codelineno-14-27 name=__codelineno-14-27 href=#__codelineno-14-27></a><span class=n>gene_sequences</span> <span class=o>=</span> <span class=p>[[</span><span class=mi>1</span><span class=p>,</span> <span class=mi>2</span><span class=p>,</span> <span class=mi>3</span><span class=p>],</span> <span class=p>[</span><span class=mi>4</span><span class=p>,</span> <span class=mi>5</span><span class=p>,</span> <span class=mi>6</span><span class=p>]]</span>  <span class=c1># Gene IDs</span>
</span><span id=__span-14-28><a id=__codelineno-14-28 name=__codelineno-14-28 href=#__codelineno-14-28></a><span class=n>expr_sequences</span> <span class=o>=</span> <span class=p>[[</span><span class=mf>0.5</span><span class=p>,</span> <span class=mf>0.8</span><span class=p>,</span> <span class=mf>0.2</span><span class=p>],</span> <span class=p>[</span><span class=mf>0.9</span><span class=p>,</span> <span class=mf>0.1</span><span class=p>,</span> <span class=mf>0.7</span><span class=p>]]</span>  <span class=c1># Expression values</span>
</span><span id=__span-14-29><a id=__codelineno-14-29 name=__codelineno-14-29 href=#__codelineno-14-29></a><span class=n>input_ids</span><span class=p>,</span> <span class=n>attention_mask</span> <span class=o>=</span> <span class=n>tokenizer</span><span class=o>.</span><span class=n>tokenize</span><span class=p>(</span>
</span><span id=__span-14-30><a id=__codelineno-14-30 name=__codelineno-14-30 href=#__codelineno-14-30></a>    <span class=n>gene_sequences</span><span class=p>,</span>
</span><span id=__span-14-31><a id=__codelineno-14-31 name=__codelineno-14-31 href=#__codelineno-14-31></a>    <span class=n>expr_sequences</span><span class=o>=</span><span class=n>expr_sequences</span><span class=p>,</span>
</span><span id=__span-14-32><a id=__codelineno-14-32 name=__codelineno-14-32 href=#__codelineno-14-32></a>    <span class=n>max_genes</span><span class=o>=</span><span class=mi>1024</span>
</span><span id=__span-14-33><a id=__codelineno-14-33 name=__codelineno-14-33 href=#__codelineno-14-33></a><span class=p>)</span>
</span></code></pre></div> <h3 id=dataloader-for-training>DataLoader for Training<a class=headerlink href=#dataloader-for-training title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-15-1><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf.ml</span><span class=w> </span><span class=kn>import</span> <span class=n>SLAFDataLoader</span>
</span><span id=__span-15-2><a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a>
</span><span id=__span-15-3><a id=__codelineno-15-3 name=__codelineno-15-3 href=#__codelineno-15-3></a><span class=c1># Create DataLoader (uses MoS by default for high entropy)</span>
</span><span id=__span-15-4><a id=__codelineno-15-4 name=__codelineno-15-4 href=#__codelineno-15-4></a><span class=n>dataloader</span> <span class=o>=</span> <span class=n>SLAFDataLoader</span><span class=p>(</span>
</span><span id=__span-15-5><a id=__codelineno-15-5 name=__codelineno-15-5 href=#__codelineno-15-5></a>    <span class=n>slaf_array</span><span class=o>=</span><span class=n>slaf_array</span><span class=p>,</span>
</span><span id=__span-15-6><a id=__codelineno-15-6 name=__codelineno-15-6 href=#__codelineno-15-6></a>    <span class=n>tokenizer_type</span><span class=o>=</span><span class=s2>&quot;geneformer&quot;</span><span class=p>,</span>  <span class=c1># or &quot;scgpt&quot;</span>
</span><span id=__span-15-7><a id=__codelineno-15-7 name=__codelineno-15-7 href=#__codelineno-15-7></a>    <span class=n>batch_size</span><span class=o>=</span><span class=mi>32</span><span class=p>,</span>
</span><span id=__span-15-8><a id=__codelineno-15-8 name=__codelineno-15-8 href=#__codelineno-15-8></a>    <span class=n>max_genes</span><span class=o>=</span><span class=mi>2048</span>
</span><span id=__span-15-9><a id=__codelineno-15-9 name=__codelineno-15-9 href=#__codelineno-15-9></a><span class=p>)</span>
</span><span id=__span-15-10><a id=__codelineno-15-10 name=__codelineno-15-10 href=#__codelineno-15-10></a>
</span><span id=__span-15-11><a id=__codelineno-15-11 name=__codelineno-15-11 href=#__codelineno-15-11></a><span class=c1># Use with PyTorch training</span>
</span><span id=__span-15-12><a id=__codelineno-15-12 name=__codelineno-15-12 href=#__codelineno-15-12></a><span class=k>for</span> <span class=n>batch</span> <span class=ow>in</span> <span class=n>dataloader</span><span class=p>:</span>
</span><span id=__span-15-13><a id=__codelineno-15-13 name=__codelineno-15-13 href=#__codelineno-15-13></a>    <span class=n>input_ids</span> <span class=o>=</span> <span class=n>batch</span><span class=p>[</span><span class=s2>&quot;input_ids&quot;</span><span class=p>]</span>
</span><span id=__span-15-14><a id=__codelineno-15-14 name=__codelineno-15-14 href=#__codelineno-15-14></a>    <span class=n>attention_mask</span> <span class=o>=</span> <span class=n>batch</span><span class=p>[</span><span class=s2>&quot;attention_mask&quot;</span><span class=p>]</span>
</span><span id=__span-15-15><a id=__codelineno-15-15 name=__codelineno-15-15 href=#__codelineno-15-15></a>    <span class=n>cell_ids</span> <span class=o>=</span> <span class=n>batch</span><span class=p>[</span><span class=s2>&quot;cell_ids&quot;</span><span class=p>]</span>
</span><span id=__span-15-16><a id=__codelineno-15-16 name=__codelineno-15-16 href=#__codelineno-15-16></a>
</span><span id=__span-15-17><a id=__codelineno-15-17 name=__codelineno-15-17 href=#__codelineno-15-17></a>    <span class=c1># Your training loop here</span>
</span><span id=__span-15-18><a id=__codelineno-15-18 name=__codelineno-15-18 href=#__codelineno-15-18></a>    <span class=n>loss</span> <span class=o>=</span> <span class=n>model</span><span class=p>(</span><span class=n>input_ids</span><span class=p>,</span> <span class=n>attention_mask</span><span class=o>=</span><span class=n>attention_mask</span><span class=p>)</span>
</span><span id=__span-15-19><a id=__codelineno-15-19 name=__codelineno-15-19 href=#__codelineno-15-19></a>    <span class=n>loss</span><span class=o>.</span><span class=n>backward</span><span class=p>()</span>
</span></code></pre></div> <h2 id=using-the-slaf-cli>Using the SLAF CLI<a class=headerlink href=#using-the-slaf-cli title="Permanent link">&para;</a></h2> <p>SLAF includes a powerful command-line interface for common data operations:</p> <h3 id=getting-help>Getting Help<a class=headerlink href=#getting-help title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-16-1><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=c1># Show general help</span>
</span><span id=__span-16-2><a id=__codelineno-16-2 name=__codelineno-16-2 href=#__codelineno-16-2></a>slaf<span class=w> </span>--help
</span><span id=__span-16-3><a id=__codelineno-16-3 name=__codelineno-16-3 href=#__codelineno-16-3></a>
</span><span id=__span-16-4><a id=__codelineno-16-4 name=__codelineno-16-4 href=#__codelineno-16-4></a><span class=c1># Show help for specific command</span>
</span><span id=__span-16-5><a id=__codelineno-16-5 name=__codelineno-16-5 href=#__codelineno-16-5></a>slaf<span class=w> </span>convert<span class=w> </span>--help
</span><span id=__span-16-6><a id=__codelineno-16-6 name=__codelineno-16-6 href=#__codelineno-16-6></a>slaf<span class=w> </span>query<span class=w> </span>--help
</span></code></pre></div> <h3 id=basic-commands>Basic Commands<a class=headerlink href=#basic-commands title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-17-1><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=c1># Check SLAF version</span>
</span><span id=__span-17-2><a id=__codelineno-17-2 name=__codelineno-17-2 href=#__codelineno-17-2></a>slaf<span class=w> </span>version
</span><span id=__span-17-3><a id=__codelineno-17-3 name=__codelineno-17-3 href=#__codelineno-17-3></a>
</span><span id=__span-17-4><a id=__codelineno-17-4 name=__codelineno-17-4 href=#__codelineno-17-4></a><span class=c1># Show info about your dataset</span>
</span><span id=__span-17-5><a id=__codelineno-17-5 name=__codelineno-17-5 href=#__codelineno-17-5></a>slaf<span class=w> </span>info<span class=w> </span>my_dataset.slaf
</span><span id=__span-17-6><a id=__codelineno-17-6 name=__codelineno-17-6 href=#__codelineno-17-6></a>
</span><span id=__span-17-7><a id=__codelineno-17-7 name=__codelineno-17-7 href=#__codelineno-17-7></a><span class=c1># Execute a SQL query</span>
</span><span id=__span-17-8><a id=__codelineno-17-8 name=__codelineno-17-8 href=#__codelineno-17-8></a>slaf<span class=w> </span>query<span class=w> </span>my_dataset.slaf<span class=w> </span><span class=s2>&quot;SELECT COUNT(*) FROM cells&quot;</span>
</span><span id=__span-17-9><a id=__codelineno-17-9 name=__codelineno-17-9 href=#__codelineno-17-9></a>
</span><span id=__span-17-10><a id=__codelineno-17-10 name=__codelineno-17-10 href=#__codelineno-17-10></a><span class=c1># Convert other formats to SLAF</span>
</span><span id=__span-17-11><a id=__codelineno-17-11 name=__codelineno-17-11 href=#__codelineno-17-11></a>slaf<span class=w> </span>convert<span class=w> </span>data.h5ad<span class=w> </span>output.slaf
</span></code></pre></div> <h3 id=common-use-cases>Common Use Cases<a class=headerlink href=#common-use-cases title="Permanent link">&para;</a></h3> <p><strong>Converting Datasets:</strong></p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-18-1><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=c1># Convert an AnnData file</span>
</span><span id=__span-18-2><a id=__codelineno-18-2 name=__codelineno-18-2 href=#__codelineno-18-2></a>slaf<span class=w> </span>convert<span class=w> </span>pbmc3k.h5ad<span class=w> </span>pbmc3k.slaf
</span><span id=__span-18-3><a id=__codelineno-18-3 name=__codelineno-18-3 href=#__codelineno-18-3></a>
</span><span id=__span-18-4><a id=__codelineno-18-4 name=__codelineno-18-4 href=#__codelineno-18-4></a><span class=c1># Convert with verbose output to see details</span>
</span><span id=__span-18-5><a id=__codelineno-18-5 name=__codelineno-18-5 href=#__codelineno-18-5></a>slaf<span class=w> </span>convert<span class=w> </span>pbmc3k.h5ad<span class=w> </span>pbmc3k.slaf<span class=w> </span>--verbose
</span></code></pre></div> <p><strong>Exploring Datasets:</strong></p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-19-1><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=c1># Get basic info about a dataset</span>
</span><span id=__span-19-2><a id=__codelineno-19-2 name=__codelineno-19-2 href=#__codelineno-19-2></a>slaf<span class=w> </span>info<span class=w> </span>pbmc3k.slaf
</span><span id=__span-19-3><a id=__codelineno-19-3 name=__codelineno-19-3 href=#__codelineno-19-3></a>
</span><span id=__span-19-4><a id=__codelineno-19-4 name=__codelineno-19-4 href=#__codelineno-19-4></a><span class=c1># Run a simple query</span>
</span><span id=__span-19-5><a id=__codelineno-19-5 name=__codelineno-19-5 href=#__codelineno-19-5></a>slaf<span class=w> </span>query<span class=w> </span>pbmc3k.slaf<span class=w> </span><span class=s2>&quot;SELECT cell_type, COUNT(*) FROM cells GROUP BY cell_type&quot;</span>
</span><span id=__span-19-6><a id=__codelineno-19-6 name=__codelineno-19-6 href=#__codelineno-19-6></a>
</span><span id=__span-19-7><a id=__codelineno-19-7 name=__codelineno-19-7 href=#__codelineno-19-7></a><span class=c1># Export query results to CSV</span>
</span><span id=__span-19-8><a id=__codelineno-19-8 name=__codelineno-19-8 href=#__codelineno-19-8></a>slaf<span class=w> </span>query<span class=w> </span>pbmc3k.slaf<span class=w> </span><span class=s2>&quot;SELECT * FROM cells WHERE cell_type = &#39;T cells&#39;&quot;</span><span class=w> </span>--output<span class=w> </span>t_cells.csv
</span></code></pre></div> <p><strong>Data Analysis Pipeline:</strong></p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-20-1><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a><span class=c1># Convert input data</span>
</span><span id=__span-20-2><a id=__codelineno-20-2 name=__codelineno-20-2 href=#__codelineno-20-2></a>slaf<span class=w> </span>convert<span class=w> </span>input.h5ad<span class=w> </span>output.slaf
</span><span id=__span-20-3><a id=__codelineno-20-3 name=__codelineno-20-3 href=#__codelineno-20-3></a>
</span><span id=__span-20-4><a id=__codelineno-20-4 name=__codelineno-20-4 href=#__codelineno-20-4></a><span class=c1># Verify conversion</span>
</span><span id=__span-20-5><a id=__codelineno-20-5 name=__codelineno-20-5 href=#__codelineno-20-5></a>slaf<span class=w> </span>info<span class=w> </span>output.slaf
</span><span id=__span-20-6><a id=__codelineno-20-6 name=__codelineno-20-6 href=#__codelineno-20-6></a>
</span><span id=__span-20-7><a id=__codelineno-20-7 name=__codelineno-20-7 href=#__codelineno-20-7></a><span class=c1># Run analysis queries</span>
</span><span id=__span-20-8><a id=__codelineno-20-8 name=__codelineno-20-8 href=#__codelineno-20-8></a>slaf<span class=w> </span>query<span class=w> </span>output.slaf<span class=w> </span><span class=s2>&quot;SELECT cell_type, AVG(total_counts) FROM cells GROUP BY cell_type&quot;</span>
</span></code></pre></div> <h2 id=next-steps>Next Steps<a class=headerlink href=#next-steps title="Permanent link">&para;</a></h2> <ul> <li>Read the <a href=../../user-guide/how-slaf-works/ >User Guide</a> for detailed concepts</li> <li>Explore <a href=../../examples/getting-started/ >Examples</a> for real-world use cases</li> <li>Check the <a href=../../api/core/ >API Reference</a> for complete documentation</li> <li>Report bugs or make feature requests on <a href=https://github.com/slaf-project/slaf>GitHub</a></li> </ul> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="December 16, 2025 23:03:09 UTC">December 16, 2025</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="July 3, 2025 19:35:35 UTC">July 3, 2025</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Contributors> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 4a4 4 0 0 1 4 4 4 4 0 0 1-4 4 4 4 0 0 1-4-4 4 4 0 0 1 4-4m0 10c4.42 0 8 1.79 8 4v2H4v-2c0-2.21 3.58-4 8-4"/></svg> </span> <nav> Pavan Ramkumar </nav> </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2026 Pavan Ramkumar </div> </div> <div class=md-social> <a href=https://github.com/slaf-project/slaf target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.sections", "content.code.copy"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../../javascripts/mathjax.js></script> <script src=../../javascripts/theme-reset.js></script> <script src=../../javascripts/copy-enhancement.js></script> <script src=../../javascripts/theme-aware-assets.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>