<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Sparse Lazy Array Format - High-performance single-cell data storage and analysis"><meta name=author content="Pavan Ramkumar"><link href=https://slaf-project.github.io/slaf/development/benchmarks/ rel=canonical><link href=../maintaining/ rel=prev><link href=../../blog/ rel=next><link rel=icon href=../../assets/slaf-icon-transparent-dark-mono.svg><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>Benchmarks - SLAF Documentation</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../assets/_mkdocstrings.css><link rel=stylesheet href=../../stylesheets/extra.css><link rel=stylesheet href=../../stylesheets/logo-theme.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config",""),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#slaf-benchmark-system class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="SLAF Documentation" class="md-header__button md-logo" aria-label="SLAF Documentation" data-md-component=logo> <img src=../../assets/slaf-icon-transparent-dark-mono.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> SLAF Documentation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Benchmarks </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/slaf-project/slaf title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> slaf-project/slaf </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="SLAF Documentation" class="md-nav__button md-logo" aria-label="SLAF Documentation" data-md-component=logo> <img src=../../assets/slaf-icon-transparent-dark-mono.svg alt=logo> </a> SLAF Documentation </label> <div class=md-nav__source> <a href=https://github.com/slaf-project/slaf title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> slaf-project/slaf </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class=md-nav__item> <a href=../../getting-started/quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex> <span class=md-ellipsis> User Guide </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> User Guide </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../user-guide/how-slaf-works/ class=md-nav__link> <span class=md-ellipsis> How SLAF Works </span> </a> </li> <li class=md-nav__item> <a href=../../user-guide/migrating-to-slaf/ class=md-nav__link> <span class=md-ellipsis> Migrating to SLAF </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex> <span class=md-ellipsis> Examples </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Examples </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../examples/getting-started/ class=md-nav__link> <span class=md-ellipsis> Getting Started </span> </a> </li> <li class=md-nav__item> <a href=../../examples/lazy-processing/ class=md-nav__link> <span class=md-ellipsis> Lazy Processing </span> </a> </li> <li class=md-nav__item> <a href=../../examples/ml-training/ class=md-nav__link> <span class=md-ellipsis> ML Training </span> </a> </li> <li class=md-nav__item> <a href=../../examples/sql-queries/ class=md-nav__link> <span class=md-ellipsis> SQL Queries </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex> <span class=md-ellipsis> Benchmarks </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Benchmarks </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../benchmarks/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../benchmarks/bioinformatics_benchmarks/ class=md-nav__link> <span class=md-ellipsis> For Bioinformaticians </span> </a> </li> <li class=md-nav__item> <a href=../../benchmarks/ml_benchmarks/ class=md-nav__link> <span class=md-ellipsis> For ML Engineers </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/core/ class=md-nav__link> <span class=md-ellipsis> Core </span> </a> </li> <li class=md-nav__item> <a href=../../api/data/ class=md-nav__link> <span class=md-ellipsis> Data </span> </a> </li> <li class=md-nav__item> <a href=../../api/integrations/ class=md-nav__link> <span class=md-ellipsis> Integrations </span> </a> </li> <li class=md-nav__item> <a href=../../api/ml/ class=md-nav__link> <span class=md-ellipsis> ML </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_7 checked> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex> <span class=md-ellipsis> Development </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=true> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Development </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../contributing/ class=md-nav__link> <span class=md-ellipsis> For Contributors </span> </a> </li> <li class=md-nav__item> <a href=../maintaining/ class=md-nav__link> <span class=md-ellipsis> For Maintainers </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Benchmarks </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Benchmarks </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#quick-start-recommended class=md-nav__link> <span class=md-ellipsis> üöÄ Quick Start (Recommended) </span> </a> <nav class=md-nav aria-label="üöÄ Quick Start (Recommended)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bioinformatics-benchmarks-cli-integration class=md-nav__link> <span class=md-ellipsis> Bioinformatics Benchmarks (CLI Integration) </span> </a> </li> <li class=md-nav__item> <a href=#ml-benchmarks-standalone-scripts class=md-nav__link> <span class=md-ellipsis> ML Benchmarks (Standalone Scripts) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#file-structure class=md-nav__link> <span class=md-ellipsis> üìÅ File Structure </span> </a> <nav class=md-nav aria-label="üìÅ File Structure"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#core-files class=md-nav__link> <span class=md-ellipsis> Core Files </span> </a> </li> <li class=md-nav__item> <a href=#bioinformatics-benchmark-modules-cli-integrated class=md-nav__link> <span class=md-ellipsis> Bioinformatics Benchmark Modules (CLI Integrated) </span> </a> </li> <li class=md-nav__item> <a href=#ml-benchmark-modules-standalone class=md-nav__link> <span class=md-ellipsis> ML Benchmark Modules (Standalone) </span> </a> </li> <li class=md-nav__item> <a href=#output-files class=md-nav__link> <span class=md-ellipsis> Output Files </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#cli-commands-bioinformatics-only class=md-nav__link> <span class=md-ellipsis> üîß CLI Commands (Bioinformatics Only) </span> </a> <nav class=md-nav aria-label="üîß CLI Commands (Bioinformatics Only)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#run-benchmarks class=md-nav__link> <span class=md-ellipsis> Run Benchmarks </span> </a> </li> <li class=md-nav__item> <a href=#generate-summary class=md-nav__link> <span class=md-ellipsis> Generate Summary </span> </a> </li> <li class=md-nav__item> <a href=#update-documentation class=md-nav__link> <span class=md-ellipsis> Update Documentation </span> </a> </li> <li class=md-nav__item> <a href=#complete-workflow class=md-nav__link> <span class=md-ellipsis> Complete Workflow </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#available-benchmark-types class=md-nav__link> <span class=md-ellipsis> üìä Available Benchmark Types </span> </a> <nav class=md-nav aria-label="üìä Available Benchmark Types"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bioinformatics-benchmarks-cli-integrated class=md-nav__link> <span class=md-ellipsis> Bioinformatics Benchmarks (CLI Integrated) </span> </a> </li> <li class=md-nav__item> <a href=#ml-benchmarks-standalone-scripts_1 class=md-nav__link> <span class=md-ellipsis> ML Benchmarks (Standalone Scripts) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#usage-examples class=md-nav__link> <span class=md-ellipsis> üéØ Usage Examples </span> </a> <nav class=md-nav aria-label="üéØ Usage Examples"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bioinformatics-development-workflow class=md-nav__link> <span class=md-ellipsis> Bioinformatics Development Workflow </span> </a> </li> <li class=md-nav__item> <a href=#ml-development-workflow class=md-nav__link> <span class=md-ellipsis> ML Development Workflow </span> </a> </li> <li class=md-nav__item> <a href=#performance-analysis class=md-nav__link> <span class=md-ellipsis> Performance Analysis </span> </a> </li> <li class=md-nav__item> <a href=#multi-dataset-testing class=md-nav__link> <span class=md-ellipsis> Multi-Dataset Testing </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#output-files_1 class=md-nav__link> <span class=md-ellipsis> üìà Output Files </span> </a> <nav class=md-nav aria-label="üìà Output Files"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bioinformatics-results-files class=md-nav__link> <span class=md-ellipsis> Bioinformatics Results Files </span> </a> </li> <li class=md-nav__item> <a href=#ml-results-files class=md-nav__link> <span class=md-ellipsis> ML Results Files </span> </a> </li> <li class=md-nav__item> <a href=#documentation-integration class=md-nav__link> <span class=md-ellipsis> Documentation Integration </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> üîç Troubleshooting </span> </a> <nav class=md-nav aria-label="üîç Troubleshooting"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#common-issues class=md-nav__link> <span class=md-ellipsis> Common Issues </span> </a> </li> <li class=md-nav__item> <a href=#debug-mode class=md-nav__link> <span class=md-ellipsis> Debug Mode </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#contributing class=md-nav__link> <span class=md-ellipsis> üìù Contributing </span> </a> <nav class=md-nav aria-label="üìù Contributing"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#adding-bioinformatics-benchmarks class=md-nav__link> <span class=md-ellipsis> Adding Bioinformatics Benchmarks </span> </a> </li> <li class=md-nav__item> <a href=#adding-ml-benchmarks class=md-nav__link> <span class=md-ellipsis> Adding ML Benchmarks </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#architecture class=md-nav__link> <span class=md-ellipsis> üèóÔ∏è Architecture </span> </a> <nav class=md-nav aria-label="üèóÔ∏è Architecture"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bioinformatics-benchmarks-cli-integrated_1 class=md-nav__link> <span class=md-ellipsis> Bioinformatics Benchmarks (CLI Integrated) </span> </a> </li> <li class=md-nav__item> <a href=#ml-benchmarks-standalone class=md-nav__link> <span class=md-ellipsis> ML Benchmarks (Standalone) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#future-integration class=md-nav__link> <span class=md-ellipsis> üîÑ Future Integration </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex> <span class=md-ellipsis> Blog </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> Blog </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../blog/ class=md-nav__link> <span class=md-ellipsis> Blog Home </span> </a> </li> <li class=md-nav__item> <a href=../../blog/huggingface/ class=md-nav__link> <span class=md-ellipsis> SLAF on Hugging Face </span> </a> </li> <li class=md-nav__item> <a href=../../blog/introducing-slaf/ class=md-nav__link> <span class=md-ellipsis> Introducing SLAF </span> </a> </li> <li class=md-nav__item> <a href=../../blog/blazing-fast-dataloaders/ class=md-nav__link> <span class=md-ellipsis> Blazing Fast Dataloaders </span> </a> </li> <li class=md-nav__item> <a href=../../blog/blazing-fast-dataloaders-2/ class=md-nav__link> <span class=md-ellipsis> Mixture of Scanners </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#quick-start-recommended class=md-nav__link> <span class=md-ellipsis> üöÄ Quick Start (Recommended) </span> </a> <nav class=md-nav aria-label="üöÄ Quick Start (Recommended)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bioinformatics-benchmarks-cli-integration class=md-nav__link> <span class=md-ellipsis> Bioinformatics Benchmarks (CLI Integration) </span> </a> </li> <li class=md-nav__item> <a href=#ml-benchmarks-standalone-scripts class=md-nav__link> <span class=md-ellipsis> ML Benchmarks (Standalone Scripts) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#file-structure class=md-nav__link> <span class=md-ellipsis> üìÅ File Structure </span> </a> <nav class=md-nav aria-label="üìÅ File Structure"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#core-files class=md-nav__link> <span class=md-ellipsis> Core Files </span> </a> </li> <li class=md-nav__item> <a href=#bioinformatics-benchmark-modules-cli-integrated class=md-nav__link> <span class=md-ellipsis> Bioinformatics Benchmark Modules (CLI Integrated) </span> </a> </li> <li class=md-nav__item> <a href=#ml-benchmark-modules-standalone class=md-nav__link> <span class=md-ellipsis> ML Benchmark Modules (Standalone) </span> </a> </li> <li class=md-nav__item> <a href=#output-files class=md-nav__link> <span class=md-ellipsis> Output Files </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#cli-commands-bioinformatics-only class=md-nav__link> <span class=md-ellipsis> üîß CLI Commands (Bioinformatics Only) </span> </a> <nav class=md-nav aria-label="üîß CLI Commands (Bioinformatics Only)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#run-benchmarks class=md-nav__link> <span class=md-ellipsis> Run Benchmarks </span> </a> </li> <li class=md-nav__item> <a href=#generate-summary class=md-nav__link> <span class=md-ellipsis> Generate Summary </span> </a> </li> <li class=md-nav__item> <a href=#update-documentation class=md-nav__link> <span class=md-ellipsis> Update Documentation </span> </a> </li> <li class=md-nav__item> <a href=#complete-workflow class=md-nav__link> <span class=md-ellipsis> Complete Workflow </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#available-benchmark-types class=md-nav__link> <span class=md-ellipsis> üìä Available Benchmark Types </span> </a> <nav class=md-nav aria-label="üìä Available Benchmark Types"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bioinformatics-benchmarks-cli-integrated class=md-nav__link> <span class=md-ellipsis> Bioinformatics Benchmarks (CLI Integrated) </span> </a> </li> <li class=md-nav__item> <a href=#ml-benchmarks-standalone-scripts_1 class=md-nav__link> <span class=md-ellipsis> ML Benchmarks (Standalone Scripts) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#usage-examples class=md-nav__link> <span class=md-ellipsis> üéØ Usage Examples </span> </a> <nav class=md-nav aria-label="üéØ Usage Examples"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bioinformatics-development-workflow class=md-nav__link> <span class=md-ellipsis> Bioinformatics Development Workflow </span> </a> </li> <li class=md-nav__item> <a href=#ml-development-workflow class=md-nav__link> <span class=md-ellipsis> ML Development Workflow </span> </a> </li> <li class=md-nav__item> <a href=#performance-analysis class=md-nav__link> <span class=md-ellipsis> Performance Analysis </span> </a> </li> <li class=md-nav__item> <a href=#multi-dataset-testing class=md-nav__link> <span class=md-ellipsis> Multi-Dataset Testing </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#output-files_1 class=md-nav__link> <span class=md-ellipsis> üìà Output Files </span> </a> <nav class=md-nav aria-label="üìà Output Files"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bioinformatics-results-files class=md-nav__link> <span class=md-ellipsis> Bioinformatics Results Files </span> </a> </li> <li class=md-nav__item> <a href=#ml-results-files class=md-nav__link> <span class=md-ellipsis> ML Results Files </span> </a> </li> <li class=md-nav__item> <a href=#documentation-integration class=md-nav__link> <span class=md-ellipsis> Documentation Integration </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#troubleshooting class=md-nav__link> <span class=md-ellipsis> üîç Troubleshooting </span> </a> <nav class=md-nav aria-label="üîç Troubleshooting"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#common-issues class=md-nav__link> <span class=md-ellipsis> Common Issues </span> </a> </li> <li class=md-nav__item> <a href=#debug-mode class=md-nav__link> <span class=md-ellipsis> Debug Mode </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#contributing class=md-nav__link> <span class=md-ellipsis> üìù Contributing </span> </a> <nav class=md-nav aria-label="üìù Contributing"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#adding-bioinformatics-benchmarks class=md-nav__link> <span class=md-ellipsis> Adding Bioinformatics Benchmarks </span> </a> </li> <li class=md-nav__item> <a href=#adding-ml-benchmarks class=md-nav__link> <span class=md-ellipsis> Adding ML Benchmarks </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#architecture class=md-nav__link> <span class=md-ellipsis> üèóÔ∏è Architecture </span> </a> <nav class=md-nav aria-label="üèóÔ∏è Architecture"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#bioinformatics-benchmarks-cli-integrated_1 class=md-nav__link> <span class=md-ellipsis> Bioinformatics Benchmarks (CLI Integrated) </span> </a> </li> <li class=md-nav__item> <a href=#ml-benchmarks-standalone class=md-nav__link> <span class=md-ellipsis> ML Benchmarks (Standalone) </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#future-integration class=md-nav__link> <span class=md-ellipsis> üîÑ Future Integration </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=slaf-benchmark-system>SLAF Benchmark System<a class=headerlink href=#slaf-benchmark-system title="Permanent link">&para;</a></h1> <p>This document describes the SLAF benchmark suite for performance testing and documentation generation. The benchmark system has been refactored to separate bioinformatics and ML benchmarks.</p> <h2 id=quick-start-recommended>üöÄ Quick Start (Recommended)<a class=headerlink href=#quick-start-recommended title="Permanent link">&para;</a></h2> <h3 id=bioinformatics-benchmarks-cli-integration>Bioinformatics Benchmarks (CLI Integration)<a class=headerlink href=#bioinformatics-benchmarks-cli-integration title="Permanent link">&para;</a></h3> <p>Use the unified CLI interface for bioinformatics benchmark operations:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1># Run bioinformatics benchmarks</span>
</span><span id=__span-0-2><a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>slaf<span class=w> </span>benchmark<span class=w> </span>run<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>--types<span class=w> </span>cell_filtering,expression_queries<span class=w> </span>--verbose
</span><span id=__span-0-3><a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a>
</span><span id=__span-0-4><a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=c1># Generate summary from results</span>
</span><span id=__span-0-5><a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a>slaf<span class=w> </span>benchmark<span class=w> </span>summary<span class=w> </span>--results<span class=w> </span>comprehensive_benchmark_results.json
</span><span id=__span-0-6><a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a>
</span><span id=__span-0-7><a id=__codelineno-0-7 name=__codelineno-0-7 href=#__codelineno-0-7></a><span class=c1># Update documentation</span>
</span><span id=__span-0-8><a id=__codelineno-0-8 name=__codelineno-0-8 href=#__codelineno-0-8></a>slaf<span class=w> </span>benchmark<span class=w> </span>docs<span class=w> </span>--summary<span class=w> </span>benchmark_summary.json
</span><span id=__span-0-9><a id=__codelineno-0-9 name=__codelineno-0-9 href=#__codelineno-0-9></a>
</span><span id=__span-0-10><a id=__codelineno-0-10 name=__codelineno-0-10 href=#__codelineno-0-10></a><span class=c1># Run complete workflow</span>
</span><span id=__span-0-11><a id=__codelineno-0-11 name=__codelineno-0-11 href=#__codelineno-0-11></a>slaf<span class=w> </span>benchmark<span class=w> </span>all<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>--auto-convert
</span></code></pre></div> <h3 id=ml-benchmarks-standalone-scripts>ML Benchmarks (Standalone Scripts)<a class=headerlink href=#ml-benchmarks-standalone-scripts title="Permanent link">&para;</a></h3> <p>ML benchmarks are run as standalone scripts:</p> <div class="language-bash highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=c1># External dataloader comparisons</span>
</span><span id=__span-1-2><a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>python<span class=w> </span>benchmarks/benchmark_dataloaders_external.py
</span><span id=__span-1-3><a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>
</span><span id=__span-1-4><a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=c1># Internal tokenization strategies</span>
</span><span id=__span-1-5><a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a>python<span class=w> </span>benchmarks/benchmark_dataloaders_internal.py
</span><span id=__span-1-6><a id=__codelineno-1-6 name=__codelineno-1-6 href=#__codelineno-1-6></a>
</span><span id=__span-1-7><a id=__codelineno-1-7 name=__codelineno-1-7 href=#__codelineno-1-7></a><span class=c1># Prefetcher performance analysis</span>
</span><span id=__span-1-8><a id=__codelineno-1-8 name=__codelineno-1-8 href=#__codelineno-1-8></a>python<span class=w> </span>benchmarks/benchmark_prefetcher.py
</span></code></pre></div> <h2 id=file-structure>üìÅ File Structure<a class=headerlink href=#file-structure title="Permanent link">&para;</a></h2> <h3 id=core-files>Core Files<a class=headerlink href=#core-files title="Permanent link">&para;</a></h3> <ul> <li><code>benchmarks/benchmark.py</code> - Main bioinformatics benchmark runner with CLI integration</li> <li><code>benchmarks/benchmark_utils.py</code> - Shared utilities for bioinformatics benchmarks</li> </ul> <h3 id=bioinformatics-benchmark-modules-cli-integrated>Bioinformatics Benchmark Modules (CLI Integrated)<a class=headerlink href=#bioinformatics-benchmark-modules-cli-integrated title="Permanent link">&para;</a></h3> <ul> <li><code>benchmarks/benchmark_cell_filtering.py</code> - Cell filtering performance tests</li> <li><code>benchmarks/benchmark_gene_filtering.py</code> - Gene filtering performance tests</li> <li><code>benchmarks/benchmark_expression_queries.py</code> - Expression query performance tests</li> <li><code>benchmarks/benchmark_anndata_ops.py</code> - AnnData operation performance tests</li> <li><code>benchmarks/benchmark_scanpy_preprocessing.py</code> - Scanpy preprocessing performance tests</li> </ul> <h3 id=ml-benchmark-modules-standalone>ML Benchmark Modules (Standalone)<a class=headerlink href=#ml-benchmark-modules-standalone title="Permanent link">&para;</a></h3> <ul> <li><code>benchmarks/benchmark_dataloaders_external.py</code> - External dataloader comparisons (SLAF vs scDataset, BioNeMo, etc.)</li> <li><code>benchmarks/benchmark_dataloaders_internal.py</code> - Internal tokenization strategy comparisons (scGPT, Geneformer, etc.)</li> <li><code>benchmarks/benchmark_prefetcher.py</code> - Prefetcher pipeline performance analysis</li> </ul> <h3 id=output-files>Output Files<a class=headerlink href=#output-files title="Permanent link">&para;</a></h3> <ul> <li><code>benchmarks/comprehensive_benchmark_results.json</code> - Complete bioinformatics benchmark results</li> <li><code>benchmarks/benchmark_summary.json</code> - Documentation-ready summary</li> <li><code>benchmarks/benchmark_output.txt</code> - Detailed benchmark output</li> <li><code>benchmarks/benchmark_results.json</code> - Legacy results file</li> </ul> <h2 id=cli-commands-bioinformatics-only>üîß CLI Commands (Bioinformatics Only)<a class=headerlink href=#cli-commands-bioinformatics-only title="Permanent link">&para;</a></h2> <h3 id=run-benchmarks>Run Benchmarks<a class=headerlink href=#run-benchmarks title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1># Run all bioinformatics benchmark types</span>
</span><span id=__span-2-2><a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a>slaf<span class=w> </span>benchmark<span class=w> </span>run<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>--auto-convert
</span><span id=__span-2-3><a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a>
</span><span id=__span-2-4><a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a><span class=c1># Run specific benchmark types</span>
</span><span id=__span-2-5><a id=__codelineno-2-5 name=__codelineno-2-5 href=#__codelineno-2-5></a>slaf<span class=w> </span>benchmark<span class=w> </span>run<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>--types<span class=w> </span>cell_filtering,expression_queries
</span><span id=__span-2-6><a id=__codelineno-2-6 name=__codelineno-2-6 href=#__codelineno-2-6></a>
</span><span id=__span-2-7><a id=__codelineno-2-7 name=__codelineno-2-7 href=#__codelineno-2-7></a><span class=c1># Run with verbose output</span>
</span><span id=__span-2-8><a id=__codelineno-2-8 name=__codelineno-2-8 href=#__codelineno-2-8></a>slaf<span class=w> </span>benchmark<span class=w> </span>run<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>--verbose<span class=w> </span>--auto-convert
</span><span id=__span-2-9><a id=__codelineno-2-9 name=__codelineno-2-9 href=#__codelineno-2-9></a>
</span><span id=__span-2-10><a id=__codelineno-2-10 name=__codelineno-2-10 href=#__codelineno-2-10></a><span class=c1># Run on multiple datasets</span>
</span><span id=__span-2-11><a id=__codelineno-2-11 name=__codelineno-2-11 href=#__codelineno-2-11></a>slaf<span class=w> </span>benchmark<span class=w> </span>run<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>pbmc_68k<span class=w> </span>--auto-convert
</span></code></pre></div> <h3 id=generate-summary>Generate Summary<a class=headerlink href=#generate-summary title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-3-1><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1># Generate summary from existing results</span>
</span><span id=__span-3-2><a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a>slaf<span class=w> </span>benchmark<span class=w> </span>summary<span class=w> </span>--results<span class=w> </span>comprehensive_benchmark_results.json
</span><span id=__span-3-3><a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a>
</span><span id=__span-3-4><a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a><span class=c1># Generate summary with custom output</span>
</span><span id=__span-3-5><a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a>slaf<span class=w> </span>benchmark<span class=w> </span>summary<span class=w> </span>--results<span class=w> </span>comprehensive_benchmark_results.json<span class=w> </span>--output<span class=w> </span>custom_summary.json
</span></code></pre></div> <h3 id=update-documentation>Update Documentation<a class=headerlink href=#update-documentation title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-4-1><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1># Update bioinformatics_benchmarks.md with summary data</span>
</span><span id=__span-4-2><a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>slaf<span class=w> </span>benchmark<span class=w> </span>docs<span class=w> </span>--summary<span class=w> </span>benchmark_summary.json
</span><span id=__span-4-3><a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>
</span><span id=__span-4-4><a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a><span class=c1># Update with custom summary file</span>
</span><span id=__span-4-5><a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a>slaf<span class=w> </span>benchmark<span class=w> </span>docs<span class=w> </span>--summary<span class=w> </span>custom_summary.json
</span></code></pre></div> <h3 id=complete-workflow>Complete Workflow<a class=headerlink href=#complete-workflow title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-5-1><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=c1># Run benchmarks, generate summary, and update docs</span>
</span><span id=__span-5-2><a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a>slaf<span class=w> </span>benchmark<span class=w> </span>all<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>--auto-convert<span class=w> </span>--verbose
</span></code></pre></div> <h2 id=available-benchmark-types>üìä Available Benchmark Types<a class=headerlink href=#available-benchmark-types title="Permanent link">&para;</a></h2> <h3 id=bioinformatics-benchmarks-cli-integrated>Bioinformatics Benchmarks (CLI Integrated)<a class=headerlink href=#bioinformatics-benchmarks-cli-integrated title="Permanent link">&para;</a></h3> <ul> <li><strong>cell_filtering</strong> - Metadata-based cell filtering performance</li> <li><strong>gene_filtering</strong> - Metadata-based gene filtering performance</li> <li><strong>expression_queries</strong> - Expression matrix slicing performance</li> <li><strong>anndata_ops</strong> - AnnData operation performance</li> <li><strong>scanpy_preprocessing</strong> - Scanpy preprocessing pipeline performance</li> </ul> <h3 id=ml-benchmarks-standalone-scripts_1>ML Benchmarks (Standalone Scripts)<a class=headerlink href=#ml-benchmarks-standalone-scripts_1 title="Permanent link">&para;</a></h3> <ul> <li><strong>External Dataloader Comparisons</strong> - SLAF vs scDataset, BioNeMo SCDL, AnnDataLoader</li> <li><strong>Internal Tokenization Strategies</strong> - scGPT, Geneformer, raw data loading</li> <li><strong>Prefetcher Performance</strong> - Pipeline timing analysis across configurations</li> </ul> <h2 id=usage-examples>üéØ Usage Examples<a class=headerlink href=#usage-examples title="Permanent link">&para;</a></h2> <h3 id=bioinformatics-development-workflow>Bioinformatics Development Workflow<a class=headerlink href=#bioinformatics-development-workflow title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-6-1><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=c1># Quick test of cell filtering</span>
</span><span id=__span-6-2><a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a>slaf<span class=w> </span>benchmark<span class=w> </span>run<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>--types<span class=w> </span>cell_filtering<span class=w> </span>--verbose
</span><span id=__span-6-3><a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a>
</span><span id=__span-6-4><a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a><span class=c1># Comprehensive testing</span>
</span><span id=__span-6-5><a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a>slaf<span class=w> </span>benchmark<span class=w> </span>all<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>--auto-convert<span class=w> </span>--verbose
</span></code></pre></div> <h3 id=ml-development-workflow>ML Development Workflow<a class=headerlink href=#ml-development-workflow title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-7-1><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># Compare against external dataloaders</span>
</span><span id=__span-7-2><a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a>python<span class=w> </span>benchmarks/benchmark_dataloaders_external.py
</span><span id=__span-7-3><a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a>
</span><span id=__span-7-4><a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a><span class=c1># Test different tokenization strategies</span>
</span><span id=__span-7-5><a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a>python<span class=w> </span>benchmarks/benchmark_dataloaders_internal.py
</span><span id=__span-7-6><a id=__codelineno-7-6 name=__codelineno-7-6 href=#__codelineno-7-6></a>
</span><span id=__span-7-7><a id=__codelineno-7-7 name=__codelineno-7-7 href=#__codelineno-7-7></a><span class=c1># Analyze prefetcher performance</span>
</span><span id=__span-7-8><a id=__codelineno-7-8 name=__codelineno-7-8 href=#__codelineno-7-8></a>python<span class=w> </span>benchmarks/benchmark_prefetcher.py
</span></code></pre></div> <h3 id=performance-analysis>Performance Analysis<a class=headerlink href=#performance-analysis title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-8-1><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=c1># Generate bioinformatics performance summary</span>
</span><span id=__span-8-2><a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a>slaf<span class=w> </span>benchmark<span class=w> </span>summary<span class=w> </span>--results<span class=w> </span>comprehensive_benchmark_results.json
</span><span id=__span-8-3><a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a>
</span><span id=__span-8-4><a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=c1># Update bioinformatics documentation with latest results</span>
</span><span id=__span-8-5><a id=__codelineno-8-5 name=__codelineno-8-5 href=#__codelineno-8-5></a>slaf<span class=w> </span>benchmark<span class=w> </span>docs<span class=w> </span>--summary<span class=w> </span>benchmark_summary.json
</span></code></pre></div> <h3 id=multi-dataset-testing>Multi-Dataset Testing<a class=headerlink href=#multi-dataset-testing title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-9-1><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=c1># Test bioinformatics benchmarks on multiple datasets</span>
</span><span id=__span-9-2><a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a>slaf<span class=w> </span>benchmark<span class=w> </span>run<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>pbmc_68k<span class=w> </span>--types<span class=w> </span>cell_filtering,expression_queries<span class=w> </span>--auto-convert
</span></code></pre></div> <h2 id=output-files_1>üìà Output Files<a class=headerlink href=#output-files_1 title="Permanent link">&para;</a></h2> <h3 id=bioinformatics-results-files>Bioinformatics Results Files<a class=headerlink href=#bioinformatics-results-files title="Permanent link">&para;</a></h3> <ul> <li><code>comprehensive_benchmark_results.json</code> - Complete benchmark results with detailed timing and memory data</li> <li><code>benchmark_summary.json</code> - Condensed summary for documentation updates</li> <li><code>benchmark_output.txt</code> - Human-readable benchmark output with tables and analysis</li> </ul> <h3 id=ml-results-files>ML Results Files<a class=headerlink href=#ml-results-files title="Permanent link">&para;</a></h3> <ul> <li>ML benchmarks output results directly to console with rich formatting</li> <li>Results are not automatically saved to files (manual documentation updates required)</li> </ul> <h3 id=documentation-integration>Documentation Integration<a class=headerlink href=#documentation-integration title="Permanent link">&para;</a></h3> <p>The bioinformatics benchmark system automatically updates <code>docs/benchmarks/bioinformatics_benchmarks.md</code> with the latest performance data, ensuring documentation stays current with benchmark results. ML benchmarks are documented separately in <code>docs/benchmarks/ml_benchmarks.md</code> and require manual updates.</p> <h2 id=troubleshooting>üîç Troubleshooting<a class=headerlink href=#troubleshooting title="Permanent link">&para;</a></h2> <h3 id=common-issues>Common Issues<a class=headerlink href=#common-issues title="Permanent link">&para;</a></h3> <ol> <li><strong>Dataset not found</strong>: Ensure datasets are in the correct directory and use <code>--auto-convert</code> to convert h5ad files</li> <li><strong>Benchmark failures</strong>: Check that SLAF files exist and are properly formatted</li> <li><strong>Memory issues</strong>: Some benchmarks require significant memory for large datasets</li> <li><strong>ML benchmark dependencies</strong>: Ensure all ML dependencies are installed for standalone ML benchmarks</li> </ol> <h3 id=debug-mode>Debug Mode<a class=headerlink href=#debug-mode title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-10-1><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=c1># Run bioinformatics benchmarks with verbose output for debugging</span>
</span><span id=__span-10-2><a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a>slaf<span class=w> </span>benchmark<span class=w> </span>run<span class=w> </span>--datasets<span class=w> </span>pbmc3k_processed<span class=w> </span>--types<span class=w> </span>cell_filtering<span class=w> </span>--verbose
</span><span id=__span-10-3><a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a>
</span><span id=__span-10-4><a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=c1># Run ML benchmarks with debug output</span>
</span><span id=__span-10-5><a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a>python<span class=w> </span>benchmarks/benchmark_dataloaders_external.py<span class=w> </span>--debug
</span></code></pre></div> <h2 id=contributing>üìù Contributing<a class=headerlink href=#contributing title="Permanent link">&para;</a></h2> <h3 id=adding-bioinformatics-benchmarks>Adding Bioinformatics Benchmarks<a class=headerlink href=#adding-bioinformatics-benchmarks title="Permanent link">&para;</a></h3> <p>When adding new bioinformatics benchmarks:</p> <ol> <li>Create a new benchmark module following the existing pattern</li> <li>Add the benchmark type to the CLI in <code>slaf/cli.py</code></li> <li>Update this documentation with the new benchmark type</li> <li>Test with <code>slaf benchmark run --types your_new_benchmark</code></li> </ol> <h3 id=adding-ml-benchmarks>Adding ML Benchmarks<a class=headerlink href=#adding-ml-benchmarks title="Permanent link">&para;</a></h3> <p>When adding new ML benchmarks:</p> <ol> <li>Create a new standalone benchmark script following the existing pattern</li> <li>Add appropriate documentation in <code>docs/benchmarks/ml_benchmarks.md</code></li> <li>Test the standalone script directly</li> <li>Consider integration with CLI system in the future</li> </ol> <h2 id=architecture>üèóÔ∏è Architecture<a class=headerlink href=#architecture title="Permanent link">&para;</a></h2> <p>The benchmark system uses a modular design with two distinct approaches:</p> <h3 id=bioinformatics-benchmarks-cli-integrated_1>Bioinformatics Benchmarks (CLI Integrated)<a class=headerlink href=#bioinformatics-benchmarks-cli-integrated_1 title="Permanent link">&para;</a></h3> <ul> <li><strong>CLI Interface</strong>: Unified command-line interface in <code>slaf/cli.py</code></li> <li><strong>Benchmark Runner</strong>: Main orchestration in <code>benchmarks/benchmark.py</code></li> <li><strong>Individual Modules</strong>: Specialized benchmark tests in separate files</li> <li><strong>Utilities</strong>: Shared functions in <code>benchmarks/benchmark_utils.py</code></li> <li><strong>Documentation</strong>: Automatic updates to <code>docs/benchmarks/bioinformatics_benchmarks.md</code></li> </ul> <h3 id=ml-benchmarks-standalone>ML Benchmarks (Standalone)<a class=headerlink href=#ml-benchmarks-standalone title="Permanent link">&para;</a></h3> <ul> <li><strong>Standalone Scripts</strong>: Independent benchmark scripts with rich console output</li> <li><strong>External Comparisons</strong>: <code>benchmark_dataloaders_external.py</code> for competitor analysis</li> <li><strong>Internal Analysis</strong>: <code>benchmark_dataloaders_internal.py</code> for tokenization strategies</li> <li><strong>Pipeline Analysis</strong>: <code>benchmark_prefetcher.py</code> for prefetcher performance</li> <li><strong>Documentation</strong>: Manual updates to <code>docs/benchmarks/ml_benchmarks.md</code></li> </ul> <h2 id=future-integration>üîÑ Future Integration<a class=headerlink href=#future-integration title="Permanent link">&para;</a></h2> <p>The ML benchmarks are currently standalone but may be integrated with the CLI system in the future to provide:</p> <ul> <li>Unified benchmark execution</li> <li>Automatic result aggregation</li> <li>Integrated documentation updates</li> <li>Consistent output formatting</li> </ul> <p>For now, ML benchmarks provide immediate value as standalone tools for development and performance analysis.</p> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="July 28, 2025 19:41:23 UTC">July 28, 2025</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="July 12, 2025 23:14:02 UTC">July 12, 2025</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Contributors> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 4a4 4 0 0 1 4 4 4 4 0 0 1-4 4 4 4 0 0 1-4-4 4 4 0 0 1 4-4m0 10c4.42 0 8 1.79 8 4v2H4v-2c0-2.21 3.58-4 8-4"/></svg> </span> <nav> Pavan Ramkumar </nav> </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2026 Pavan Ramkumar </div> </div> <div class=md-social> <a href=https://github.com/slaf-project/slaf target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.sections", "content.code.copy"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../../javascripts/mathjax.js></script> <script src=../../javascripts/theme-reset.js></script> <script src=../../javascripts/copy-enhancement.js></script> <script src=../../javascripts/theme-aware-assets.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>