<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Sparse Lazy Array Format - High-performance single-cell data storage and analysis"><meta name=author content="Pavan Ramkumar"><link href=https://slaf-project.github.io/slaf/blog/introducing-slaf/ rel=canonical><link href=../huggingface/ rel=prev><link href=../blazing-fast-dataloaders/ rel=next><link rel=icon href=../../assets/slaf-icon-transparent-dark-mono.svg><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>Introducing SLAF - SLAF Documentation</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../assets/_mkdocstrings.css><link rel=stylesheet href=../../stylesheets/extra.css><link rel=stylesheet href=../../stylesheets/logo-theme.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config",""),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#introducing-slaf-the-single-cell-data-format-for-the-virtual-cell-era class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="SLAF Documentation" class="md-header__button md-logo" aria-label="SLAF Documentation" data-md-component=logo> <img src=../../assets/slaf-icon-transparent-dark-mono.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> SLAF Documentation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Introducing SLAF </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/slaf-project/slaf title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> slaf-project/slaf </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="SLAF Documentation" class="md-nav__button md-logo" aria-label="SLAF Documentation" data-md-component=logo> <img src=../../assets/slaf-icon-transparent-dark-mono.svg alt=logo> </a> SLAF Documentation </label> <div class=md-nav__source> <a href=https://github.com/slaf-project/slaf title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> slaf-project/slaf </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class=md-nav__item> <a href=../../getting-started/quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex> <span class=md-ellipsis> User Guide </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> User Guide </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../user-guide/how-slaf-works/ class=md-nav__link> <span class=md-ellipsis> How SLAF Works </span> </a> </li> <li class=md-nav__item> <a href=../../user-guide/migrating-to-slaf/ class=md-nav__link> <span class=md-ellipsis> Migrating to SLAF </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex> <span class=md-ellipsis> Examples </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Examples </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../examples/getting-started/ class=md-nav__link> <span class=md-ellipsis> Getting Started </span> </a> </li> <li class=md-nav__item> <a href=../../examples/lazy-processing/ class=md-nav__link> <span class=md-ellipsis> Lazy Processing </span> </a> </li> <li class=md-nav__item> <a href=../../examples/ml-training/ class=md-nav__link> <span class=md-ellipsis> ML Training </span> </a> </li> <li class=md-nav__item> <a href=../../examples/sql-queries/ class=md-nav__link> <span class=md-ellipsis> SQL Queries </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex> <span class=md-ellipsis> Benchmarks </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Benchmarks </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../benchmarks/ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../../benchmarks/bioinformatics_benchmarks/ class=md-nav__link> <span class=md-ellipsis> For Bioinformaticians </span> </a> </li> <li class=md-nav__item> <a href=../../benchmarks/ml_benchmarks/ class=md-nav__link> <span class=md-ellipsis> For ML Engineers </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/core/ class=md-nav__link> <span class=md-ellipsis> Core </span> </a> </li> <li class=md-nav__item> <a href=../../api/data/ class=md-nav__link> <span class=md-ellipsis> Data </span> </a> </li> <li class=md-nav__item> <a href=../../api/integrations/ class=md-nav__link> <span class=md-ellipsis> Integrations </span> </a> </li> <li class=md-nav__item> <a href=../../api/ml/ class=md-nav__link> <span class=md-ellipsis> ML </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex> <span class=md-ellipsis> Development </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Development </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../development/contributing/ class=md-nav__link> <span class=md-ellipsis> For Contributors </span> </a> </li> <li class=md-nav__item> <a href=../../development/maintaining/ class=md-nav__link> <span class=md-ellipsis> For Maintainers </span> </a> </li> <li class=md-nav__item> <a href=../../development/benchmarks/ class=md-nav__link> <span class=md-ellipsis> Benchmarks </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_8 checked> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex> <span class=md-ellipsis> Blog </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=true> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> Blog </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../ class=md-nav__link> <span class=md-ellipsis> Blog Home </span> </a> </li> <li class=md-nav__item> <a href=../huggingface/ class=md-nav__link> <span class=md-ellipsis> SLAF on Hugging Face </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Introducing SLAF </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Introducing SLAF </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#a-10-year-lookback-at-storage-and-compute-revolutions class=md-nav__link> <span class=md-ellipsis> A 10-year lookback at storage and compute revolutions </span> </a> <nav class=md-nav aria-label="A 10-year lookback at storage and compute revolutions"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#the-zarr-revolution-compressed-chunked-storage class=md-nav__link> <span class=md-ellipsis> The Zarr Revolution: Compressed, Chunked Storage </span> </a> </li> <li class=md-nav__item> <a href=#the-zarrdask-collaboration-remote-storage-concurrency class=md-nav__link> <span class=md-ellipsis> The Zarr/Dask Collaboration: Remote Storage &amp; Concurrency </span> </a> </li> <li class=md-nav__item> <a href=#the-lazy-computation-revolution class=md-nav__link> <span class=md-ellipsis> The Lazy Computation Revolution </span> </a> </li> <li class=md-nav__item> <a href=#the-parquet-revolution-skip-what-you-dont-need class=md-nav__link> <span class=md-ellipsis> The Parquet Revolution: Skip What You Don't Need </span> </a> </li> <li class=md-nav__item> <a href=#the-lance-revolution-zero-infrastructure-embedded-databases class=md-nav__link> <span class=md-ellipsis> The Lance Revolution: Zero-Infrastructure Embedded Databases </span> </a> </li> <li class=md-nav__item> <a href=#the-olap-query-engine-revolution-fast-languages-lazy-execution class=md-nav__link> <span class=md-ellipsis> The OLAP Query Engine Revolution: Fast Languages + Lazy Execution </span> </a> </li> <li class=md-nav__item> <a href=#the-single-cell-gap class=md-nav__link> <span class=md-ellipsis> The Single-Cell Gap </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#the-evolution-of-single-cell-data-storage class=md-nav__link> <span class=md-ellipsis> The Evolution of Single-Cell Data Storage </span> </a> </li> <li class=md-nav__item> <a href=#the-scale-problem-why-current-tools-are-breaking class=md-nav__link> <span class=md-ellipsis> The Scale Problem: Why Current Tools Are Breaking </span> </a> </li> <li class=md-nav__item> <a href=#who-is-slaf-for class=md-nav__link> <span class=md-ellipsis> Who is SLAF For? </span> </a> </li> <li class=md-nav__item> <a href=#how-slaf-works class=md-nav__link> <span class=md-ellipsis> How SLAF Works </span> </a> </li> <li class=md-nav__item> <a href=#getting-started class=md-nav__link> <span class=md-ellipsis> Getting Started </span> </a> <nav class=md-nav aria-label="Getting Started"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#installation class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=#basic-usage class=md-nav__link> <span class=md-ellipsis> Basic Usage </span> </a> </li> <li class=md-nav__item> <a href=#migration-from-h5ad class=md-nav__link> <span class=md-ellipsis> Migration from H5AD </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#future-directions-and-community-feedback class=md-nav__link> <span class=md-ellipsis> Future Directions and Community Feedback </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../blazing-fast-dataloaders/ class=md-nav__link> <span class=md-ellipsis> Blazing Fast Dataloaders </span> </a> </li> <li class=md-nav__item> <a href=../blazing-fast-dataloaders-2/ class=md-nav__link> <span class=md-ellipsis> Mixture of Scanners </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#a-10-year-lookback-at-storage-and-compute-revolutions class=md-nav__link> <span class=md-ellipsis> A 10-year lookback at storage and compute revolutions </span> </a> <nav class=md-nav aria-label="A 10-year lookback at storage and compute revolutions"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#the-zarr-revolution-compressed-chunked-storage class=md-nav__link> <span class=md-ellipsis> The Zarr Revolution: Compressed, Chunked Storage </span> </a> </li> <li class=md-nav__item> <a href=#the-zarrdask-collaboration-remote-storage-concurrency class=md-nav__link> <span class=md-ellipsis> The Zarr/Dask Collaboration: Remote Storage &amp; Concurrency </span> </a> </li> <li class=md-nav__item> <a href=#the-lazy-computation-revolution class=md-nav__link> <span class=md-ellipsis> The Lazy Computation Revolution </span> </a> </li> <li class=md-nav__item> <a href=#the-parquet-revolution-skip-what-you-dont-need class=md-nav__link> <span class=md-ellipsis> The Parquet Revolution: Skip What You Don't Need </span> </a> </li> <li class=md-nav__item> <a href=#the-lance-revolution-zero-infrastructure-embedded-databases class=md-nav__link> <span class=md-ellipsis> The Lance Revolution: Zero-Infrastructure Embedded Databases </span> </a> </li> <li class=md-nav__item> <a href=#the-olap-query-engine-revolution-fast-languages-lazy-execution class=md-nav__link> <span class=md-ellipsis> The OLAP Query Engine Revolution: Fast Languages + Lazy Execution </span> </a> </li> <li class=md-nav__item> <a href=#the-single-cell-gap class=md-nav__link> <span class=md-ellipsis> The Single-Cell Gap </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#the-evolution-of-single-cell-data-storage class=md-nav__link> <span class=md-ellipsis> The Evolution of Single-Cell Data Storage </span> </a> </li> <li class=md-nav__item> <a href=#the-scale-problem-why-current-tools-are-breaking class=md-nav__link> <span class=md-ellipsis> The Scale Problem: Why Current Tools Are Breaking </span> </a> </li> <li class=md-nav__item> <a href=#who-is-slaf-for class=md-nav__link> <span class=md-ellipsis> Who is SLAF For? </span> </a> </li> <li class=md-nav__item> <a href=#how-slaf-works class=md-nav__link> <span class=md-ellipsis> How SLAF Works </span> </a> </li> <li class=md-nav__item> <a href=#getting-started class=md-nav__link> <span class=md-ellipsis> Getting Started </span> </a> <nav class=md-nav aria-label="Getting Started"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#installation class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=#basic-usage class=md-nav__link> <span class=md-ellipsis> Basic Usage </span> </a> </li> <li class=md-nav__item> <a href=#migration-from-h5ad class=md-nav__link> <span class=md-ellipsis> Migration from H5AD </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#future-directions-and-community-feedback class=md-nav__link> <span class=md-ellipsis> Future Directions and Community Feedback </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=introducing-slaf-the-single-cell-data-format-for-the-virtual-cell-era>Introducing SLAF: The Single-Cell Data Format for the Virtual Cell Era<a class=headerlink href=#introducing-slaf-the-single-cell-data-format-for-the-virtual-cell-era title="Permanent link">&para;</a></h1> <hr> <p>Single-cell datasets have grown from 50k to 100M cells in less than a decade. What used to fit comfortably in memory now requires out-of-core, distributed computing. What used to transfer in minutes now takes days. What used to be a simple analysis pipeline for a bioinformatician now requires a team of infrastructure engineers.</p> <p>The single-cell data explosion has created a fundamental mismatch between our tools and our needs. We're trying to solve 2025 problems with 2015 technology. But it doesn't have to be that way.</p> <p>This is the story of <strong>SLAF</strong> (Sparse Lazy Array Format) --- a cloud-native, SQL-powered format designed for the modern single-cell era.</p> <hr> <h2 id=a-10-year-lookback-at-storage-and-compute-revolutions>A 10-year lookback at storage and compute revolutions<a class=headerlink href=#a-10-year-lookback-at-storage-and-compute-revolutions title="Permanent link">&para;</a></h2> <p>Before diving into single-cell data, I thought it would be worth a historical detour through some key innovations over the past decade that directly inspired SLAF's architecture.</p> <div class="admonition tip"> <p class=admonition-title>The Big Picture</p> <p>We'll explore six key innovations that revolutionized data storage and compute:</p> <ol> <li><strong>Compressed, chunked storage</strong> with optimized I/O</li> <li><strong>Cloud-native chunked storage</strong> enabling concurrent I/O and computation</li> <li><strong>Lazy computation</strong> for complex workflows (Dask, Polars)</li> <li><strong>Skip what you don't need</strong> (Parquet's predicate/projection pushdown)</li> <li><strong>Zero-infrastructure embedded databases on disk</strong> (Lance, DuckDB)</li> <li><strong>High-performance query engines</strong> (Polars, DuckDB, Data Fusion)</li> </ol> </div> <h3 id=the-zarr-revolution-compressed-chunked-storage>The Zarr Revolution: Compressed, Chunked Storage<a class=headerlink href=#the-zarr-revolution-compressed-chunked-storage title="Permanent link">&para;</a></h3> <div class="admonition info"> <p class=admonition-title>The Problem</p> <p>In 2016, <a href=https://alimanfoo.github.io/ >Alistair Miles</a>, a genomics scientist at the Wellcome Trust Sanger Institute, was struggling with genetic variation data containing around <strong>20 billion elements</strong> - too large for memory, too small for distributed computing. Traditional compute would OOM-kill his research before it began.</p> </div> <p>Miles discovered HDF5 with the h5py Python library, which provided a great solution for storing multi-dimensional arrays. The arrays were divided into chunks, each compressed, enabling efficient storage and fast access. But there was a problem: HDF5 was still too slow for interactive work.</p> <p>He then discovered <a href=https://bcolz.readthedocs.io/ >Bcolz</a>, which used the Blosc compression library. Blosc could use multiple threads internally, worked well with CPU cache architecture, and was much faster than HDF5. In his <a href=https://alimanfoo.github.io/2016/04/14/to-hdf5-and-beyond.html>benchmarks</a>, Bcolz was more than 10 times faster at storing data than HDF5.</p> <p>But Bcolz had a limitation: it could only be chunked along the first dimension. Taking slices of other dimensions required reading and decompressing the entire array.</p> <p>So Miles created <a href=https://zarr.readthedocs.io/ >Zarr</a>, which like Bcolz used Blosc internally but supported chunking along multiple dimensions. This enabled better performance for multiple data access patterns.</p> <div class="admonition success"> <p class=admonition-title>Key Innovation #1</p> <p><strong>Compressed, chunked storage with optimized encoding/decoding for speed of I/O.</strong></p> </div> <p>From these early beginnings, Zarr has grown far beyond genomics. Today it powers massive array formats across climate science, geospatial data, medical imaging, and microscopy --- any domain dealing with large, multi-dimensional arrays that need efficient cloud-native access.</p> <h3 id=the-zarrdask-collaboration-remote-storage-concurrency>The Zarr/Dask Collaboration: Remote Storage &amp; Concurrency<a class=headerlink href=#the-zarrdask-collaboration-remote-storage-concurrency title="Permanent link">&para;</a></h3> <div class="admonition info"> <p class=admonition-title>The Problem</p> <p><a href=https://matthewrocklin.com/ >Matt Rocklin</a> (then, at the Anaconda foundation) was tackling a similar challenge: how to work with arrays too big for memory in numpy for embarassingly parallel computations.</p> </div> <p>Miles created <a href=https://zarr.readthedocs.io/ >Zarr</a> for compressed, chunked storage. Rocklin built <a href=https://dask.org/ >Dask</a> for lazy, out-of-core computation (<a href=https://blog.dask.org/2014/12/27/Towards-OOC>early ideas</a>, <a href=https://blog.dask.org/2015/05/19/State-of-Dask>consolidation</a>).</p> <p>But there were two separate technical challenges to solve for true cloud-native array access:</p> <p><strong>Challenge 1: Remote Storage Access</strong> Zarr could read chunks efficiently from local storage, but accessing remote storage (like S3) required a different approach. Traditional file I/O was designed for local filesystems, not HTTP-based object storage with different latency characteristics. Zarr solved this by implementing HTTP-based chunk access that could work directly with cloud object storage.</p> <p><strong>Challenge 2: The Python GIL Bottleneck</strong> The fundamental problem was the Python Global Interpreter Lock (GIL). As <a href=https://alimanfoo.github.io/2016/05/16/cpu-blues.html>Alistair Miles documented</a>, h5py doesn't release the GIL during I/O operations, which means other threads cannot run while h5py is doing anything - even if those threads want to do something unrelated to HDF5 I/O. This serialized all parallel operations, limiting CPU utilization to just over 1 core (~130% CPU).</p> <p>The breakthrough came when Zarr was designed to <strong>release the GIL during compression and decompression</strong>. This simple but crucial design decision meant other threads could carry on working in parallel. This GIL-aware design enabled true concurrent I/O and computation --- a fundamental requirement for cloud-native array access.</p> <p>When they <a href=https://alimanfoo.github.io/2016/05/16/cpu-blues.html>collaborated</a>, combining Zarr's remote chunked storage with Dask's process-based distributed computation, they finally achieved distributed computation on remote arrays for truly concurrent cloud-native array access.</p> <div class="admonition success"> <p class=admonition-title>Key Innovation #2</p> <p><strong>Cloud-native chunked storage enabling concurrent I/O and computation.</strong></p> </div> <h3 id=the-lazy-computation-revolution>The Lazy Computation Revolution<a class=headerlink href=#the-lazy-computation-revolution title="Permanent link">&para;</a></h3> <div class="admonition info"> <p class=admonition-title>The Problem</p> <p>Traditional eager computation loads data into memory immediately and processes it step by step. This works fine for small datasets, but becomes impossible when your data is larger than RAM. You can't even start the computation because the first step fails with "Out of Memory".</p> </div> <p>As <a href=https://blog.dask.org/2015/05/19/State-of-Dask>Matt Rocklin described</a>, Dask started as "a parallel on-disk array" that solved a fundamental problem: <strong>how do you compute on data that's too big for memory?</strong></p> <p>The key insight was <strong>lazy computation graphs</strong>. Instead of immediately loading and processing data, Dask builds a recipe (a directed acyclic graph) of all the operations you want to perform. This graph is just functions, dicts, and tuples - lightweight and easy to understand. Only when you call <code>.compute()</code> does Dask actually execute the operations, and it can optimize the entire graph before execution.</p> <p>For example, if you want to sum a trillion numbers, Dask breaks this into a million smaller sums that each fit in memory, then sums the sums. A previously impossible task becomes a million and one easy ones.</p> <p><a href=https://polars.rs/ >Polars</a> took this to the next level for dataframes, enabling complex query chains that only touch the data they need. The entire query is optimized before any data is loaded.</p> <div class="admonition success"> <p class=admonition-title>Key Innovation #3</p> <p><strong>Lazy computation graphs that optimize before execution.</strong></p> </div> <h3 id=the-parquet-revolution-skip-what-you-dont-need>The Parquet Revolution: Skip What You Don't Need<a class=headerlink href=#the-parquet-revolution-skip-what-you-dont-need title="Permanent link">&para;</a></h3> <div class="admonition info"> <p class=admonition-title>The Problem</p> <p>The Zarr/Dask combination was fantastic for embarrassingly parallel work across <em>entire</em> datasets. But what if you only wanted to analyze a subset of data that met certain criteria?</p> </div> <p>Meanwhile, <a href=https://parquet.apache.org/ >Parquet</a> didn't solve the partial compute problem but its design taught us how we might. Parquet was born at Twitter and Cloudera during the early days of the transition from OLTP to OLAP databases, and from row-based to <a href=https://en.wikipedia.org/wiki/Column-oriented_DBMS>columnar storage</a>. The insight was simple: analytical queries often access only a few columns, so storing data column-by-column enables much better compression and faster queries.</p> <p>Parquet stores data in compressed columnar chunks called "row groups" and stores metadata about these chunks. At query time, you only materialize the metadata in memory, then use it to decide:</p> <ul> <li>Which row groups to skip reading (predicate pushdown)</li> <li>Which columns to skip reading (projection pushdown)</li> </ul> <div class="admonition success"> <p class=admonition-title>Key Innovation #4</p> <p><strong>Don't just make I/O faster, skip most I/O you don't need to do.</strong></p> </div> <h3 id=the-lance-revolution-zero-infrastructure-embedded-databases>The Lance Revolution: Zero-Infrastructure Embedded Databases<a class=headerlink href=#the-lance-revolution-zero-infrastructure-embedded-databases title="Permanent link">&para;</a></h3> <div class="admonition info"> <p class=admonition-title>The Problem</p> <p>Parquet was revolutionary for analytical queries, but it had fundamental limitations that became apparent as data workloads evolved:</p> <ul> <li> <p><strong>Database features missing</strong>: Schema evolution, partitioning, time travel, and ACID transactions weren't first-class citizens. <a href=https://iceberg.apache.org/ >Apache Iceberg</a> emerged to address these limitations, but locked into Parquet as a format.</p> </li> <li> <p><strong>AI-era data types</strong>: Parquet preceded the AI era, so vector embeddings, multimodal data, and vector search weren't native. Modern AI workloads need to store and query embeddings alongside structured data.</p> </li> <li> <p><strong>Metadata overhead</strong>: Parquet's metadata can become expensive and suboptimal for random access patterns, especially as datasets grow.</p> </li> <li> <p><strong>Infrastructure complexity</strong>: Traditional databases require servers, configuration, and ongoing maintenance --- a barrier for many use cases.</p> </li> </ul> </div> <p><a href=https://lancedb.github.io/lance/ >Lance</a> directly addressed these limitations by creating a table format that provides:</p> <ul> <li><strong>Database features</strong>: Schema evolution, time travel, partitioning, and ACID transactions as first-class citizens (parity with Iceberg)</li> <li><strong>AI-native design</strong>: Native support for vector embeddings, multimodal data, and vector search</li> <li><strong>Optimized metadata</strong>: Efficient random access patterns with minimal metadata overhead</li> <li><strong>Zero infrastructure</strong>: A file on disk is a full database, like the next generation SQLite</li> </ul> <p><a href=https://pola.rs/ >Polars</a> takes a different but complementary approach: a high-performance DataFrame library written in Rust that provides blazing-fast SQL query execution with zero infrastructure. While Lance focuses on the table format, Polars focuses on both its own DataFrame operations and a highly extensible query engine. Both eliminate the need for servers and configuration.</p> <div class="admonition success"> <p class=admonition-title>Key Innovation #5</p> <p><strong>Zero-infrastructure embedded databases on disk.</strong></p> </div> <h3 id=the-olap-query-engine-revolution-fast-languages-lazy-execution>The OLAP Query Engine Revolution: Fast Languages + Lazy Execution<a class=headerlink href=#the-olap-query-engine-revolution-fast-languages-lazy-execution title="Permanent link">&para;</a></h3> <p>The final piece was query engines written in faster languages than Python. <a href=https://polars.rs/ >Polars</a> is written in Rust and executes chained query workloads lazily with blazing-fast SQL execution that can push down complex queries to the storage layer.</p> <p>By combining high-performance query engines with metadata-rich table formats like Parquet and Lance, you get superpowers: complex queries that only touch the data they need.</p> <div class="admonition success"> <p class=admonition-title>Key Innovation #6</p> <p><strong>High-performance query engines that can skip unnecessary computation.</strong></p> </div> <h3 id=the-single-cell-gap>The Single-Cell Gap<a class=headerlink href=#the-single-cell-gap title="Permanent link">&para;</a></h3> <p>The single-cell community evolved from simple <a href=https://math.nist.gov/MatrixMarket/formats.html>MTX format</a> (essentially TSV files) to <a href=https://anndata.readthedocs.io/en/stable/ >H5AD</a> (HDF5-based <a href=https://anndata.readthedocs.io/en/stable/ >AnnData</a>). Innovation happened on the metadata storage side with AnnData, and computational workflows with <a href=https://scanpy.readthedocs.io/ >Scanpy</a> leveraging scipy sparse matrices and numpy. But storage hasn't leveraged the amazing innovations in cloud-native, out-of-core, query-optimized, lazy, metadata-rich workflows that revolutionized other domains.</p> <p>A format like SLAF can bring these proven innovations to single-cell data.</p> <hr> <h2 id=the-evolution-of-single-cell-data-storage>The Evolution of Single-Cell Data Storage<a class=headerlink href=#the-evolution-of-single-cell-data-storage title="Permanent link">&para;</a></h2> <p>Single-cell data storage has evolved from simple <a href=https://math.nist.gov/MatrixMarket/formats.html>MTX format</a> (TSV files) to <a href=https://anndata.readthedocs.io/en/stable/ >H5AD</a> (HDF5-based <a href=https://anndata.readthedocs.io/en/stable/ >AnnData</a>) as the de facto standard. While H5AD brought compression and <a href=https://scanpy.readthedocs.io/ >Scanpy</a> integration, it has fundamental limitations: single-file format (no concurrent access), in-memory operations (everything must fit in RAM), and no cloud-native access.</p> <p>Recent attempts like <a href=https://github.com/single-cell-data/SOMA>SOMA</a> (TileDB/CZI collaboration) and <a href=https://github.com/chanzuckerberg/cellxgene-census>cellxgene-census</a> recognize these limitations but have different focuses: SOMA is broad-scope sparse arrays, while cellxgene-census is tied to specific datasets. The gap remains: we need cloud-native storage that works seamlessly with existing bioinformatics tools while expanding to AI use cases.</p> <hr> <h2 id=the-scale-problem-why-current-tools-are-breaking>The Scale Problem: Why Current Tools Are Breaking<a class=headerlink href=#the-scale-problem-why-current-tools-are-breaking title="Permanent link">&para;</a></h2> <p>Let's paint a picture of what happens when you try to analyze a 100M-cell dataset with current tools.</p> <div class=grid cols=2> <div> <p><span class=twemoji><svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M64 32C28.7 32 0 60.7 0 96v64c0 35.3 28.7 64 64 64h320c35.3 0 64-28.7 64-64V96c0-35.3-28.7-64-64-64zm216 72a24 24 0 1 1 0 48 24 24 0 1 1 0-48m56 24a24 24 0 1 1 48 0 24 24 0 1 1-48 0M64 288c-35.3 0-64 28.7-64 64v64c0 35.3 28.7 64 64 64h320c35.3 0 64-28.7 64-64v-64c0-35.3-28.7-64-64-64zm216 72a24 24 0 1 1 0 48 24 24 0 1 1 0-48m56 24a24 24 0 1 1 48 0 24 24 0 1 1-48 0"/></svg></span> <strong>Infrastructure Bottlenecks</strong></p> <p>A typical 100M-cell dataset might be 500GB-1TB. With H5AD: download takes hours to days, must fit entire dataset in memory (impossible on most machines), all operations happen in memory.</p> <p><strong>You can't analyze data larger than your RAM.</strong></p> </div> <div> <p><span class=twemoji><svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M320 16a104 104 0 1 1 0 208 104 104 0 1 1 0-208M96 88a72 72 0 1 1 0 144 72 72 0 1 1 0-144M0 416c0-70.7 57.3-128 128-128 12.8 0 25.2 1.9 36.9 5.4C132 330.2 112 378.8 112 432v16c0 11.4 2.4 22.2 6.7 32H32c-17.7 0-32-14.3-32-32zm521.3 64c4.3-9.8 6.7-20.6 6.7-32v-16c0-53.2-20-101.8-52.9-138.6 11.7-3.5 24.1-5.4 36.9-5.4 70.7 0 128 57.3 128 128v32c0 17.7-14.3 32-32 32zM472 160a72 72 0 1 1 144 0 72 72 0 1 1-144 0M160 432c0-88.4 71.6-160 160-160s160 71.6 160 160v16c0 17.7-14.3 32-32 32H192c-17.7 0-32-14.3-32-32z"/></svg></span> <strong>Human Bottlenecks</strong></p> <p>Bioinformatician wants to analyze large dataset → requests infrastructure changes → waits weeks → downloads massive dataset → runs into memory issues → requests more RAM → waits again.</p> <p><strong>Every scale-up requires infrastructure engineers.</strong></p> </div> <div> <p><span class=twemoji><svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M192 0c-35.3 0-64 28.7-64 64v256c0 35.3 28.7 64 64 64h192c35.3 0 64-28.7 64-64V119.4c0-17.4-7.1-34.1-19.7-46.2l-57.7-55.4A64.1 64.1 0 0 0 326.3 0zM64 128c-35.3 0-64 28.7-64 64v256c0 35.3 28.7 64 64 64h192c35.3 0 64-28.7 64-64v-16h-64v16H64V192h16v-64z"/></svg></span> <strong>Data Duplication</strong></p> <p>Team of 5 bioinformaticians × 500GB dataset × 5 experiments = 12.5TB for what should be 500GB. Each researcher needs their own copy, each experiment needs a copy.</p> <p><strong>This data multiplication problem is unsustainable at scale.</strong></p> </div> <div> <p><span class=twemoji><svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M120 56c0-30.9 25.1-56 56-56h24c17.7 0 32 14.3 32 32v448c0 17.7-14.3 32-32 32h-32c-29.8 0-54.9-20.4-62-48h-2c-44.2 0-80-35.8-80-80 0-18 6-34.6 16-48-19.4-14.6-32-37.8-32-64 0-30.9 17.6-57.8 43.2-71.1-7.1-12-11.2-26-11.2-40.9 0-44.2 35.8-80 80-80zm272 0v24c44.2 0 80 35.8 80 80 0 15-4.1 29-11.2 40.9C486.5 214.2 504 241 504 272c0 26.2-12.6 49.4-32 64 10 13.4 16 30 16 48 0 44.2-35.8 80-80 80h-2c-7.1 27.6-32.2 48-62 48h-32c-17.7 0-32-14.3-32-32V32c0-17.7 14.3-32 32-32h24c30.9 0 56 25.1 56 56"/></svg></span> <strong>AI Workload Mismatch</strong></p> <p>Traditional workflows: filtering, normalization, PCA, clustering. New AI-native workflows: nearest neighbor search on embeddings, gene-gene relationship ranking, transformer training, distributed training.</p> <p><strong>Current tools weren't designed for these workloads.</strong></p> </div> </div> <hr> <h2 id=who-is-slaf-for>Who is SLAF For?<a class=headerlink href=#who-is-slaf-for title="Permanent link">&para;</a></h2> <p>SLAF is designed for the modern single-cell ecosystem facing scale challenges:</p> <table> <thead> <tr> <th>Persona</th> <th>Problem</th> <th>Current Reality</th> <th>How SLAF Helps</th> </tr> </thead> <tbody> <tr> <td><strong>Bioinformaticians</strong></td> <td>OOM errors on 10M+ cell datasets, can't do self-service analysis</td> <td>Need infrastructure engineers, wait weeks for provisioning</td> <td>Lazy evaluation eliminates OOM errors, enables 100M-cell analysis on laptop</td> </tr> <tr> <td><strong>Foundation Model Builders</strong></td> <td>Spending more time on data engineering than model training, data transfer bottlenecks</td> <td>Copy 500GB datasets to each node, experiments bottlenecked by throughput</td> <td>Cloud-native streaming eliminates duplication, SQL queries enable efficient tokenization</td> </tr> <tr> <td><strong>Tech Leaders &amp; Architects</strong></td> <td>Storage costs exploding, data multiplication problem</td> <td>5 bioinformaticians × 500GB × 5 experiments = 12.5TB for 500GB data</td> <td>Zero-copy, query-in-place storage means one copy serves everyone</td> </tr> <tr> <td><strong>Tool Builders</strong> (cellxgene, LatchBio etc.)</td> <td>Can't provide interactive experiences on massive datasets</td> <td>Users wait minutes, need expensive infrastructure</td> <td>Concurrent, cloud-scale access with high QPS, commodity web services</td> </tr> <tr> <td><strong>Atlas Builders</strong> (CZI, etc.)</td> <td>Can't serve massive datasets to research community</td> <td>Datasets too large to download, expensive serving infrastructure</td> <td>Cloud-native, zero-copy storage enables global access without downloads</td> </tr> <tr> <td><strong>Data Integrators</strong> (Elucidata, etc.)</td> <td>Complex data integration workflows don't scale</td> <td>Multiple data transfers, expensive compute for joins/aggregations</td> <td>SQL-native design enables complex integration with pushdown optimization</td> </tr> </tbody> </table> <h2 id=how-slaf-works>How SLAF Works<a class=headerlink href=#how-slaf-works title="Permanent link">&para;</a></h2> <p>SLAF combines the best ideas from multiple domains into a unified solution:</p> <ul> <li><strong>Zarr-inspired</strong>: Zero-copy, query-in-place access to cloud storage</li> <li><strong>Dask-inspired</strong>: Lazy computation graphs that optimize before execution</li> <li><strong>Lance + Polars-inspired</strong>: OLAP-powered SQL with pushdown optimization</li> <li><strong>Scanpy-compatible</strong>: Drop-in replacement for existing workflows</li> </ul> <p>For a detailed technical deep dive into SLAF's architecture, including SQL-native relational schema, lazy evaluation, cloud-native concurrent access, and foundation model training support, see <a href=../../user-guide/how-slaf-works/ >How SLAF Works</a>.</p> <hr> <h2 id=getting-started>Getting Started<a class=headerlink href=#getting-started title="Permanent link">&para;</a></h2> <p>SLAF is designed to be easy to adopt. Here's how to get started:</p> <h3 id=installation>Installation<a class=headerlink href=#installation title="Permanent link">&para;</a></h3> <div class="language-bash highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1># Using uv (recommended)</span>
</span><span id=__span-0-2><a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>uv<span class=w> </span>add<span class=w> </span>slafdb
</span><span id=__span-0-3><a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a>
</span><span id=__span-0-4><a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=c1># Or pip</span>
</span><span id=__span-0-5><a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a>pip<span class=w> </span>install<span class=w> </span>slafdb
</span></code></pre></div> <h3 id=basic-usage>Basic Usage<a class=headerlink href=#basic-usage title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf</span><span class=w> </span><span class=kn>import</span> <span class=n>SLAFArray</span>
</span><span id=__span-1-2><a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>
</span><span id=__span-1-3><a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a><span class=c1># Load a SLAF dataset</span>
</span><span id=__span-1-4><a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a><span class=n>slaf_array</span> <span class=o>=</span> <span class=n>SLAFArray</span><span class=p>(</span><span class=s2>&quot;path/to/dataset.slaf&quot;</span><span class=p>)</span>
</span><span id=__span-1-5><a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a>
</span><span id=__span-1-6><a id=__codelineno-1-6 name=__codelineno-1-6 href=#__codelineno-1-6></a><span class=c1># Query with SQL</span>
</span><span id=__span-1-7><a id=__codelineno-1-7 name=__codelineno-1-7 href=#__codelineno-1-7></a><span class=n>results</span> <span class=o>=</span> <span class=n>slaf_array</span><span class=o>.</span><span class=n>query</span><span class=p>(</span><span class=s2>&quot;&quot;&quot;</span>
</span><span id=__span-1-8><a id=__codelineno-1-8 name=__codelineno-1-8 href=#__codelineno-1-8></a><span class=s2>    SELECT cell_type, COUNT(*) as count</span>
</span><span id=__span-1-9><a id=__codelineno-1-9 name=__codelineno-1-9 href=#__codelineno-1-9></a><span class=s2>    FROM cells</span>
</span><span id=__span-1-10><a id=__codelineno-1-10 name=__codelineno-1-10 href=#__codelineno-1-10></a><span class=s2>    GROUP BY cell_type</span>
</span><span id=__span-1-11><a id=__codelineno-1-11 name=__codelineno-1-11 href=#__codelineno-1-11></a><span class=s2>&quot;&quot;&quot;</span><span class=p>)</span>
</span><span id=__span-1-12><a id=__codelineno-1-12 name=__codelineno-1-12 href=#__codelineno-1-12></a>
</span><span id=__span-1-13><a id=__codelineno-1-13 name=__codelineno-1-13 href=#__codelineno-1-13></a><span class=c1># Use with Scanpy</span>
</span><span id=__span-1-14><a id=__codelineno-1-14 name=__codelineno-1-14 href=#__codelineno-1-14></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf.integrations</span><span class=w> </span><span class=kn>import</span> <span class=n>read_slaf</span>
</span><span id=__span-1-15><a id=__codelineno-1-15 name=__codelineno-1-15 href=#__codelineno-1-15></a><span class=n>adata</span> <span class=o>=</span> <span class=n>read_slaf</span><span class=p>(</span><span class=s2>&quot;path/to/dataset.slaf&quot;</span><span class=p>)</span>
</span></code></pre></div> <h3 id=migration-from-h5ad>Migration from H5AD<a class=headerlink href=#migration-from-h5ad title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=c1># Convert existing H5AD to SLAF</span>
</span><span id=__span-2-2><a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=kn>from</span><span class=w> </span><span class=nn>slaf.data</span><span class=w> </span><span class=kn>import</span> <span class=n>SLAFConverter</span>
</span><span id=__span-2-3><a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a>
</span><span id=__span-2-4><a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a><span class=c1># Convert h5ad file to SLAF format</span>
</span><span id=__span-2-5><a id=__codelineno-2-5 name=__codelineno-2-5 href=#__codelineno-2-5></a><span class=n>converter</span> <span class=o>=</span> <span class=n>SLAFConverter</span><span class=p>()</span>
</span><span id=__span-2-6><a id=__codelineno-2-6 name=__codelineno-2-6 href=#__codelineno-2-6></a><span class=n>converter</span><span class=o>.</span><span class=n>convert</span><span class=p>(</span><span class=s2>&quot;input.h5ad&quot;</span><span class=p>,</span> <span class=s2>&quot;output.slaf&quot;</span><span class=p>)</span>
</span><span id=__span-2-7><a id=__codelineno-2-7 name=__codelineno-2-7 href=#__codelineno-2-7></a>
</span><span id=__span-2-8><a id=__codelineno-2-8 name=__codelineno-2-8 href=#__codelineno-2-8></a><span class=c1># Or convert from AnnData object</span>
</span><span id=__span-2-9><a id=__codelineno-2-9 name=__codelineno-2-9 href=#__codelineno-2-9></a><span class=kn>import</span><span class=w> </span><span class=nn>scanpy</span><span class=w> </span><span class=k>as</span><span class=w> </span><span class=nn>sc</span>
</span><span id=__span-2-10><a id=__codelineno-2-10 name=__codelineno-2-10 href=#__codelineno-2-10></a><span class=n>adata</span> <span class=o>=</span> <span class=n>sc</span><span class=o>.</span><span class=n>read_h5ad</span><span class=p>(</span><span class=s2>&quot;input.h5ad&quot;</span><span class=p>)</span>
</span><span id=__span-2-11><a id=__codelineno-2-11 name=__codelineno-2-11 href=#__codelineno-2-11></a><span class=n>converter</span><span class=o>.</span><span class=n>convert_anndata</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span> <span class=s2>&quot;output.slaf&quot;</span><span class=p>)</span>
</span></code></pre></div> <hr> <h2 id=future-directions-and-community-feedback>Future Directions and Community Feedback<a class=headerlink href=#future-directions-and-community-feedback title="Permanent link">&para;</a></h2> <p>SLAF could go in several directions and I'd like to get feedback on which of them would benefit the community the most.</p> <p><strong>Visualization Integration</strong>: We lack direct visualization support. A sketch of how SLAF could be more directly compatible with cellxgene might be super useful for Atlas builders to transition to SaaS-based analytics. This could enable interactive exploration of 100M-cell datasets without the infrastructure headaches that currently plague large-scale visualization. What visualization workflows would you prioritize?</p> <p><strong>Embeddings Support</strong>: Native storage and querying of cell/gene embeddings could unlock entirely new AI-native workflows. Instead of computing embeddings separately and storing them in different systems, SLAF could store embeddings alongside expression data and enable queries like "find cells similar to this embedding" or "rank genes by embedding similarity." This would be particularly valuable for foundation model training and retrieval-augmented generation. What embedding use cases are most critical for your work?</p> <p><strong>Hosted Solution</strong>: A cloud-hosted database version of SLAF could push query performance even further by eliminating the need for local storage entirely. This would enable true zero-infrastructure analysis where researchers query massive datasets directly from the cloud without any local setup. The challenge is balancing performance with cost. What query patterns and performance requirements would make this viable for your workflows?</p> <p><strong>Migration Tools</strong>: Better tools to migrate from mtx/h5ad to SLAF could accelerate adoption significantly. This includes not just format conversion but intelligent optimization of hydrating the SLAF format for different dataset characteristics. What migration pain points are most blocking your adoption of new formats?</p> <p><strong>Distributed Computing</strong>: Better support for distributed analysis workflows could enable truly massive-scale analysis beyond what any single machine can handle. This includes both distributed query execution and distributed training support for foundation models. What distributed computing patterns are most important for your large-scale analysis needs?</p> <p><strong>Scanpy Feature Parity</strong>: We currently only support a couple of limited use cases from scanpy's preprocessing module. Building lazy operations for broader needs like PCA, UMAP, or differential expression computations at scale could make SLAF a true drop-in replacement for existing scanpy workflows. This would enable researchers to run familiar analysis pipelines on datasets that would otherwise cause memory explosions. Which scanpy operations are most critical for your large-scale analysis workflows?</p> <p><strong>Foundation Model Training Lifecycles</strong>: Our current dataloader is basic - no smart shuffling, shard-aware streaming, async pre-fetching, or real-world benchmarks on multi-node training yet. SLAF could help frame and drive moonshots like "the $100 overnight scGPT training experiment" by driving up the efficiencies of multi-node, multi-GPU training and fine-tuning workloads. What training bottlenecks are most limiting your foundation model experiments?</p> <p><strong>Cloud-Native Benchmarks</strong>: Limited real-world benchmarks for cloud-native datasets. Today's benchmarks are all on local storage for small datasets. Would this be a blocker to adoption? What performance characteristics and benchmarks would give you confidence to adopt SLAF for production workloads?</p> <hr> <p>If you've read this far, thank you! You're definitely invested in the space.</p> <p>The single-cell data explosion isn't going to slow down. If anything, it's accelerating. We need tools that can keep up. SLAF is an early attempt to build the modern companion for bioinformaticians turned AI engineers.</p> <p>I'm excited to see how you adopt and extend SLAF. Together, we can build the infrastructure needed for the next decade of single-cell research.</p> <hr> <p><em>Want to learn more? Check out the <a href=https://slaf-project.github.io/slaf/ >SLAF documentation</a> or join the conversation on <a href=https://github.com/slaf-project/slaf>GitHub</a>.</em></p> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="August 14, 2025 18:23:50 UTC">August 14, 2025</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="July 13, 2025 19:37:44 UTC">July 13, 2025</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Contributors> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 4a4 4 0 0 1 4 4 4 4 0 0 1-4 4 4 4 0 0 1-4-4 4 4 0 0 1 4-4m0 10c4.42 0 8 1.79 8 4v2H4v-2c0-2.21 3.58-4 8-4"/></svg> </span> <nav> Pavan Ramkumar </nav> </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2026 Pavan Ramkumar </div> </div> <div class=md-social> <a href=https://github.com/slaf-project/slaf target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.sections", "content.code.copy"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../../javascripts/mathjax.js></script> <script src=../../javascripts/theme-reset.js></script> <script src=../../javascripts/copy-enhancement.js></script> <script src=../../javascripts/theme-aware-assets.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>