<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Sparse Lazy Array Format - High-performance single-cell data storage and analysis"><meta name=author content="Pavan Ramkumar"><link href=https://slaf-project.github.io/slaf/benchmarks/slaf_vs_tiledb_benchmarks/ rel=canonical><link rel=icon href=../../assets/slaf-icon-transparent-dark-mono.svg><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>SLAF vs TileDB Performance Benchmarks - SLAF Documentation</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../assets/_mkdocstrings.css><link rel=stylesheet href=../../stylesheets/extra.css><link rel=stylesheet href=../../stylesheets/logo-theme.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config",""),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#slaf-vs-tiledb-performance-benchmarks class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="SLAF Documentation" class="md-header__button md-logo" aria-label="SLAF Documentation" data-md-component=logo> <img src=../../assets/slaf-icon-transparent-dark-mono.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> SLAF Documentation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> SLAF vs TileDB Performance Benchmarks </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/slaf-project/slaf title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> slaf-project/slaf </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="SLAF Documentation" class="md-nav__button md-logo" aria-label="SLAF Documentation" data-md-component=logo> <img src=../../assets/slaf-icon-transparent-dark-mono.svg alt=logo> </a> SLAF Documentation </label> <div class=md-nav__source> <a href=https://github.com/slaf-project/slaf title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> slaf-project/slaf </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class=md-nav__item> <a href=../../getting-started/quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex> <span class=md-ellipsis> User Guide </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> User Guide </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../user-guide/how-slaf-works/ class=md-nav__link> <span class=md-ellipsis> How SLAF Works </span> </a> </li> <li class=md-nav__item> <a href=../../user-guide/migrating-to-slaf/ class=md-nav__link> <span class=md-ellipsis> Migrating to SLAF </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex> <span class=md-ellipsis> Examples </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Examples </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../examples/getting-started/ class=md-nav__link> <span class=md-ellipsis> Getting Started </span> </a> </li> <li class=md-nav__item> <a href=../../examples/lazy-processing/ class=md-nav__link> <span class=md-ellipsis> Lazy Processing </span> </a> </li> <li class=md-nav__item> <a href=../../examples/ml-training/ class=md-nav__link> <span class=md-ellipsis> ML Training </span> </a> </li> <li class=md-nav__item> <a href=../../examples/sql-queries/ class=md-nav__link> <span class=md-ellipsis> SQL Queries </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex> <span class=md-ellipsis> Benchmarks </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=false> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Benchmarks </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class=md-nav__item> <a href=../bioinformatics_benchmarks/ class=md-nav__link> <span class=md-ellipsis> For Bioinformaticians </span> </a> </li> <li class=md-nav__item> <a href=../ml_benchmarks/ class=md-nav__link> <span class=md-ellipsis> For ML Engineers </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/core/ class=md-nav__link> <span class=md-ellipsis> Core </span> </a> </li> <li class=md-nav__item> <a href=../../api/data/ class=md-nav__link> <span class=md-ellipsis> Data </span> </a> </li> <li class=md-nav__item> <a href=../../api/integrations/ class=md-nav__link> <span class=md-ellipsis> Integrations </span> </a> </li> <li class=md-nav__item> <a href=../../api/ml/ class=md-nav__link> <span class=md-ellipsis> ML </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex> <span class=md-ellipsis> Development </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Development </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../development/contributing/ class=md-nav__link> <span class=md-ellipsis> For Contributors </span> </a> </li> <li class=md-nav__item> <a href=../../development/maintaining/ class=md-nav__link> <span class=md-ellipsis> For Maintainers </span> </a> </li> <li class=md-nav__item> <a href=../../development/benchmarks/ class=md-nav__link> <span class=md-ellipsis> Benchmarks </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex> <span class=md-ellipsis> Blog </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> Blog </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../blog/ class=md-nav__link> <span class=md-ellipsis> Blog Home </span> </a> </li> <li class=md-nav__item> <a href=../../blog/huggingface/ class=md-nav__link> <span class=md-ellipsis> SLAF on Hugging Face </span> </a> </li> <li class=md-nav__item> <a href=../../blog/introducing-slaf/ class=md-nav__link> <span class=md-ellipsis> Introducing SLAF </span> </a> </li> <li class=md-nav__item> <a href=../../blog/blazing-fast-dataloaders/ class=md-nav__link> <span class=md-ellipsis> Blazing Fast Dataloaders </span> </a> </li> <li class=md-nav__item> <a href=../../blog/blazing-fast-dataloaders-2/ class=md-nav__link> <span class=md-ellipsis> Mixture of Scanners </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> <nav class=md-nav aria-label=Overview> <ul class=md-nav__list> <li class=md-nav__item> <a href=#key-performance-summary class=md-nav__link> <span class=md-ellipsis> Key Performance Summary </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#conversion-performance class=md-nav__link> <span class=md-ellipsis> Conversion Performance </span> </a> <nav class=md-nav aria-label="Conversion Performance"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#tiledb-to-slaf-conversion class=md-nav__link> <span class=md-ellipsis> TileDB to SLAF Conversion </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#bioinformatics-benchmarks class=md-nav__link> <span class=md-ellipsis> Bioinformatics Benchmarks </span> </a> <nav class=md-nav aria-label="Bioinformatics Benchmarks"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#cell-filtering-performance class=md-nav__link> <span class=md-ellipsis> Cell Filtering Performance </span> </a> </li> <li class=md-nav__item> <a href=#gene-filtering-performance class=md-nav__link> <span class=md-ellipsis> Gene Filtering Performance </span> </a> </li> <li class=md-nav__item> <a href=#expression-queries-performance class=md-nav__link> <span class=md-ellipsis> Expression Queries Performance </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#machine-learning-benchmarks class=md-nav__link> <span class=md-ellipsis> Machine Learning Benchmarks </span> </a> <nav class=md-nav aria-label="Machine Learning Benchmarks"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#raw-data-loading-performance class=md-nav__link> <span class=md-ellipsis> Raw Data Loading Performance </span> </a> </li> <li class=md-nav__item> <a href=#gpu-ready-output-performance class=md-nav__link> <span class=md-ellipsis> GPU-Ready Output Performance </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#technical-implementation-comparison class=md-nav__link> <span class=md-ellipsis> Technical Implementation Comparison </span> </a> <nav class=md-nav aria-label="Technical Implementation Comparison"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#key-technical-advantages class=md-nav__link> <span class=md-ellipsis> Key Technical Advantages </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#migration-benefits class=md-nav__link> <span class=md-ellipsis> Migration Benefits </span> </a> <nav class=md-nav aria-label="Migration Benefits"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#performance-improvements class=md-nav__link> <span class=md-ellipsis> Performance Improvements </span> </a> </li> <li class=md-nav__item> <a href=#developer-experience class=md-nav__link> <span class=md-ellipsis> Developer Experience </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#use-case-recommendations class=md-nav__link> <span class=md-ellipsis> Use Case Recommendations </span> </a> <nav class=md-nav aria-label="Use Case Recommendations"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#choose-slaf-for class=md-nav__link> <span class=md-ellipsis> Choose SLAF for: </span> </a> </li> <li class=md-nav__item> <a href=#choose-tiledb-for class=md-nav__link> <span class=md-ellipsis> Choose TileDB for: </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#conclusion class=md-nav__link> <span class=md-ellipsis> Conclusion </span> </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=slaf-vs-tiledb-performance-benchmarks>SLAF vs TileDB Performance Benchmarks<a class=headerlink href=#slaf-vs-tiledb-performance-benchmarks title="Permanent link">&para;</a></h1> <p>This document provides a comprehensive performance comparison between SLAF and TileDB SOMA across bioinformatics and machine learning workflows, focusing on modern columnar storage formats.</p> <h2 id=overview><strong>Overview</strong><a class=headerlink href=#overview title="Permanent link">&para;</a></h2> <p>Both SLAF and TileDB represent modern approaches to single-cell data storage, using Arrow-interoperable columnar formats. However, SLAF demonstrates <strong>consistent performance advantages</strong> across metadata filtering and dataloader throughput, highlighting the benefits of its optimized streaming architecture and efficient data access patterns.</p> <h3 id=key-performance-summary><strong>Key Performance Summary</strong><a class=headerlink href=#key-performance-summary title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>Category</th> <th>SLAF vs TileDB Speedup</th> <th>Dataset</th> </tr> </thead> <tbody> <tr> <td><strong>Conversion Performance</strong></td> <td><strong>9.9x faster</strong></td> <td>synthetic_50k_processed</td> </tr> <tr> <td><strong>Cell Filtering</strong></td> <td><strong>9.4x faster</strong></td> <td>synthetic_50k_processed</td> </tr> <tr> <td><strong>Gene Filtering</strong></td> <td><strong>8.6x faster</strong></td> <td>synthetic_50k_processed</td> </tr> <tr> <td><strong>Cell-centric Queries</strong></td> <td><strong>1.1x-1.4x faster</strong></td> <td>synthetic_50k_processed</td> </tr> <tr> <td><strong>Gene-centric Queries</strong></td> <td><strong>2.5x-3.3x slower</strong></td> <td>synthetic_50k_processed</td> </tr> <tr> <td><strong>Submatrix Queries</strong></td> <td><strong>2.5x-5.0x slower</strong></td> <td>synthetic_50k_processed</td> </tr> <tr> <td><strong>ML Data Loading</strong></td> <td><strong>45.7x faster</strong></td> <td>Tahoe-100M</td> </tr> </tbody> </table> <div class="admonition success"> <p class=admonition-title>Modern Format Advantage</p> <p>Both SLAF and TileDB significantly outperform traditional h5ad-based approaches, demonstrating the benefits of modern cloud-native storage. SLAF provides additional performance optimizations for streaming and data access.</p> </div> <h2 id=conversion-performance><strong>Conversion Performance</strong><a class=headerlink href=#conversion-performance title="Permanent link">&para;</a></h2> <p>Conversion from h5ad to modern formats demonstrates SLAF's efficiency in data ingestion workflows.</p> <p><strong>Input Dataset</strong>: synthetic_50k_processed (49,955 cells × 25,000 genes, ~722MB h5ad file)</p> <table> <thead> <tr> <th>Metric</th> <th>SLAF</th> <th>TileDB SOMA</th> <th>SLAF vs TileDB Improvement</th> </tr> </thead> <tbody> <tr> <td><strong>Conversion Time</strong></td> <td><strong>1.87s</strong></td> <td>18.44s</td> <td><strong>9.9x faster</strong></td> </tr> <tr> <td><strong>Output Size</strong></td> <td><strong>349.2MB</strong></td> <td>561.1MB</td> <td><strong>38% smaller</strong></td> </tr> <tr> <td><strong>Peak Memory</strong></td> <td><strong>826.6MB</strong></td> <td>4,399.6MB</td> <td><strong>5.3x more memory efficient</strong></td> </tr> </tbody> </table> <p><strong>Key Advantages:</strong></p> <ul> <li><strong>9.9x faster conversion</strong> from h5ad to SLAF format</li> <li><strong>38% smaller output size</strong> compared to TileDB SOMA</li> <li><strong>5.3x more memory efficient</strong> - SLAF uses only 827MB vs TileDB's 4.4GB peak memory</li> <li><strong>Optimized chunked processing</strong> with efficient memory management</li> </ul> <h3 id=tiledb-to-slaf-conversion><strong>TileDB to SLAF Conversion</strong><a class=headerlink href=#tiledb-to-slaf-conversion title="Permanent link">&para;</a></h3> <p>Converting from TileDB SOMA to SLAF format is straightforward and efficient:</p> <p><strong>TileDB to SLAF Performance</strong>: synthetic_50k_processed (49,955 cells × 25,000 genes)</p> <p><strong>Migration Benefits:</strong></p> <ul> <li><strong>Fast conversion</strong>: 50k cell dataset converts in just 2 seconds</li> <li><strong>Linear scaling</strong>: Performance scales linearly with the number of cells</li> <li><strong>Simple command</strong>: <code>slaf convert data.tiledb output.slaf</code></li> </ul> <div class="admonition success"> <p class=admonition-title>Easy Migration from TileDB</p> <p>Converting from TileDB to SLAF is simple and fast. A 50k cell dataset takes only 2 seconds to convert, with linear scaling performance. The conversion preserves all data types and metadata while providing significant performance improvements for downstream analysis.</p> </div> <div class="admonition success"> <p class=admonition-title>Fast Migration Path</p> <p>SLAF's superior conversion performance enables rapid migration of existing h5ad datasets, with conversion times under 2 seconds for 50k cell datasets and significantly smaller output files.</p> </div> <h2 id=bioinformatics-benchmarks><strong>Bioinformatics Benchmarks</strong><a class=headerlink href=#bioinformatics-benchmarks title="Permanent link">&para;</a></h2> <h3 id=cell-filtering-performance><strong>Cell Filtering Performance</strong><a class=headerlink href=#cell-filtering-performance title="Permanent link">&para;</a></h3> <p>Cell filtering operations demonstrate the efficiency of modern columnar storage for metadata operations.</p> <table> <thead> <tr> <th>Scenario</th> <th>TileDB Total (ms)</th> <th>SLAF Total (ms)</th> <th>Speedup</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td>S1</td> <td>20.9</td> <td>2.9</td> <td><strong>7.3x</strong></td> <td>Cells with &gt;=500 genes</td> </tr> <tr> <td>S2</td> <td>21.3</td> <td>2.0</td> <td><strong>10.5x</strong></td> <td>High UMI count (total_counts &gt; 2000)</td> </tr> <tr> <td>S3</td> <td>21.3</td> <td>1.9</td> <td><strong>11.5x</strong></td> <td>Mitochondrial fraction &lt; 0.1</td> </tr> <tr> <td>S4</td> <td>18.6</td> <td>2.0</td> <td><strong>9.1x</strong></td> <td>Complex multi-condition filter</td> </tr> <tr> <td>S5</td> <td>18.1</td> <td>2.8</td> <td><strong>6.5x</strong></td> <td>Cell type annotation filter</td> </tr> <tr> <td>S6</td> <td>20.9</td> <td>2.0</td> <td><strong>10.5x</strong></td> <td>Cells from batch_1</td> </tr> <tr> <td>S7</td> <td>23.8</td> <td>2.3</td> <td><strong>10.3x</strong></td> <td>Cells in clusters 0,1 from batch_1</td> </tr> <tr> <td>S8</td> <td>23.0</td> <td>2.1</td> <td><strong>10.7x</strong></td> <td>High-quality cells (&gt;=1000 genes, &lt;=10% mt)</td> </tr> <tr> <td>S9</td> <td>19.3</td> <td>2.5</td> <td><strong>7.9x</strong></td> <td>Cells with 800-2000 total counts</td> </tr> <tr> <td>S10</td> <td>20.8</td> <td>2.1</td> <td><strong>10.1x</strong></td> <td>Cells with 200-1500 genes</td> </tr> </tbody> </table> <p><strong>Average Performance: 9.4x faster</strong></p> <h3 id=gene-filtering-performance><strong>Gene Filtering Performance</strong><a class=headerlink href=#gene-filtering-performance title="Permanent link">&para;</a></h3> <p>Gene filtering operations show consistent performance advantages for SLAF's optimized Polars operations.</p> <table> <thead> <tr> <th>Scenario</th> <th>TileDB Total (ms)</th> <th>SLAF Total (ms)</th> <th>Speedup</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td>S1</td> <td>22.3</td> <td>3.0</td> <td><strong>7.5x</strong></td> <td>Genes expressed in &gt;=10 cells</td> </tr> <tr> <td>S2</td> <td>19.5</td> <td>1.7</td> <td><strong>11.7x</strong></td> <td>Genes with &gt;=100 total counts</td> </tr> <tr> <td>S3</td> <td>9.3</td> <td>1.8</td> <td><strong>5.0x</strong></td> <td>Genes with mean expression &gt;=0.1</td> </tr> <tr> <td>S4</td> <td>15.8</td> <td>1.6</td> <td><strong>10.1x</strong></td> <td>Exclude mitochondrial genes</td> </tr> <tr> <td>S5</td> <td>16.9</td> <td>1.7</td> <td><strong>10.2x</strong></td> <td>Highly variable genes</td> </tr> <tr> <td>S6</td> <td>15.8</td> <td>2.1</td> <td><strong>7.7x</strong></td> <td>Non-highly variable genes</td> </tr> <tr> <td>S7</td> <td>18.2</td> <td>2.0</td> <td><strong>9.1x</strong></td> <td>Genes in &gt;=50 cells with &gt;=500 total counts</td> </tr> <tr> <td>S8</td> <td>19.8</td> <td>1.9</td> <td><strong>10.6x</strong></td> <td>Genes with 100-10000 total counts</td> </tr> <tr> <td>S9</td> <td>11.3</td> <td>2.0</td> <td><strong>5.6x</strong></td> <td>Genes in 5-1000 cells</td> </tr> </tbody> </table> <p><strong>Average Performance: 8.6x faster</strong></p> <h3 id=expression-queries-performance><strong>Expression Queries Performance</strong><a class=headerlink href=#expression-queries-performance title="Permanent link">&para;</a></h3> <p>Expression queries demonstrate the efficiency of optimized sparse matrix operations, where TileDB often wins.</p> <table> <thead> <tr> <th>Scenario</th> <th>TileDB Total (ms)</th> <th>SLAF Total (ms)</th> <th>Speedup</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td>S1</td> <td>63.8</td> <td>16.1</td> <td><strong>4.0x</strong></td> <td>Single cell expression</td> </tr> <tr> <td>S2</td> <td>19.3</td> <td>13.9</td> <td><strong>1.4x</strong></td> <td>Another single cell</td> </tr> <tr> <td>S3</td> <td>18.1</td> <td>14.2</td> <td><strong>1.3x</strong></td> <td>Two cells</td> </tr> <tr> <td>S4</td> <td>19.0</td> <td>15.5</td> <td><strong>1.2x</strong></td> <td>Three cells</td> </tr> <tr> <td>S5</td> <td>150.8</td> <td>523.7</td> <td><strong>0.3x</strong></td> <td>Single gene across all cells</td> </tr> <tr> <td>S6</td> <td>84.2</td> <td>442.6</td> <td><strong>0.2x</strong></td> <td>Another single gene</td> </tr> <tr> <td>S7</td> <td>97.7</td> <td>303.0</td> <td><strong>0.3x</strong></td> <td>Two genes</td> </tr> <tr> <td>S8</td> <td>83.3</td> <td>655.9</td> <td><strong>0.1x</strong></td> <td>Three genes</td> </tr> <tr> <td>S9</td> <td>9.9</td> <td>22.5</td> <td><strong>0.4x</strong></td> <td>100x50 submatrix</td> </tr> <tr> <td>S10</td> <td>12.1</td> <td>61.9</td> <td><strong>0.2x</strong></td> <td>500x100 submatrix</td> </tr> <tr> <td>S11</td> <td>10.8</td> <td>63.2</td> <td><strong>0.2x</strong></td> <td>500x500 submatrix</td> </tr> </tbody> </table> <p><strong>Average Performance:</strong></p> <ul> <li>SLAF wins cell-centric (1.2x-4.0x),</li> <li>TileDB wins gene-centric (3.3x-10x) and submatrix (2.5x-5.0x)</li> </ul> <h2 id=machine-learning-benchmarks><strong>Machine Learning Benchmarks</strong><a class=headerlink href=#machine-learning-benchmarks title="Permanent link">&para;</a></h2> <h3 id=raw-data-loading-performance><strong>Raw Data Loading Performance</strong><a class=headerlink href=#raw-data-loading-performance title="Permanent link">&para;</a></h3> <p>Raw data loading performance demonstrates the advantages of SLAF's optimized streaming architecture.</p> <table> <thead> <tr> <th>System</th> <th>Throughput (cells/sec)</th> <th>Notes</th> </tr> </thead> <tbody> <tr> <td><strong>SLAF</strong></td> <td><strong>24,587</strong></td> <td>Optimized streaming</td> </tr> <tr> <td>TileDB DataLoader</td> <td>518</td> <td>Custom PyTorch loader</td> </tr> </tbody> </table> <p><strong>Performance Comparison: 47.5x faster</strong></p> <h3 id=gpu-ready-output-performance><strong>GPU-Ready Output Performance</strong><a class=headerlink href=#gpu-ready-output-performance title="Permanent link">&para;</a></h3> <p>SLAF provides pre-tokenized sequences ready for GPU training, while TileDB DataLoader only provides raw data.</p> <table> <thead> <tr> <th>System</th> <th>Throughput (cells/sec)</th> <th>Throughput (tokens/sec)</th> <th>Output Type</th> </tr> </thead> <tbody> <tr> <td><strong>SLAF</strong></td> <td><strong>7,487</strong></td> <td><strong>15,332,896</strong></td> <td>Pre-tokenized sequences</td> </tr> <tr> <td>TileDB DataLoader</td> <td>N/A</td> <td>N/A</td> <td>Raw data only</td> </tr> </tbody> </table> <div class="admonition success"> <p class=admonition-title>GPU Training Advantage</p> <p>SLAF is the only system providing GPU-ready tokenized output, enabling efficient training of foundation models like Geneformer and scGPT.</p> </div> <h2 id=technical-implementation-comparison><strong>Technical Implementation Comparison</strong><a class=headerlink href=#technical-implementation-comparison title="Permanent link">&para;</a></h2> <table> <thead> <tr> <th>Aspect</th> <th>SLAF</th> <th>TileDB</th> </tr> </thead> <tbody> <tr> <td><strong>Storage</strong></td> <td>Arrow-interoperable columnar storage with Lance backend</td> <td>Arrow-interoperable array-native storage with TileDB backend</td> </tr> <tr> <td><strong>Metadata</strong></td> <td>Polars DataFrames for efficient filtering operations</td> <td>Arrow tables with Polars for filtering operations</td> </tr> <tr> <td><strong>Expression</strong></td> <td>Optimized sparse COO matrices with zero-copy access</td> <td>Smart indexing for both cell and gene based slicing</td> </tr> <tr> <td><strong>ML Integration</strong></td> <td>Native PyTorch DataLoader with tokenization support</td> <td>Needs third party custom PyTorch DataLoader for raw data</td> </tr> </tbody> </table> <h3 id=key-technical-advantages><strong>Key Technical Advantages</strong><a class=headerlink href=#key-technical-advantages title="Permanent link">&para;</a></h3> <div class="admonition success"> <p class=admonition-title>Optimized Streaming Architecture</p> <p>SLAF's asynchronous prefetching and background processing provide significant performance advantages over TileDB's synchronous loading approach.</p> </div> <div class="admonition success"> <p class=admonition-title>Enhanced ML Integration</p> <p>SLAF provides native PyTorch DataLoader integration with built-in tokenization support, while TileDB requires custom loader implementation.</p> </div> <div class="admonition success"> <p class=admonition-title>Memory Efficiency</p> <p>SLAF's optimized memory management and zero-copy operations result in 1.5-2.3x memory efficiency gains over TileDB.</p> </div> <div class="admonition success"> <p class=admonition-title>Simplified API</p> <p>SLAF provides a more streamlined API for common bioinformatics operations, while TileDB requires more manual configuration.</p> </div> <h2 id=migration-benefits><strong>Migration Benefits</strong><a class=headerlink href=#migration-benefits title="Permanent link">&para;</a></h2> <h3 id=performance-improvements><strong>Performance Improvements</strong><a class=headerlink href=#performance-improvements title="Permanent link">&para;</a></h3> <ul> <li><strong>9.4x faster</strong> bioinformatics operations</li> <li><strong>47.5x faster</strong> machine learning data loading</li> <li><strong>Faster conversion from h5ad</strong> to SLAF format, enabling rapid migration of existing datasets</li> <li><strong>Easy TileDB migration</strong>: 2-second conversion for 50k cell datasets with linear scaling</li> </ul> <h3 id=developer-experience><strong>Developer Experience</strong><a class=headerlink href=#developer-experience title="Permanent link">&para;</a></h3> <ul> <li><strong>Simplified API</strong> with familiar Polars operations</li> <li><strong>Scanpy-native lazy workflows</strong> with drop-in replacement of AnnData objects, enabling efficient lazy computation graphs for preprocessing and filtering pipelines</li> <li><strong>Enhanced ML integration</strong> with native PyTorch support</li> <li><strong>Built-in tokenization</strong> for foundation model training</li> </ul> <h2 id=use-case-recommendations><strong>Use Case Recommendations</strong><a class=headerlink href=#use-case-recommendations title="Permanent link">&para;</a></h2> <h3 id=choose-slaf-for><strong>Choose SLAF for:</strong><a class=headerlink href=#choose-slaf-for title="Permanent link">&para;</a></h3> <ul> <li><strong>High-throughput machine learning</strong> workflows requiring fast data loading</li> <li><strong>Foundation model training</strong> requiring GPU-ready tokenized sequences</li> <li><strong>Streaming applications</strong> where continuous data flow is critical</li> <li><strong>Memory-constrained environments</strong> where efficiency is important</li> </ul> <h3 id=choose-tiledb-for><strong>Choose TileDB for:</strong><a class=headerlink href=#choose-tiledb-for title="Permanent link">&para;</a></h3> <ul> <li><strong>Existing TileDB infrastructure</strong> where migration costs are high</li> <li><strong>Multi-modal data</strong> where TileDB's broader ecosystem is beneficial</li> </ul> <h2 id=conclusion><strong>Conclusion</strong><a class=headerlink href=#conclusion title="Permanent link">&para;</a></h2> <p>SLAF provides <strong>strong performance advantages</strong> over TileDB across most benchmark categories:</p> <ul> <li><strong>Metadata filtering</strong>: 9.4x-8.6x faster (cell and gene filtering)</li> <li><strong>Machine learning data loading</strong>: 47.5x faster</li> <li><strong>Expression queries</strong>: Mixed performance (SLAF wins cell-centric, TileDB wins gene-centric)</li> </ul> <p>While both systems represent modern approaches to single-cell data storage, SLAF's optimized streaming architecture, enhanced ML integration, and simplified API provide significant advantages for most use cases. The benchmarks demonstrate that SLAF's performance optimizations and developer-friendly design make it the preferred choice for high-throughput bioinformatics and machine learning workflows.</p> <hr> <p><em>For detailed migration guidance including TileDB to SLAF conversion, see <a href=../../user-guide/migrating-to-slaf/ >Migrating to SLAF</a>. For comprehensive benchmark results, see <a href=../bioinformatics_benchmarks/ >Bioinformatics Benchmarks</a> and <a href=../ml_benchmarks/ >ML Benchmarks</a>.</em></p> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="October 1, 2025 23:10:01 UTC">October 1, 2025</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="September 4, 2025 22:15:51 UTC">September 4, 2025</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Contributors> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 4a4 4 0 0 1 4 4 4 4 0 0 1-4 4 4 4 0 0 1-4-4 4 4 0 0 1 4-4m0 10c4.42 0 8 1.79 8 4v2H4v-2c0-2.21 3.58-4 8-4"/></svg> </span> <nav> Pavan Ramkumar </nav> </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2026 Pavan Ramkumar </div> </div> <div class=md-social> <a href=https://github.com/slaf-project/slaf target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.sections", "content.code.copy"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../../javascripts/mathjax.js></script> <script src=../../javascripts/theme-reset.js></script> <script src=../../javascripts/copy-enhancement.js></script> <script src=../../javascripts/theme-aware-assets.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>