<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Sparse Lazy Array Format - High-performance single-cell data storage and analysis"><meta name=author content="Pavan Ramkumar"><link href=https://slaf-project.github.io/slaf/benchmarks/bioinformatics_benchmarks/ rel=canonical><link href=../ rel=prev><link href=../ml_benchmarks/ rel=next><link rel=icon href=../../assets/slaf-icon-transparent-dark-mono.svg><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.7.1"><title>For Bioinformaticians - SLAF Documentation</title><link rel=stylesheet href=../../assets/stylesheets/main.484c7ddc.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.ab4e12ef.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../assets/_mkdocstrings.css><link rel=stylesheet href=../../stylesheets/extra.css><link rel=stylesheet href=../../stylesheets/logo-theme.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script id=__analytics>function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config",""),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#bioinformatics-benchmarks-slaf-vs-traditional-formats class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="SLAF Documentation" class="md-header__button md-logo" aria-label="SLAF Documentation" data-md-component=logo> <img src=../../assets/slaf-icon-transparent-dark-mono.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> SLAF Documentation </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> For Bioinformaticians </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg> </label> </form> <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap tabindex=0 data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/slaf-project/slaf title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> slaf-project/slaf </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="SLAF Documentation" class="md-nav__button md-logo" aria-label="SLAF Documentation" data-md-component=logo> <img src=../../assets/slaf-icon-transparent-dark-mono.svg alt=logo> </a> SLAF Documentation </label> <div class=md-nav__source> <a href=https://github.com/slaf-project/slaf title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg> </div> <div class=md-source__repository> slaf-project/slaf </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> <span class=md-ellipsis> Home </span> </a> </li> <li class=md-nav__item> <a href=../../getting-started/quickstart/ class=md-nav__link> <span class=md-ellipsis> Quick Start </span> </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex> <span class=md-ellipsis> User Guide </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> User Guide </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../user-guide/how-slaf-works/ class=md-nav__link> <span class=md-ellipsis> How SLAF Works </span> </a> </li> <li class=md-nav__item> <a href=../../user-guide/migrating-to-slaf/ class=md-nav__link> <span class=md-ellipsis> Migrating to SLAF </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4> <label class=md-nav__link for=__nav_4 id=__nav_4_label tabindex> <span class=md-ellipsis> Examples </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Examples </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../examples/getting-started/ class=md-nav__link> <span class=md-ellipsis> Getting Started </span> </a> </li> <li class=md-nav__item> <a href=../../examples/lazy-processing/ class=md-nav__link> <span class=md-ellipsis> Lazy Processing </span> </a> </li> <li class=md-nav__item> <a href=../../examples/ml-training/ class=md-nav__link> <span class=md-ellipsis> ML Training </span> </a> </li> <li class=md-nav__item> <a href=../../examples/sql-queries/ class=md-nav__link> <span class=md-ellipsis> SQL Queries </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5 id=__nav_5_label tabindex> <span class=md-ellipsis> Benchmarks </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_5_label aria-expanded=true> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> Benchmarks </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../ class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> For Bioinformaticians </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> For Bioinformaticians </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> <nav class=md-nav aria-label=Overview> <ul class=md-nav__list> <li class=md-nav__item> <a href=#test-dataset-synthetic_50k_processed class=md-nav__link> <span class=md-ellipsis> Test Dataset: synthetic_50k_processed </span> </a> </li> <li class=md-nav__item> <a href=#hardware-configuration class=md-nav__link> <span class=md-ellipsis> Hardware Configuration </span> </a> </li> <li class=md-nav__item> <a href=#performance-summary-across-all-categories class=md-nav__link> <span class=md-ellipsis> Performance Summary Across All Categories </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#cell-filtering-benchmarks class=md-nav__link> <span class=md-ellipsis> Cell Filtering Benchmarks </span> </a> <nav class=md-nav aria-label="Cell Filtering Benchmarks"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#performance-results class=md-nav__link> <span class=md-ellipsis> Performance Results </span> </a> </li> <li class=md-nav__item> <a href=#key-insights class=md-nav__link> <span class=md-ellipsis> Key Insights </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#gene-filtering-benchmarks class=md-nav__link> <span class=md-ellipsis> Gene Filtering Benchmarks </span> </a> <nav class=md-nav aria-label="Gene Filtering Benchmarks"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#performance-results_1 class=md-nav__link> <span class=md-ellipsis> Performance Results </span> </a> </li> <li class=md-nav__item> <a href=#key-insights_1 class=md-nav__link> <span class=md-ellipsis> Key Insights </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#expression-queries-benchmarks class=md-nav__link> <span class=md-ellipsis> Expression Queries Benchmarks </span> </a> <nav class=md-nav aria-label="Expression Queries Benchmarks"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#performance-results_2 class=md-nav__link> <span class=md-ellipsis> Performance Results </span> </a> </li> <li class=md-nav__item> <a href=#key-insights_2 class=md-nav__link> <span class=md-ellipsis> Key Insights </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#lazy-computation-preprocessing-pipelines class=md-nav__link> <span class=md-ellipsis> Lazy Computation (Preprocessing Pipelines) </span> </a> <nav class=md-nav aria-label="Lazy Computation (Preprocessing Pipelines)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#traditional-approach-eager-processing class=md-nav__link> <span class=md-ellipsis> Traditional Approach (Eager Processing) </span> </a> </li> <li class=md-nav__item> <a href=#slaf-approach-lazy-computation-graph class=md-nav__link> <span class=md-ellipsis> SLAF Approach (Lazy Computation Graph) </span> </a> </li> <li class=md-nav__item> <a href=#performance-results_3 class=md-nav__link> <span class=md-ellipsis> Performance Results </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../ml_benchmarks/ class=md-nav__link> <span class=md-ellipsis> For ML Engineers </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_6> <label class=md-nav__link for=__nav_6 id=__nav_6_label tabindex> <span class=md-ellipsis> API Reference </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_6_label aria-expanded=false> <label class=md-nav__title for=__nav_6> <span class="md-nav__icon md-icon"></span> API Reference </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../api/core/ class=md-nav__link> <span class=md-ellipsis> Core </span> </a> </li> <li class=md-nav__item> <a href=../../api/data/ class=md-nav__link> <span class=md-ellipsis> Data </span> </a> </li> <li class=md-nav__item> <a href=../../api/integrations/ class=md-nav__link> <span class=md-ellipsis> Integrations </span> </a> </li> <li class=md-nav__item> <a href=../../api/ml/ class=md-nav__link> <span class=md-ellipsis> ML </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_7> <label class=md-nav__link for=__nav_7 id=__nav_7_label tabindex> <span class=md-ellipsis> Development </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_7_label aria-expanded=false> <label class=md-nav__title for=__nav_7> <span class="md-nav__icon md-icon"></span> Development </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../development/contributing/ class=md-nav__link> <span class=md-ellipsis> For Contributors </span> </a> </li> <li class=md-nav__item> <a href=../../development/maintaining/ class=md-nav__link> <span class=md-ellipsis> For Maintainers </span> </a> </li> <li class=md-nav__item> <a href=../../development/benchmarks/ class=md-nav__link> <span class=md-ellipsis> Benchmarks </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_8> <label class=md-nav__link for=__nav_8 id=__nav_8_label tabindex> <span class=md-ellipsis> Blog </span> <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_8_label aria-expanded=false> <label class=md-nav__title for=__nav_8> <span class="md-nav__icon md-icon"></span> Blog </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../blog/ class=md-nav__link> <span class=md-ellipsis> Blog Home </span> </a> </li> <li class=md-nav__item> <a href=../../blog/huggingface/ class=md-nav__link> <span class=md-ellipsis> SLAF on Hugging Face </span> </a> </li> <li class=md-nav__item> <a href=../../blog/introducing-slaf/ class=md-nav__link> <span class=md-ellipsis> Introducing SLAF </span> </a> </li> <li class=md-nav__item> <a href=../../blog/blazing-fast-dataloaders/ class=md-nav__link> <span class=md-ellipsis> Blazing Fast Dataloaders </span> </a> </li> <li class=md-nav__item> <a href=../../blog/blazing-fast-dataloaders-2/ class=md-nav__link> <span class=md-ellipsis> Mixture of Scanners </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> <span class=md-ellipsis> Overview </span> </a> <nav class=md-nav aria-label=Overview> <ul class=md-nav__list> <li class=md-nav__item> <a href=#test-dataset-synthetic_50k_processed class=md-nav__link> <span class=md-ellipsis> Test Dataset: synthetic_50k_processed </span> </a> </li> <li class=md-nav__item> <a href=#hardware-configuration class=md-nav__link> <span class=md-ellipsis> Hardware Configuration </span> </a> </li> <li class=md-nav__item> <a href=#performance-summary-across-all-categories class=md-nav__link> <span class=md-ellipsis> Performance Summary Across All Categories </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#cell-filtering-benchmarks class=md-nav__link> <span class=md-ellipsis> Cell Filtering Benchmarks </span> </a> <nav class=md-nav aria-label="Cell Filtering Benchmarks"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#performance-results class=md-nav__link> <span class=md-ellipsis> Performance Results </span> </a> </li> <li class=md-nav__item> <a href=#key-insights class=md-nav__link> <span class=md-ellipsis> Key Insights </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#gene-filtering-benchmarks class=md-nav__link> <span class=md-ellipsis> Gene Filtering Benchmarks </span> </a> <nav class=md-nav aria-label="Gene Filtering Benchmarks"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#performance-results_1 class=md-nav__link> <span class=md-ellipsis> Performance Results </span> </a> </li> <li class=md-nav__item> <a href=#key-insights_1 class=md-nav__link> <span class=md-ellipsis> Key Insights </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#expression-queries-benchmarks class=md-nav__link> <span class=md-ellipsis> Expression Queries Benchmarks </span> </a> <nav class=md-nav aria-label="Expression Queries Benchmarks"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#performance-results_2 class=md-nav__link> <span class=md-ellipsis> Performance Results </span> </a> </li> <li class=md-nav__item> <a href=#key-insights_2 class=md-nav__link> <span class=md-ellipsis> Key Insights </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#lazy-computation-preprocessing-pipelines class=md-nav__link> <span class=md-ellipsis> Lazy Computation (Preprocessing Pipelines) </span> </a> <nav class=md-nav aria-label="Lazy Computation (Preprocessing Pipelines)"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#traditional-approach-eager-processing class=md-nav__link> <span class=md-ellipsis> Traditional Approach (Eager Processing) </span> </a> </li> <li class=md-nav__item> <a href=#slaf-approach-lazy-computation-graph class=md-nav__link> <span class=md-ellipsis> SLAF Approach (Lazy Computation Graph) </span> </a> </li> <li class=md-nav__item> <a href=#performance-results_3 class=md-nav__link> <span class=md-ellipsis> Performance Results </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=bioinformatics-benchmarks-slaf-vs-traditional-formats>Bioinformatics Benchmarks: SLAF vs Traditional Formats<a class=headerlink href=#bioinformatics-benchmarks-slaf-vs-traditional-formats title="Permanent link">&para;</a></h1> <p>SLAF provides <strong>dramatic performance improvements</strong> over traditional single-cell data formats for common bioinformatics operations. This document presents comprehensive benchmarks comparing SLAF against h5ad (AnnData) and TileDB SOMA across realistic bioinformatics workflows.</p> <h2 id=overview><strong>Overview</strong><a class=headerlink href=#overview title="Permanent link">&para;</a></h2> <p>These benchmarks demonstrate SLAF's performance advantages in three key areas:</p> <ol> <li><strong>Metadata Filtering</strong> - Cell and gene filtering operations</li> <li><strong>Expression Queries</strong> - Retrieving expression data for specific cells/genes</li> <li><strong>Preprocessing Pipelines</strong> - Scanpy-based preprocessing workflows</li> </ol> <h3 id=test-dataset-synthetic_50k_processed><strong>Test Dataset: synthetic_50k_processed</strong><a class=headerlink href=#test-dataset-synthetic_50k_processed title="Permanent link">&para;</a></h3> <ul> <li><strong>Cells</strong>: 49,955 cells</li> <li><strong>Genes</strong>: 25,000 genes</li> <li><strong>Input File Size</strong>: ~722MB h5ad file</li> </ul> <h3 id=hardware-configuration><strong>Hardware Configuration</strong><a class=headerlink href=#hardware-configuration title="Permanent link">&para;</a></h3> <ul> <li><strong>Machine</strong>: Apple MacBook Pro with M1 Max</li> <li><strong>Memory</strong>: 32 GB RAM</li> <li><strong>Storage</strong>: 1 TB NVMe SSD (local disk)</li> <li><strong>OS</strong>: macOS 13.6.1</li> <li><strong>Python</strong>: 3.12.0</li> </ul> <h3 id=performance-summary-across-all-categories><strong>Performance Summary Across All Categories</strong><a class=headerlink href=#performance-summary-across-all-categories title="Permanent link">&para;</a></h3> <p><strong>SLAF vs h5ad Speedup:</strong></p> <div class="language-text highlight"><pre><span></span><code><span id=__span-0-1><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a>Speedup (x faster)
</span><span id=__span-0-2><a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a>─────────────────────────────────────────────────────────
</span><span id=__span-0-3><a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a>Cell Filtering     ████████████████████████████████████ 92.3x
</span><span id=__span-0-4><a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a>Gene Filtering     ████████████████ 17.3x
</span><span id=__span-0-5><a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a>Expression Queries █████████ 9.5x
</span><span id=__span-0-6><a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a>─────────────────────────────────────────────────────────
</span><span id=__span-0-7><a id=__codelineno-0-7 name=__codelineno-0-7 href=#__codelineno-0-7></a>                    0   20   40   60   80  100  120
</span></code></pre></div> <p><strong>SLAF vs TileDB Speedup:</strong></p> <div class="language-text highlight"><pre><span></span><code><span id=__span-1-1><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a>Speedup (x faster)
</span><span id=__span-1-2><a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a>─────────────────────────────────────────────────────────
</span><span id=__span-1-3><a id=__codelineno-1-3 name=__codelineno-1-3 href=#__codelineno-1-3></a>Cell Filtering     ████ 9.4x
</span><span id=__span-1-4><a id=__codelineno-1-4 name=__codelineno-1-4 href=#__codelineno-1-4></a>Gene Filtering     ████████ 8.6x
</span><span id=__span-1-5><a id=__codelineno-1-5 name=__codelineno-1-5 href=#__codelineno-1-5></a>Expression Queries █ 0.9x (competitive)
</span><span id=__span-1-6><a id=__codelineno-1-6 name=__codelineno-1-6 href=#__codelineno-1-6></a>─────────────────────────────────────────────────────────
</span><span id=__span-1-7><a id=__codelineno-1-7 name=__codelineno-1-7 href=#__codelineno-1-7></a>                    0    2    4    6    8   10   12
</span></code></pre></div> <h2 id=cell-filtering-benchmarks><strong>Cell Filtering Benchmarks</strong><a class=headerlink href=#cell-filtering-benchmarks title="Permanent link">&para;</a></h2> <p>Cell filtering is a fundamental operation in single-cell analysis, used for quality control, cell type selection, and data subsetting.</p> <h3 id=performance-results><strong>Performance Results</strong><a class=headerlink href=#performance-results title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>Scenario</th> <th>h5ad Total (ms)</th> <th>SLAF Total (ms)</th> <th>TileDB Total (ms)</th> <th>SLAF vs h5ad</th> <th>SLAF vs TileDB</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td>S1</td> <td>530.0</td> <td>2.9</td> <td>20.9</td> <td><strong>183.6x</strong></td> <td><strong>7.3x</strong></td> <td>Cells with &gt;=500 genes</td> </tr> <tr> <td>S2</td> <td>169.7</td> <td>2.0</td> <td>21.3</td> <td><strong>83.3x</strong></td> <td><strong>10.5x</strong></td> <td>High UMI count (total_counts &gt; 2000)</td> </tr> <tr> <td>S3</td> <td>170.7</td> <td>1.9</td> <td>21.3</td> <td><strong>92.2x</strong></td> <td><strong>11.5x</strong></td> <td>Mitochondrial fraction &lt; 0.1</td> </tr> <tr> <td>S4</td> <td>177.1</td> <td>2.0</td> <td>18.6</td> <td><strong>86.7x</strong></td> <td><strong>9.1x</strong></td> <td>Complex multi-condition filter</td> </tr> <tr> <td>S5</td> <td>186.6</td> <td>2.8</td> <td>18.1</td> <td><strong>67.2x</strong></td> <td><strong>6.5x</strong></td> <td>Cell type annotation filter</td> </tr> <tr> <td>S6</td> <td>171.4</td> <td>2.0</td> <td>20.9</td> <td><strong>86.3x</strong></td> <td><strong>10.5x</strong></td> <td>Cells from batch_1</td> </tr> <tr> <td>S7</td> <td>207.0</td> <td>2.3</td> <td>23.8</td> <td><strong>89.4x</strong></td> <td><strong>10.3x</strong></td> <td>Cells in clusters 0,1 from batch_1</td> </tr> <tr> <td>S8</td> <td>170.9</td> <td>2.1</td> <td>23.0</td> <td><strong>79.6x</strong></td> <td><strong>10.7x</strong></td> <td>High-quality cells (&gt;=1000 genes, &lt;=10% mt)</td> </tr> <tr> <td>S9</td> <td>172.1</td> <td>2.5</td> <td>19.3</td> <td><strong>70.2x</strong></td> <td><strong>7.9x</strong></td> <td>Cells with 800-2000 total counts</td> </tr> <tr> <td>S10</td> <td>173.5</td> <td>2.1</td> <td>20.8</td> <td><strong>84.6x</strong></td> <td><strong>10.1x</strong></td> <td>Cells with 200-1500 genes</td> </tr> </tbody> </table> <p><strong>Average Performance:</strong></p> <ul> <li><strong>SLAF vs h5ad</strong>: <strong>92.3x faster</strong></li> <li><strong>SLAF vs TileDB</strong>: <strong>9.4x faster</strong></li> <li><strong>Memory Usage</strong>: SLAF uses 115.7x less memory than h5ad</li> </ul> <p><strong>Cell Filtering Performance Summary:</strong></p> <div class="language-text highlight"><pre><span></span><code><span id=__span-2-1><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a>Speedup (x faster)
</span><span id=__span-2-2><a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a>─────────────────────────────────────────────────────────
</span><span id=__span-2-3><a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a>SLAF vs h5ad        ████████████████████████████████████ 92.3x
</span><span id=__span-2-4><a id=__codelineno-2-4 name=__codelineno-2-4 href=#__codelineno-2-4></a>SLAF vs TileDB      ████ 9.4x
</span><span id=__span-2-5><a id=__codelineno-2-5 name=__codelineno-2-5 href=#__codelineno-2-5></a>─────────────────────────────────────────────────────────
</span><span id=__span-2-6><a id=__codelineno-2-6 name=__codelineno-2-6 href=#__codelineno-2-6></a>                    0   20   40   60   80  100  120
</span></code></pre></div> <h3 id=key-insights><strong>Key Insights</strong><a class=headerlink href=#key-insights title="Permanent link">&para;</a></h3> <div class="admonition success"> <p class=admonition-title>Dramatic Performance Advantage</p> <p>SLAF achieves <strong>92.3x average speedup</strong> over h5ad for cell filtering operations, demonstrating the massive performance benefits of modern columnar storage and optimized querying.</p> </div> <div class="admonition info"> <p class=admonition-title>Columnar Format Efficiency</p> <p>Both SLAF and TileDB (Arrow-interoperable formats) significantly outperform h5ad, with SLAF providing an additional 9.4x advantage over TileDB through its optimized streaming architecture.</p> </div> <h2 id=gene-filtering-benchmarks><strong>Gene Filtering Benchmarks</strong><a class=headerlink href=#gene-filtering-benchmarks title="Permanent link">&para;</a></h2> <p>Gene filtering operations are essential for feature selection, quality control, and differential expression in single-cell analysis.</p> <h3 id=performance-results_1><strong>Performance Results</strong><a class=headerlink href=#performance-results_1 title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>Scenario</th> <th>h5ad Total (ms)</th> <th>SLAF Total (ms)</th> <th>TileDB Total (ms)</th> <th>SLAF vs h5ad</th> <th>SLAF vs TileDB</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td>S1</td> <td>43.4</td> <td>3.0</td> <td>22.3</td> <td><strong>14.6x</strong></td> <td><strong>7.5x</strong></td> <td>Genes expressed in &gt;=10 cells</td> </tr> <tr> <td>S2</td> <td>32.3</td> <td>1.7</td> <td>19.5</td> <td><strong>19.4x</strong></td> <td><strong>11.7x</strong></td> <td>Genes with &gt;=100 total counts</td> </tr> <tr> <td>S3</td> <td>32.1</td> <td>1.8</td> <td>9.3</td> <td><strong>17.4x</strong></td> <td><strong>5.0x</strong></td> <td>Genes with mean expression &gt;=0.1</td> </tr> <tr> <td>S4</td> <td>31.1</td> <td>1.6</td> <td>15.8</td> <td><strong>19.9x</strong></td> <td><strong>10.1x</strong></td> <td>Exclude mitochondrial genes</td> </tr> <tr> <td>S5</td> <td>32.7</td> <td>1.7</td> <td>16.9</td> <td><strong>19.7x</strong></td> <td><strong>10.2x</strong></td> <td>Highly variable genes</td> </tr> <tr> <td>S6</td> <td>31.7</td> <td>2.1</td> <td>15.8</td> <td><strong>15.4x</strong></td> <td><strong>7.7x</strong></td> <td>Non-highly variable genes</td> </tr> <tr> <td>S7</td> <td>31.5</td> <td>2.0</td> <td>18.2</td> <td><strong>15.8x</strong></td> <td><strong>9.1x</strong></td> <td>Genes in &gt;=50 cells with &gt;=500 total counts</td> </tr> <tr> <td>S8</td> <td>31.7</td> <td>1.9</td> <td>19.8</td> <td><strong>17.0x</strong></td> <td><strong>10.6x</strong></td> <td>Genes with 100-10000 total counts</td> </tr> <tr> <td>S9</td> <td>33.2</td> <td>2.0</td> <td>11.3</td> <td><strong>16.4x</strong></td> <td><strong>5.6x</strong></td> <td>Genes in 5-1000 cells</td> </tr> </tbody> </table> <p><strong>Average Performance:</strong></p> <ul> <li><strong>SLAF vs h5ad</strong>: <strong>17.3x faster</strong></li> <li><strong>SLAF vs TileDB</strong>: <strong>8.6x faster</strong></li> <li><strong>Memory Usage</strong>: SLAF uses 2.2x less memory than h5ad</li> </ul> <p><strong>Gene Filtering Performance Summary:</strong></p> <div class="language-text highlight"><pre><span></span><code><span id=__span-3-1><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a>Speedup (x faster)
</span><span id=__span-3-2><a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a>─────────────────────────────────────────────────────────
</span><span id=__span-3-3><a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a>SLAF vs h5ad        ████████████████ 17.3x
</span><span id=__span-3-4><a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a>SLAF vs TileDB      ████████ 8.6x
</span><span id=__span-3-5><a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a>─────────────────────────────────────────────────────────
</span><span id=__span-3-6><a id=__codelineno-3-6 name=__codelineno-3-6 href=#__codelineno-3-6></a>                    0    5   10   15   20   25   30
</span></code></pre></div> <h3 id=key-insights_1><strong>Key Insights</strong><a class=headerlink href=#key-insights_1 title="Permanent link">&para;</a></h3> <div class="admonition success"> <p class=admonition-title>Consistent High Performance</p> <p>SLAF maintains consistent 14x+ speedups across all gene filtering scenarios, demonstrating robust optimization of Polars operations and modern storage formats.</p> </div> <div class="admonition info"> <p class=admonition-title>Memory Efficiency</p> <p>Gene filtering operations show moderate memory efficiency gains, with SLAF using 2.2x less memory than h5ad for equivalent operations.</p> </div> <h2 id=expression-queries-benchmarks><strong>Expression Queries Benchmarks</strong><a class=headerlink href=#expression-queries-benchmarks title="Permanent link">&para;</a></h2> <p>Expression queries retrieve specific expression data for cells or genes, supporting analysis workflows that require targeted data access.</p> <h3 id=performance-results_2><strong>Performance Results</strong><a class=headerlink href=#performance-results_2 title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>Scenario</th> <th>h5ad Total (ms)</th> <th>SLAF Total (ms)</th> <th>TileDB Total (ms)</th> <th>SLAF vs h5ad</th> <th>SLAF vs TileDB</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td>S1</td> <td>484.5</td> <td>16.1</td> <td>63.8</td> <td><strong>30.1x</strong></td> <td><strong>4.0x</strong></td> <td>Single cell expression</td> </tr> <tr> <td>S2</td> <td>251.3</td> <td>13.9</td> <td>19.3</td> <td><strong>18.1x</strong></td> <td><strong>1.4x</strong></td> <td>Another single cell</td> </tr> <tr> <td>S3</td> <td>328.3</td> <td>14.2</td> <td>18.1</td> <td><strong>23.1x</strong></td> <td><strong>1.3x</strong></td> <td>Two cells</td> </tr> <tr> <td>S4</td> <td>233.2</td> <td>15.5</td> <td>19.0</td> <td><strong>15.1x</strong></td> <td><strong>1.2x</strong></td> <td>Three cells</td> </tr> <tr> <td>S5</td> <td>232.7</td> <td>523.7</td> <td>150.8</td> <td><strong>0.4x</strong></td> <td><strong>0.3x</strong></td> <td>Single gene across all cells</td> </tr> <tr> <td>S6</td> <td>203.4</td> <td>442.6</td> <td>84.2</td> <td><strong>0.5x</strong></td> <td><strong>0.2x</strong></td> <td>Another single gene</td> </tr> <tr> <td>S7</td> <td>256.1</td> <td>303.0</td> <td>97.7</td> <td><strong>0.8x</strong></td> <td><strong>0.3x</strong></td> <td>Two genes</td> </tr> <tr> <td>S8</td> <td>212.0</td> <td>655.9</td> <td>83.3</td> <td><strong>0.3x</strong></td> <td><strong>0.1x</strong></td> <td>Three genes</td> </tr> <tr> <td>S9</td> <td>221.4</td> <td>22.5</td> <td>9.9</td> <td><strong>9.9x</strong></td> <td><strong>0.4x</strong></td> <td>100x50 submatrix</td> </tr> <tr> <td>S10</td> <td>168.3</td> <td>61.9</td> <td>12.1</td> <td><strong>2.7x</strong></td> <td><strong>0.2x</strong></td> <td>500x100 submatrix</td> </tr> <tr> <td>S11</td> <td>212.2</td> <td>63.2</td> <td>10.8</td> <td><strong>3.4x</strong></td> <td><strong>0.2x</strong></td> <td>500x500 submatrix</td> </tr> </tbody> </table> <p><strong>Average Performance:</strong></p> <ul> <li><strong>SLAF vs h5ad</strong>: <strong>9.5x faster</strong></li> <li><strong>SLAF vs TileDB</strong>: <strong>0.9x faster</strong></li> <li><strong>Memory Usage</strong>: SLAF uses 154.6x less memory than h5ad</li> </ul> <p><strong>Expression Queries Performance Summary:</strong></p> <div class="language-text highlight"><pre><span></span><code><span id=__span-4-1><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a>Speedup (x faster)
</span><span id=__span-4-2><a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a>─────────────────────────────────────────────────────────
</span><span id=__span-4-3><a id=__codelineno-4-3 name=__codelineno-4-3 href=#__codelineno-4-3></a>SLAF vs h5ad        █████████ 9.5x
</span><span id=__span-4-4><a id=__codelineno-4-4 name=__codelineno-4-4 href=#__codelineno-4-4></a>SLAF vs TileDB      █ 0.9x (competitive)
</span><span id=__span-4-5><a id=__codelineno-4-5 name=__codelineno-4-5 href=#__codelineno-4-5></a>─────────────────────────────────────────────────────────
</span><span id=__span-4-6><a id=__codelineno-4-6 name=__codelineno-4-6 href=#__codelineno-4-6></a>                    0    2    4    6    8   10   12
</span></code></pre></div> <h3 id=key-insights_2><strong>Key Insights</strong><a class=headerlink href=#key-insights_2 title="Permanent link">&para;</a></h3> <div class="admonition success"> <p class=admonition-title>Query Optimization</p> <p>SLAF's expression query performance demonstrates efficient sparse matrix operations and optimized data access patterns, achieving 9.5x average speedup over h5ad.</p> </div> <div class="admonition info"> <p class=admonition-title>TileDB's Expression Query Strengths</p> <p>TileDB demonstrates impressive performance for gene expression queries and submatrix operations, often outperforming both SLAF and h5ad. For single gene queries across all cells (S5-S8), TileDB shows 2.5-10x speedup over SLAF, highlighting its optimized columnar access patterns for gene-centric operations.</p> </div> <div class="admonition info"> <p class=admonition-title>SLAF's Cell-Centric Advantages</p> <p>SLAF maintains strong performance for cell-centric queries (S1-S4, S9-S11), achieving 15-30x speedup over h5ad for single cell and submatrix operations, while TileDB shows competitive or superior performance for larger submatrices.</p> </div> <div class="admonition info"> <p class=admonition-title>Mixed Performance Profile</p> <p>The benchmarks reveal a nuanced performance landscape: SLAF excels at cell-centric operations and metadata filtering, while TileDB demonstrates superior performance for gene-centric expression queries and large submatrix operations. This suggests different systems may be optimal for different analysis workflows.</p> </div> <h2 id=lazy-computation-preprocessing-pipelines><strong>Lazy Computation (Preprocessing Pipelines)</strong><a class=headerlink href=#lazy-computation-preprocessing-pipelines title="Permanent link">&para;</a></h2> <p>SLAF enables <strong>lazy computation graphs</strong> that build complex preprocessing pipelines and only execute them on the slice of interest, similar to Dask's delayed computation patterns.</p> <h3 id=traditional-approach-eager-processing>Traditional Approach (Eager Processing)<a class=headerlink href=#traditional-approach-eager-processing title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-5-1><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=c1># Each step loads data into memory</span>
</span><span id=__span-5-2><a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=n>adata</span> <span class=o>=</span> <span class=n>sc</span><span class=o>.</span><span class=n>read_h5ad</span><span class=p>(</span><span class=s2>&quot;data.h5ad&quot;</span><span class=p>,</span> <span class=n>backed</span><span class=o>=</span><span class=s2>&quot;r&quot;</span><span class=p>)</span>
</span><span id=__span-5-3><a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a>
</span><span id=__span-5-4><a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=c1># QC metrics calculation</span>
</span><span id=__span-5-5><a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a><span class=n>sc</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>calculate_qc_metrics</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
</span><span id=__span-5-6><a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a>
</span><span id=__span-5-7><a id=__codelineno-5-7 name=__codelineno-5-7 href=#__codelineno-5-7></a><span class=c1># Cell filtering with min_counts and min_genes</span>
</span><span id=__span-5-8><a id=__codelineno-5-8 name=__codelineno-5-8 href=#__codelineno-5-8></a><span class=n>sc</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>filter_cells</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span> <span class=n>min_counts</span><span class=o>=</span><span class=mi>500</span><span class=p>,</span> <span class=n>min_genes</span><span class=o>=</span><span class=mi>200</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
</span><span id=__span-5-9><a id=__codelineno-5-9 name=__codelineno-5-9 href=#__codelineno-5-9></a>
</span><span id=__span-5-10><a id=__codelineno-5-10 name=__codelineno-5-10 href=#__codelineno-5-10></a><span class=c1># Gene filtering</span>
</span><span id=__span-5-11><a id=__codelineno-5-11 name=__codelineno-5-11 href=#__codelineno-5-11></a><span class=n>sc</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>filter_genes</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span> <span class=n>min_counts</span><span class=o>=</span><span class=mi>10</span><span class=p>,</span> <span class=n>min_cells</span><span class=o>=</span><span class=mi>5</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
</span><span id=__span-5-12><a id=__codelineno-5-12 name=__codelineno-5-12 href=#__codelineno-5-12></a>
</span><span id=__span-5-13><a id=__codelineno-5-13 name=__codelineno-5-13 href=#__codelineno-5-13></a><span class=c1># Normalization</span>
</span><span id=__span-5-14><a id=__codelineno-5-14 name=__codelineno-5-14 href=#__codelineno-5-14></a><span class=n>sc</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>normalize_total</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span> <span class=n>target_sum</span><span class=o>=</span><span class=mf>1e4</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
</span><span id=__span-5-15><a id=__codelineno-5-15 name=__codelineno-5-15 href=#__codelineno-5-15></a>
</span><span id=__span-5-16><a id=__codelineno-5-16 name=__codelineno-5-16 href=#__codelineno-5-16></a><span class=c1># Log transformation</span>
</span><span id=__span-5-17><a id=__codelineno-5-17 name=__codelineno-5-17 href=#__codelineno-5-17></a><span class=n>sc</span><span class=o>.</span><span class=n>pp</span><span class=o>.</span><span class=n>log1p</span><span class=p>(</span><span class=n>adata</span><span class=p>)</span>
</span><span id=__span-5-18><a id=__codelineno-5-18 name=__codelineno-5-18 href=#__codelineno-5-18></a>
</span><span id=__span-5-19><a id=__codelineno-5-19 name=__codelineno-5-19 href=#__codelineno-5-19></a><span class=c1># Each operation processes the entire dataset</span>
</span><span id=__span-5-20><a id=__codelineno-5-20 name=__codelineno-5-20 href=#__codelineno-5-20></a><span class=n>expression</span> <span class=o>=</span> <span class=n>adata</span><span class=o>.</span><span class=n>X</span><span class=p>[</span><span class=n>cell_ids</span><span class=p>,</span> <span class=n>gene_ids</span><span class=p>]</span>
</span></code></pre></div> <h3 id=slaf-approach-lazy-computation-graph>SLAF Approach (Lazy Computation Graph)<a class=headerlink href=#slaf-approach-lazy-computation-graph title="Permanent link">&para;</a></h3> <div class="language-python highlight"><pre><span></span><code><span id=__span-6-1><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=c1># Build lazy computation graph</span>
</span><span id=__span-6-2><a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=n>adata</span> <span class=o>=</span> <span class=n>LazyAnnData</span><span class=p>(</span><span class=s2>&quot;data.slaf&quot;</span><span class=p>)</span>  <span class=c1># LazyAnnData object</span>
</span><span id=__span-6-3><a id=__codelineno-6-3 name=__codelineno-6-3 href=#__codelineno-6-3></a>
</span><span id=__span-6-4><a id=__codelineno-6-4 name=__codelineno-6-4 href=#__codelineno-6-4></a><span class=c1># QC metrics calculation (lazy)</span>
</span><span id=__span-6-5><a id=__codelineno-6-5 name=__codelineno-6-5 href=#__codelineno-6-5></a><span class=n>pp</span><span class=o>.</span><span class=n>calculate_qc_metrics</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
</span><span id=__span-6-6><a id=__codelineno-6-6 name=__codelineno-6-6 href=#__codelineno-6-6></a>
</span><span id=__span-6-7><a id=__codelineno-6-7 name=__codelineno-6-7 href=#__codelineno-6-7></a><span class=c1># Cell filtering (lazy)</span>
</span><span id=__span-6-8><a id=__codelineno-6-8 name=__codelineno-6-8 href=#__codelineno-6-8></a><span class=n>pp</span><span class=o>.</span><span class=n>filter_cells</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span> <span class=n>min_counts</span><span class=o>=</span><span class=mi>500</span><span class=p>,</span> <span class=n>min_genes</span><span class=o>=</span><span class=mi>200</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
</span><span id=__span-6-9><a id=__codelineno-6-9 name=__codelineno-6-9 href=#__codelineno-6-9></a>
</span><span id=__span-6-10><a id=__codelineno-6-10 name=__codelineno-6-10 href=#__codelineno-6-10></a><span class=c1># Gene filtering (lazy)</span>
</span><span id=__span-6-11><a id=__codelineno-6-11 name=__codelineno-6-11 href=#__codelineno-6-11></a><span class=n>pp</span><span class=o>.</span><span class=n>filter_genes</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span> <span class=n>min_counts</span><span class=o>=</span><span class=mi>10</span><span class=p>,</span> <span class=n>min_cells</span><span class=o>=</span><span class=mi>5</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
</span><span id=__span-6-12><a id=__codelineno-6-12 name=__codelineno-6-12 href=#__codelineno-6-12></a>
</span><span id=__span-6-13><a id=__codelineno-6-13 name=__codelineno-6-13 href=#__codelineno-6-13></a><span class=c1># Normalization (lazy)</span>
</span><span id=__span-6-14><a id=__codelineno-6-14 name=__codelineno-6-14 href=#__codelineno-6-14></a><span class=n>pp</span><span class=o>.</span><span class=n>normalize_total</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span> <span class=n>target_sum</span><span class=o>=</span><span class=mf>1e4</span><span class=p>,</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
</span><span id=__span-6-15><a id=__codelineno-6-15 name=__codelineno-6-15 href=#__codelineno-6-15></a>
</span><span id=__span-6-16><a id=__codelineno-6-16 name=__codelineno-6-16 href=#__codelineno-6-16></a><span class=c1># Log transformation (lazy)</span>
</span><span id=__span-6-17><a id=__codelineno-6-17 name=__codelineno-6-17 href=#__codelineno-6-17></a><span class=n>pp</span><span class=o>.</span><span class=n>log1p</span><span class=p>(</span><span class=n>adata</span><span class=p>)</span>
</span><span id=__span-6-18><a id=__codelineno-6-18 name=__codelineno-6-18 href=#__codelineno-6-18></a>
</span><span id=__span-6-19><a id=__codelineno-6-19 name=__codelineno-6-19 href=#__codelineno-6-19></a><span class=c1># Only execute the computation on the slice of interest</span>
</span><span id=__span-6-20><a id=__codelineno-6-20 name=__codelineno-6-20 href=#__codelineno-6-20></a><span class=n>expression</span> <span class=o>=</span> <span class=n>adata</span><span class=o>.</span><span class=n>X</span><span class=p>[</span><span class=n>cell_ids</span><span class=p>,</span> <span class=n>gene_ids</span><span class=p>]</span><span class=o>.</span><span class=n>compute</span><span class=p>()</span>  <span class=c1># LazyExpressionMatrix.compute()</span>
</span></code></pre></div> <h3 id=performance-results_3>Performance Results<a class=headerlink href=#performance-results_3 title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>Operation</th> <th>Traditional Total (ms)</th> <th>SLAF Total (ms)</th> <th>Total Speedup</th> <th>Memory Efficiency</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td>S1</td> <td>1150.4</td> <td>2577.0</td> <td>0.4x</td> <td>148.1x</td> <td>Calculate QC metrics</td> </tr> <tr> <td>S2</td> <td>941.0</td> <td>2045.1</td> <td>0.5x</td> <td>1.5x</td> <td>Filter cells (min_counts=500, min_genes=200)</td> </tr> <tr> <td>S3</td> <td>929.7</td> <td>2058.9</td> <td>0.5x</td> <td>1.5x</td> <td>Filter cells (min_counts=100, min_genes=50)</td> </tr> <tr> <td>S4</td> <td>862.1</td> <td>2184.9</td> <td>0.4x</td> <td>1.5x</td> <td>Filter cells (max_counts=10000, max_genes=3000)</td> </tr> <tr> <td>S5</td> <td>1208.9</td> <td>2586.7</td> <td>0.5x</td> <td>1.5x</td> <td>Filter genes (min_counts=10, min_cells=5)</td> </tr> <tr> <td>S6</td> <td>1183.5</td> <td>2516.4</td> <td>0.5x</td> <td>1.5x</td> <td>Filter genes (min_counts=20, min_cells=5)</td> </tr> <tr> <td>S7</td> <td>436.2</td> <td>3675.7</td> <td>0.1x</td> <td>1.5x</td> <td>Normalize total (target_sum=1e4)</td> </tr> <tr> <td>S8</td> <td>410.9</td> <td>2843.9</td> <td>0.1x</td> <td>1.5x</td> <td>Normalize total (target_sum=1e6)</td> </tr> <tr> <td>S9</td> <td>628.2</td> <td>2083.7</td> <td>0.3x</td> <td>0.0x</td> <td>Log1p transformation</td> </tr> <tr> <td>S10</td> <td>653.9</td> <td>1660.6</td> <td>0.4x</td> <td>98.9x</td> <td>Find highly variable genes</td> </tr> <tr> <td>S11</td> <td>695.6</td> <td>1642.6</td> <td>0.4x</td> <td>98.9x</td> <td>Find top 2000 highly variable genes</td> </tr> <tr> <td>S12</td> <td>2359.9</td> <td>6418.5</td> <td>0.4x</td> <td>1.5x</td> <td>QC metrics + cell filtering + gene filtering</td> </tr> <tr> <td>S13</td> <td>338.8</td> <td>3417.2</td> <td>0.1x</td> <td>2.0x</td> <td>Normalize total + slice 100x50 submatrix (lazy)</td> </tr> <tr> <td>S14</td> <td>617.9</td> <td>2174.4</td> <td>0.3x</td> <td>2.0x</td> <td>Log1p + slice 200x100 submatrix (lazy)</td> </tr> <tr> <td>S15</td> <td>627.9</td> <td>3510.9</td> <td>0.2x</td> <td>2.0x</td> <td>Normalize + Log1p + slice 500x250 submatrix (lazy)</td> </tr> <tr> <td>S16</td> <td>922.6</td> <td>3061.3</td> <td>0.3x</td> <td>1.9x</td> <td>Normalize + Log1p + mean per gene (lazy)</td> </tr> <tr> <td>S17</td> <td>620.4</td> <td>3174.6</td> <td>0.2x</td> <td>1.8x</td> <td>Normalize + Log1p + variance per cell (lazy)</td> </tr> </tbody> </table> <hr> <p><strong>Key Insight</strong>: Lazy computation enables <strong>complex preprocessing pipelines</strong> that would cause memory explosions with traditional tools. The computation cost is paid when materializing results, but the memory efficiency enables workflows impossible with eager processing. This is similar to Dask's delayed computation patterns.</p> <p><em>For detailed migration guides, see <a href=../slaf_vs_h5ad_benchmarks/ >SLAF vs h5ad Benchmarks</a> and <a href=../slaf_vs_tiledb_benchmarks/ >SLAF vs TileDB Benchmarks</a>.</em></p> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1zM12.5 7v5.2l4 2.4-1 1L11 13V7zM11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="November 19, 2025 18:42:07 UTC">November 19, 2025</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Created> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3z"/></svg> </span> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date" title="July 8, 2025 21:26:46 UTC">July 8, 2025</span> </span> <span class=md-source-file__fact> <span class=md-icon title=Contributors> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 4a4 4 0 0 1 4 4 4 4 0 0 1-4 4 4 4 0 0 1-4-4 4 4 0 0 1 4-4m0 10c4.42 0 8 1.79 8 4v2H4v-2c0-2.21 3.58-4 8-4"/></svg> </span> <nav> Pavan Ramkumar </nav> </span> </aside> </article> </div> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2026 Pavan Ramkumar </div> </div> <div class=md-social> <a href=https://github.com/slaf-project/slaf target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"annotate": null, "base": "../..", "features": ["navigation.sections", "content.code.copy"], "search": "../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script> <script src=../../assets/javascripts/bundle.79ae519e.min.js></script> <script src=../../javascripts/mathjax.js></script> <script src=../../javascripts/theme-reset.js></script> <script src=../../javascripts/copy-enhancement.js></script> <script src=../../javascripts/theme-aware-assets.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> </body> </html>